<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>ANF MetaBioDiv - PARTII_alpha_Diversity_ANF</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ANF MetaBioDiv</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">    
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/rforbeginer.html" rel="" target="">
 <span class="dropdown-text">R for beginners</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/introduction.pdf" rel="" target="">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/dada2.html" rel="" target="_blank">
 <span class="dropdown-text">Analysing metabarcoding data with dada2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/phyloseq.html" rel="" target="_blank">
 <span class="dropdown-text">Introduction to phyloseq</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/preprocessing_dada2.html" rel="" target="">
 <span class="dropdown-text">Preprocessing: dada2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals/preprocessing_phyloseq.html" rel="" target="">
 <span class="dropdown-text">Preprocessing: phyloseq</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/alphadiv1_2025.pdf" rel="" target="_blank">
 <span class="dropdown-text">Alpha diversity</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/Statistic_test_ANF2025.pdf" rel="" target="">
 <span class="dropdown-text">Introduction to univariate statistics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/ampliseq_presentation_ANF.html" rel="" target="_blank">
 <span class="dropdown-text">Introduction to Ampliseq</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/alpha_diversity.html" rel="" target="">
 <span class="dropdown-text">Alpha diversity</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals/ampliseq.html" rel="" target="">
 <span class="dropdown-text">Workflow Ampliseq</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-4" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 4</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-4">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/beta_diversity.pdf" rel="" target="_blank">
 <span class="dropdown-text">Beta diversity</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/beta_diversity.html" rel="" target="">
 <span class="dropdown-text">Beta diversity</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ANF-MetaBioDiv/course-material" rel="" target="_blank"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#prepare-workspace" id="toc-prepare-workspace" class="nav-link active" data-scroll-target="#prepare-workspace"><span class="header-section-number">1</span> Prepare workspace</a>
  <ul class="collapse">
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries"><span class="header-section-number">1.1</span> Load libraries</a></li>
  <li><a href="#load-custom-functions" id="toc-load-custom-functions" class="nav-link" data-scroll-target="#load-custom-functions"><span class="header-section-number">1.2</span> Load custom functions</a></li>
  <li><a href="#define-output-folder" id="toc-define-output-folder" class="nav-link" data-scroll-target="#define-output-folder"><span class="header-section-number">1.3</span> Define output folder</a></li>
  <li><a href="#load-the-data-and-inspect-the-phyloseq-object" id="toc-load-the-data-and-inspect-the-phyloseq-object" class="nav-link" data-scroll-target="#load-the-data-and-inspect-the-phyloseq-object"><span class="header-section-number">1.4</span> Load the data and inspect the phyloseq object</a></li>
  </ul></li>
  <li><a href="#data-structure" id="toc-data-structure" class="nav-link" data-scroll-target="#data-structure"><span class="header-section-number">2</span> Data Structure</a>
  <ul class="collapse">
  <li><a href="#composition-of-our-phyloseq-object-physeq" id="toc-composition-of-our-phyloseq-object-physeq" class="nav-link" data-scroll-target="#composition-of-our-phyloseq-object-physeq"><span class="header-section-number">2.1</span> Composition of our phyloseq object <code>physeq</code></a>
  <ul class="collapse">
  <li><a href="#an-asv-table-with-the-absolute-counts" id="toc-an-asv-table-with-the-absolute-counts" class="nav-link" data-scroll-target="#an-asv-table-with-the-absolute-counts"><span class="header-section-number">2.1.1</span> An ASV table with the absolute counts</a></li>
  <li><a href="#a-metadata-table-with-information-e.g.-physicochemical-categorical-variables-about-samples" id="toc-a-metadata-table-with-information-e.g.-physicochemical-categorical-variables-about-samples" class="nav-link" data-scroll-target="#a-metadata-table-with-information-e.g.-physicochemical-categorical-variables-about-samples"><span class="header-section-number">2.1.2</span> A metadata table with information (e.g.&nbsp;physicochemical, categorical variables) about samples</a></li>
  <li><a href="#a-table-of-taxonomic-classification-level-of-each-asv" id="toc-a-table-of-taxonomic-classification-level-of-each-asv" class="nav-link" data-scroll-target="#a-table-of-taxonomic-classification-level-of-each-asv"><span class="header-section-number">2.1.3</span> A table of taxonomic classification level of each ASV</a></li>
  <li><a href="#a-phylogenetic-tree" id="toc-a-phylogenetic-tree" class="nav-link" data-scroll-target="#a-phylogenetic-tree"><span class="header-section-number">2.1.4</span> A Phylogenetic tree</a></li>
  <li><a href="#a-table-with-the-asv-sequences" id="toc-a-table-with-the-asv-sequences" class="nav-link" data-scroll-target="#a-table-with-the-asv-sequences"><span class="header-section-number">2.1.5</span> A table with the ASV sequences</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#subsampling-normalization" id="toc-subsampling-normalization" class="nav-link" data-scroll-target="#subsampling-normalization"><span class="header-section-number">3</span> Subsampling normalization</a>
  <ul class="collapse">
  <li><a href="#rarefaction-curves" id="toc-rarefaction-curves" class="nav-link" data-scroll-target="#rarefaction-curves"><span class="header-section-number">3.1</span> Rarefaction Curves</a>
  <ul class="collapse">
  <li><a href="#identify-your-minimum-sample-size" id="toc-identify-your-minimum-sample-size" class="nav-link" data-scroll-target="#identify-your-minimum-sample-size"><span class="header-section-number">3.1.1</span> Identify your minimum sample size</a></li>
  <li><a href="#run-rarefaction-curves-using-our-custom-function-ggrare-defined-in-ralpha_diversity.r" id="toc-run-rarefaction-curves-using-our-custom-function-ggrare-defined-in-ralpha_diversity.r" class="nav-link" data-scroll-target="#run-rarefaction-curves-using-our-custom-function-ggrare-defined-in-ralpha_diversity.r"><span class="header-section-number">3.1.2</span> Run rarefaction curves using our custom function <code>ggrare()</code> (defined in <code>R/alpha_diversity.R</code>)</a></li>
  </ul></li>
  <li><a href="#normalization-process-for-alpha-diversity-sub-sampling" id="toc-normalization-process-for-alpha-diversity-sub-sampling" class="nav-link" data-scroll-target="#normalization-process-for-alpha-diversity-sub-sampling"><span class="header-section-number">3.2</span> Normalization process for alpha diversity: sub-sampling</a></li>
  <li><a href="#run-rarefaction-curves-on-normalized-data" id="toc-run-rarefaction-curves-on-normalized-data" class="nav-link" data-scroll-target="#run-rarefaction-curves-on-normalized-data"><span class="header-section-number">3.3</span> Run rarefaction curves on normalized data</a></li>
  <li><a href="#group-separation" id="toc-group-separation" class="nav-link" data-scroll-target="#group-separation"><span class="header-section-number">3.4</span> Group separation</a></li>
  </ul></li>
  <li><a href="#iv-alpha-diversity" id="toc-iv-alpha-diversity" class="nav-link" data-scroll-target="#iv-alpha-diversity"><span class="header-section-number">4</span> IV-Alpha Diversity</a>
  <ul class="collapse">
  <li><a href="#indices" id="toc-indices" class="nav-link" data-scroll-target="#indices"><span class="header-section-number">4.1</span> Indices</a>
  <ul class="collapse">
  <li><a href="#get-taxonomy-based-diversity-indices" id="toc-get-taxonomy-based-diversity-indices" class="nav-link" data-scroll-target="#get-taxonomy-based-diversity-indices"><span class="header-section-number">4.1.1</span> Get taxonomy-based diversity indices</a></li>
  <li><a href="#add-the-alpha-indices-result-to-your-metadata-sample_data-phyloseq-object" id="toc-add-the-alpha-indices-result-to-your-metadata-sample_data-phyloseq-object" class="nav-link" data-scroll-target="#add-the-alpha-indices-result-to-your-metadata-sample_data-phyloseq-object"><span class="header-section-number">4.1.2</span> Add the alpha indices result to your metadata (sample_data) phyloseq object</a></li>
  </ul></li>
  <li><a href="#alpha-diversity-representations" id="toc-alpha-diversity-representations" class="nav-link" data-scroll-target="#alpha-diversity-representations"><span class="header-section-number">4.2</span> Alpha diversity representations</a>
  <ul class="collapse">
  <li><a href="#alpha-representations-using-phyloseqplot_richness" id="toc-alpha-representations-using-phyloseqplot_richness" class="nav-link" data-scroll-target="#alpha-representations-using-phyloseqplot_richness"><span class="header-section-number">4.2.1</span> Alpha representations using phyloseq::plot_richness()</a></li>
  <li><a href="#color-by-group-color-geo-change-sample-name" id="toc-color-by-group-color-geo-change-sample-name" class="nav-link" data-scroll-target="#color-by-group-color-geo-change-sample-name"><span class="header-section-number">4.2.2</span> Color by group: <code>color = Geo</code> &amp; change sample name</a></li>
  <li><a href="#make-box_plot-by-adding-geom_boxplot-function" id="toc-make-box_plot-by-adding-geom_boxplot-function" class="nav-link" data-scroll-target="#make-box_plot-by-adding-geom_boxplot-function"><span class="header-section-number">4.2.3</span> Make box_plot by adding geom_boxplot function</a></li>
  <li><a href="#make-box_plot-geom_boxplot-fill-color-of-boxplot-fill-transparency-with-alpha" id="toc-make-box_plot-geom_boxplot-fill-color-of-boxplot-fill-transparency-with-alpha" class="nav-link" data-scroll-target="#make-box_plot-geom_boxplot-fill-color-of-boxplot-fill-transparency-with-alpha"><span class="header-section-number">4.2.4</span> Make box_plot : geom_boxplot + fill color of boxplot (fill) + transparency (with alpha)</a></li>
  <li><a href="#alpha-representations-using-microbiomeboxplot_alpha-not-shown" id="toc-alpha-representations-using-microbiomeboxplot_alpha-not-shown" class="nav-link" data-scroll-target="#alpha-representations-using-microbiomeboxplot_alpha-not-shown"><span class="header-section-number">4.2.5</span> Alpha representations using <code>Microbiome::boxplot_alpha</code> (not shown)</a></li>
  <li><a href="#alpha-representations-using-ggplot2" id="toc-alpha-representations-using-ggplot2" class="nav-link" data-scroll-target="#alpha-representations-using-ggplot2"><span class="header-section-number">4.2.6</span> Alpha representations using ggplot2</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#statistical-hypothesis-for-alpha-diversity" id="toc-statistical-hypothesis-for-alpha-diversity" class="nav-link" data-scroll-target="#statistical-hypothesis-for-alpha-diversity"><span class="header-section-number">5</span> Statistical hypothesis for alpha diversity</a>
  <ul class="collapse">
  <li><a href="#normality-test-check-the-normal-or-not-normal-distribution-of-your-data-to-choose-the-right-test" id="toc-normality-test-check-the-normal-or-not-normal-distribution-of-your-data-to-choose-the-right-test" class="nav-link" data-scroll-target="#normality-test-check-the-normal-or-not-normal-distribution-of-your-data-to-choose-the-right-test"><span class="header-section-number">5.0.1</span> Normality test: Check the Normal or not normal distribution of your data to choose the right test!</a></li>
  <li><a href="#select-indices-to-test-run-normality-check" id="toc-select-indices-to-test-run-normality-check" class="nav-link" data-scroll-target="#select-indices-to-test-run-normality-check"><span class="header-section-number">5.0.2</span> Select indices to test &amp; run normality check</a></li>
  <li><a href="#anova-parametric-follows-normal-distribution-and-at-least-3-groups" id="toc-anova-parametric-follows-normal-distribution-and-at-least-3-groups" class="nav-link" data-scroll-target="#anova-parametric-follows-normal-distribution-and-at-least-3-groups"><span class="header-section-number">5.1</span> ANOVA: parametric (follows normal distribution) AND at least 3 groups</a>
  <ul class="collapse">
  <li><a href="#alternative-to-bartlett-levene-test-package-car-less-sensitive-to-normality-deviation" id="toc-alternative-to-bartlett-levene-test-package-car-less-sensitive-to-normality-deviation" class="nav-link" data-scroll-target="#alternative-to-bartlett-levene-test-package-car-less-sensitive-to-normality-deviation"><span class="header-section-number">5.1.1</span> Alternative to Bartlett : Levene test (package <code>car</code>), less sensitive to normality deviation</a></li>
  </ul></li>
  <li><a href="#kruskal-wallis-non-parametric-at-least-three-groups" id="toc-kruskal-wallis-non-parametric-at-least-three-groups" class="nav-link" data-scroll-target="#kruskal-wallis-non-parametric-at-least-three-groups"><span class="header-section-number">5.2</span> Kruskal-Wallis: non-parametric &amp; at least three groups</a></li>
  <li><a href="#t-test-parametric-2-groups-i.e-north-vs.-sud" id="toc-t-test-parametric-2-groups-i.e-north-vs.-sud" class="nav-link" data-scroll-target="#t-test-parametric-2-groups-i.e-north-vs.-sud"><span class="header-section-number">5.3</span> T-test: parametric, 2 groups (<em>i.e</em> North Vs. Sud)</a></li>
  <li><a href="#wilcoxon-rank-sum-non-parametric-2-groups" id="toc-wilcoxon-rank-sum-non-parametric-2-groups" class="nav-link" data-scroll-target="#wilcoxon-rank-sum-non-parametric-2-groups"><span class="header-section-number">5.4</span> Wilcoxon rank sum: non-parametric &amp; 2 Groups</a>
  <ul class="collapse">
  <li><a href="#boxplot-representation-with-p-value-information" id="toc-boxplot-representation-with-p-value-information" class="nav-link" data-scroll-target="#boxplot-representation-with-p-value-information"><span class="header-section-number">5.4.1</span> Boxplot representation with p-value information</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#taxonomy-barplot-graph" id="toc-taxonomy-barplot-graph" class="nav-link" data-scroll-target="#taxonomy-barplot-graph"><span class="header-section-number">6</span> Taxonomy: barplot graph</a>
  <ul class="collapse">
  <li><a href="#abundance-transformation" id="toc-abundance-transformation" class="nav-link" data-scroll-target="#abundance-transformation"><span class="header-section-number">6.1</span> Abundance Transformation</a>
  <ul class="collapse">
  <li><a href="#counts-in-percentage-using-phyloseqtransform_sample_counts" id="toc-counts-in-percentage-using-phyloseqtransform_sample_counts" class="nav-link" data-scroll-target="#counts-in-percentage-using-phyloseqtransform_sample_counts"><span class="header-section-number">6.1.1</span> Counts in percentage using <code>phyloseq::transform_sample_counts()</code></a></li>
  <li><a href="#summarise-at-a-given-taxonomic-level-with-phyloseqtax_glom" id="toc-summarise-at-a-given-taxonomic-level-with-phyloseqtax_glom" class="nav-link" data-scroll-target="#summarise-at-a-given-taxonomic-level-with-phyloseqtax_glom"><span class="header-section-number">6.1.2</span> Summarise at a given taxonomic level with <code>phyloseq::tax_glom()</code></a></li>
  <li><a href="#filter-phylum-mean-of-the-line-phyloseqfilter_taxa" id="toc-filter-phylum-mean-of-the-line-phyloseqfilter_taxa" class="nav-link" data-scroll-target="#filter-phylum-mean-of-the-line-phyloseqfilter_taxa"><span class="header-section-number">6.1.3</span> Filter phylum (mean of the line): <code>phyloseq::filter_taxa()</code></a></li>
  <li><a href="#how-to-save-a-table-into-a-file-exemple-of-phylum-taxonomic-table" id="toc-how-to-save-a-table-into-a-file-exemple-of-phylum-taxonomic-table" class="nav-link" data-scroll-target="#how-to-save-a-table-into-a-file-exemple-of-phylum-taxonomic-table"><span class="header-section-number">6.1.4</span> How to save a table into a file: exemple of phylum taxonomic table</a></li>
  <li><a href="#remove-black-lines" id="toc-remove-black-lines" class="nav-link" data-scroll-target="#remove-black-lines"><span class="header-section-number">6.1.5</span> Remove black lines</a></li>
  </ul></li>
  <li><a href="#microbiome-package" id="toc-microbiome-package" class="nav-link" data-scroll-target="#microbiome-package"><span class="header-section-number">6.2</span> Microbiome package</a>
  <ul class="collapse">
  <li><a href="#microbiomeaggregate_taxa" id="toc-microbiomeaggregate_taxa" class="nav-link" data-scroll-target="#microbiomeaggregate_taxa"><span class="header-section-number">6.2.1</span> <code>microbiome::aggregate_taxa()</code></a></li>
  <li><a href="#microbiomeplot_composition" id="toc-microbiomeplot_composition" class="nav-link" data-scroll-target="#microbiomeplot_composition"><span class="header-section-number">6.2.2</span> <code>microbiome::plot_composition()</code></a></li>
  <li><a href="#interactive-barplot-with-plotlyggplotly" id="toc-interactive-barplot-with-plotlyggplotly" class="nav-link" data-scroll-target="#interactive-barplot-with-plotlyggplotly"><span class="header-section-number">6.2.3</span> Interactive barplot with <code>plotly::ggplotly()</code></a></li>
  <li><a href="#how-to-manage-colors-in-barplots" id="toc-how-to-manage-colors-in-barplots" class="nav-link" data-scroll-target="#how-to-manage-colors-in-barplots"><span class="header-section-number">6.2.4</span> How to manage colors in barplots</a></li>
  <li><a href="#build-your-own-palette" id="toc-build-your-own-palette" class="nav-link" data-scroll-target="#build-your-own-palette"><span class="header-section-number">6.2.5</span> Build your own palette</a></li>
  <li><a href="#use-your-palette-in-the-p_order-barplot" id="toc-use-your-palette-in-the-p_order-barplot" class="nav-link" data-scroll-target="#use-your-palette-in-the-p_order-barplot"><span class="header-section-number">6.2.6</span> Use your palette in the p_order barplot</a></li>
  <li><a href="#to-go-even-further-in-chosing-colors" id="toc-to-go-even-further-in-chosing-colors" class="nav-link" data-scroll-target="#to-go-even-further-in-chosing-colors"><span class="header-section-number">6.2.7</span> To go even further in chosing colors</a></li>
  </ul></li>
  <li><a href="#other-data-manipulation-select-specific-taxa-merge-samples" id="toc-other-data-manipulation-select-specific-taxa-merge-samples" class="nav-link" data-scroll-target="#other-data-manipulation-select-specific-taxa-merge-samples"><span class="header-section-number">6.3</span> Other data Manipulation : select specific taxa, merge samples</a>
  <ul class="collapse">
  <li><a href="#select-actinobacteria-and-bacteroidetes-phyloseqsubset_taxa" id="toc-select-actinobacteria-and-bacteroidetes-phyloseqsubset_taxa" class="nav-link" data-scroll-target="#select-actinobacteria-and-bacteroidetes-phyloseqsubset_taxa"><span class="header-section-number">6.3.1</span> Select Actinobacteria AND Bacteroidetes: <code>phyloseq::subset_taxa()</code></a></li>
  <li><a href="#keep-all-with-the-exception-of-a-class-a-genus-etc-e.g.-contamination" id="toc-keep-all-with-the-exception-of-a-class-a-genus-etc-e.g.-contamination" class="nav-link" data-scroll-target="#keep-all-with-the-exception-of-a-class-a-genus-etc-e.g.-contamination"><span class="header-section-number">6.3.2</span> Keep all with the exception of a class, a genus etc (e.g.&nbsp;contamination)</a></li>
  <li><a href="#understand" id="toc-understand" class="nav-link" data-scroll-target="#understand"><span class="header-section-number">6.3.3</span> Understand:</a></li>
  <li><a href="#merge-samples-groups-duplicates-etc" id="toc-merge-samples-groups-duplicates-etc" class="nav-link" data-scroll-target="#merge-samples-groups-duplicates-etc"><span class="header-section-number">6.3.4</span> Merge samples (groups, duplicates etc)</a></li>
  <li><a href="#sample-selection-phyloseqsubset_samples" id="toc-sample-selection-phyloseqsubset_samples" class="nav-link" data-scroll-target="#sample-selection-phyloseqsubset_samples"><span class="header-section-number">6.3.5</span> Sample selection: <code>phyloseq::subset_samples()</code></a></li>
  <li><a href="#alternative-way-phyloseqprune_samples" id="toc-alternative-way-phyloseqprune_samples" class="nav-link" data-scroll-target="#alternative-way-phyloseqprune_samples"><span class="header-section-number">6.3.6</span> Alternative way: <code>phyloseq::prune_samples</code></a></li>
  </ul></li>
  <li><a href="#retrieve-sequences-from-a-phyloseq-object" id="toc-retrieve-sequences-from-a-phyloseq-object" class="nav-link" data-scroll-target="#retrieve-sequences-from-a-phyloseq-object"><span class="header-section-number">6.4</span> Retrieve sequences from a phyloseq object</a>
  <ul class="collapse">
  <li><a href="#one-sequence" id="toc-one-sequence" class="nav-link" data-scroll-target="#one-sequence"><span class="header-section-number">6.4.1</span> One sequence:</a></li>
  <li><a href="#by-name" id="toc-by-name" class="nav-link" data-scroll-target="#by-name"><span class="header-section-number">6.4.2</span> By name</a></li>
  <li><a href="#from-a-selection" id="toc-from-a-selection" class="nav-link" data-scroll-target="#from-a-selection"><span class="header-section-number">6.4.3</span> From a selection</a></li>
  <li><a href="#retrieve-all-sequences" id="toc-retrieve-all-sequences" class="nav-link" data-scroll-target="#retrieve-all-sequences"><span class="header-section-number">6.4.4</span> Retrieve all sequences</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#core-microbiota-analysis" id="toc-core-microbiota-analysis" class="nav-link" data-scroll-target="#core-microbiota-analysis"><span class="header-section-number">7</span> Core microbiota analysis</a>
  <ul class="collapse">
  <li><a href="#which-core-compare-north-south-core-microbiota" id="toc-which-core-compare-north-south-core-microbiota" class="nav-link" data-scroll-target="#which-core-compare-north-south-core-microbiota"><span class="header-section-number">7.0.1</span> Which core? Compare North &amp; South core microbiota</a></li>
  <li><a href="#change-first-column-name-of-taxonomy-rank" id="toc-change-first-column-name-of-taxonomy-rank" class="nav-link" data-scroll-target="#change-first-column-name-of-taxonomy-rank"><span class="header-section-number">7.0.2</span> Change first column name of taxonomy rank</a></li>
  <li><a href="#add-the-lowest-taxonomy-classification" id="toc-add-the-lowest-taxonomy-classification" class="nav-link" data-scroll-target="#add-the-lowest-taxonomy-classification"><span class="header-section-number">7.0.3</span> Add the lowest taxonomy classification</a></li>
  <li><a href="#see-the-transformation-of-tax_table" id="toc-see-the-transformation-of-tax_table" class="nav-link" data-scroll-target="#see-the-transformation-of-tax_table"><span class="header-section-number">7.0.4</span> See the transformation of tax_table</a></li>
  <li><a href="#identify-core-microbiota" id="toc-identify-core-microbiota" class="nav-link" data-scroll-target="#identify-core-microbiota"><span class="header-section-number">7.0.5</span> Identify Core microbiota</a></li>
  <li><a href="#get-core-microbiota-phyloseq-object" id="toc-get-core-microbiota-phyloseq-object" class="nav-link" data-scroll-target="#get-core-microbiota-phyloseq-object"><span class="header-section-number">7.0.6</span> Get core microbiota phyloseq object</a></li>
  <li><a href="#visualise-core-microbiome-with-microbiomeplot_core" id="toc-visualise-core-microbiome-with-microbiomeplot_core" class="nav-link" data-scroll-target="#visualise-core-microbiome-with-microbiomeplot_core"><span class="header-section-number">7.0.7</span> Visualise core microbiome with <code>microbiome::plot_core()</code></a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ANF-MetaBioDiv/course-material/edit/main/practicals/alpha_diversity.Rmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/ANF-MetaBioDiv/course-material/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">PARTII_alpha_Diversity_ANF</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Fabrice Armougom, MIO </p>
             <p>Marc Garel, MIO </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="prepare-workspace" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Prepare workspace</h1>
<section id="load-libraries" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="load-libraries"><span class="header-section-number">1.1</span> Load libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-custom-functions" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="load-custom-functions"><span class="header-section-number">1.2</span> Load custom functions</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">load_all</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Objects listed as exports, but not present in namespace:
• primer_trim</code></pre>
</div>
</div>
</section>
<section id="define-output-folder" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="define-output-folder"><span class="header-section-number">1.3</span> Define output folder</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>output_alpha <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>, <span class="st">"alpha_diversity"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(output_alpha)) <span class="fu">dir.create</span>(output_alpha, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-the-data-and-inspect-the-phyloseq-object" class="level2" data-number="1.4">
<h2 data-number="1.4" class="anchored" data-anchor-id="load-the-data-and-inspect-the-phyloseq-object"><span class="header-section-number">1.4</span> Load the data and inspect the phyloseq object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>physeq <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"asv_table"</span>,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"phyloseq_object_alpha_beta_div.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="data-structure" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Data Structure</h1>
<ul>
<li><span style="color: steelblue;"><font size="5"> Phyloseq object </font></span></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>physeq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 213 taxa and 18 samples ]
sample_data() Sample Data:       [ 18 samples by 21 sample variables ]
tax_table()   Taxonomy Table:    [ 213 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 213 tips and 212 internal nodes ]
refseq()      DNAStringSet:      [ 213 reference sequences ]</code></pre>
</div>
</div>
<section id="composition-of-our-phyloseq-object-physeq" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="composition-of-our-phyloseq-object-physeq"><span class="header-section-number">2.1</span> Composition of our phyloseq object <code>physeq</code></h2>
<section id="an-asv-table-with-the-absolute-counts" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="an-asv-table-with-the-absolute-counts"><span class="header-section-number">2.1.1</span> An ASV table with the absolute counts</h3>
<p>Be careful: Rows are samples, columns are ASVs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>physeq<span class="sc">@</span>otu_table[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>OTU Table:          [10 taxa and 10 samples]
                     taxa are columns
     ASV1 ASV2 ASV3 ASV4 ASV5 ASV6 ASV7 ASV8 ASV9 ASV10
S11B  117   25   85   70   87   40   57   34   41     0
S1B    67    0   23    0   51   48    0    0   27    58
S2B    43    0   35   15   42   52    0    0    0    43
S2S   103   87   76   12   99   43   36   72   46     0
S3B    59    0   32    0   49   73    0    0    0    57
S3S    81   10    0   20   36    0    0    0    0    50
S4B    11    6   38   33   43   46    0    8    0    37
S4S    68    6   38    0   62    0    0   11   30    46
S5B   176   18   62  109    0   35   56   13   33    82
S5S   182    0   36  101   51    0   33    0   29    42</code></pre>
</div>
</div>
</section>
<section id="a-metadata-table-with-information-e.g.-physicochemical-categorical-variables-about-samples" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="a-metadata-table-with-information-e.g.-physicochemical-categorical-variables-about-samples"><span class="header-section-number">2.1.2</span> A metadata table with information (e.g.&nbsp;physicochemical, categorical variables) about samples</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>physeq<span class="sc">@</span>sam_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     SampName   Geo Description groupe Pres PicoEuk Synec Prochloro NanoEuk
S11B     S11B South     South5B    SGF   35    5370 46830       580    6010
S1B       S1B North     North1B    NBF   52     660 32195     10675     955
S2B       S2B North     North2B    NBF   59     890 25480     16595     670
S2S       S2S North     North2S    NBS    0     890 25480     16595     670
S3B       S3B North     North3B    NBF   74     835 13340     25115    1115
S3S       S3S North     North3S    NBS    0     715 26725     16860     890
S4B       S4B North     North4B    NBF   78    2220  3130     29835    2120
S4S       S4S North     North4S    NBS   78    2220  3130     29835    2120
S5B       S5B North     North5B    NBF   42    1620 55780     23795    2555
S5S       S5S North     North5S    NBS    0    1620 56555     22835    2560
S6B       S6B South     South1B    SGF   13    2520 39050       705    3630
S6S       S6S South     South1S    SGS    0    2435 35890       915    3735
S7B       S7B South     South2B    SGF   26       0     0         0    4005
S7S       S7S South     South2S    SGS    0    4535 26545      1340    6585
S8B       S8B South     South3B    SGF   33       0     0         0    5910
S8S       S8S South     South3S    SGS    0    4260 36745       985    5470
S9B       S9B South     South4B    SGF   25    4000 31730       485    4395
S9S       S9S South     South4S    SGS    0    5465 32860       820    5045
     Crypto SiOH4   NO2   NO3   NH4   PO4    NT    PT   Chla       T       S
S11B   1690 3.324 0.083 0.756 0.467 0.115 9.539 4.138 0.0182 23.0308 38.9967
S1B     115 1.813 0.256 0.889 0.324 0.132 9.946 3.565 0.0000 22.7338 37.6204
S2B     395 2.592 0.105 1.125 0.328 0.067 9.378 3.391 0.0000 22.6824 37.6627
S2S     395 3.381 0.231 0.706 0.450 0.109 8.817 3.345 0.0000 22.6854 37.6176
S3B     165 1.438 0.057 1.159 0.369 0.174 8.989 2.568 0.0000 21.5296 37.5549
S3S     200 1.656 0.098 0.794 0.367 0.095 7.847 2.520 0.0000 22.5610 37.5960
S4B     235 2.457 0.099 1.087 0.349 0.137 8.689 3.129 0.0000 18.8515 37.4542
S4S     235 2.457 0.099 1.087 0.349 0.137 8.689 3.129 0.0000 18.8515 37.4542
S5B    1355 2.028 0.103 1.135 0.216 0.128 8.623 3.137 0.0102 24.1905 38.3192
S5S     945 2.669 0.136 0.785 0.267 0.114 9.146 3.062 0.0000 24.1789 38.3213
S6B    1295 2.206 0.249 0.768 0.629 0.236 9.013 3.455 0.0000 22.0197 39.0877
S6S    1300 3.004 0.251 0.927 0.653 0.266 8.776 3.230 0.0134 22.0515 39.0884
S7B    1600 3.016 0.157 0.895 0.491 0.176 8.968 4.116 0.0000 23.6669 38.9699
S7S    1355 1.198 0.165 1.099 0.432 0.180 8.256 3.182 0.0000 23.6814 38.9708
S8B    1590 3.868 0.253 0.567 0.533 0.169 8.395 3.126 0.0000 23.1236 39.0054
S8S    2265 3.639 0.255 0.658 0.665 0.247 8.991 3.843 0.0132 23.3147 38.9885
S9B    1180 3.910 0.107 0.672 0.490 0.134 8.954 4.042 0.0172 22.6306 38.9094
S9S    1545 3.607 0.139 0.644 0.373 0.167 9.817 3.689 0.0062 22.9545 38.7777
     Sigma_t
S11B 26.9631
S1B  26.0046
S2B  26.0521
S2S  26.0137
S3B  26.2987
S3S  26.0332
S4B  26.9415
S4S  26.9415
S5B  26.1037
S5S  26.1065
S6B  27.3241
S6S  27.3151
S7B  26.7536
S7S  26.7488
S8B  26.9423
S8S  26.8713
S9B  27.0131
S9S  26.8172</code></pre>
</div>
</div>
</section>
<section id="a-table-of-taxonomic-classification-level-of-each-asv" class="level3" data-number="2.1.3">
<h3 data-number="2.1.3" class="anchored" data-anchor-id="a-table-of-taxonomic-classification-level-of-each-asv"><span class="header-section-number">2.1.3</span> A table of taxonomic classification level of each ASV</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>physeq<span class="sc">@</span>tax_table[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Taxonomy Table:     [10 taxa by 7 taxonomic ranks]:
      Kingdom    Phylum             Class                 Order             
ASV1  "Bacteria" "Cyanobacteria"    "Cyanobacteriia"      "Synechococcales" 
ASV2  "Bacteria" "Proteobacteria"   "Gammaproteobacteria" "Enterobacterales"
ASV3  "Bacteria" "Proteobacteria"   "Alphaproteobacteria" "SAR11 clade"     
ASV4  "Archaea"  "Thermoplasmatota" "Thermoplasmata"      "Marine Group II" 
ASV5  "Bacteria" "Proteobacteria"   "Alphaproteobacteria" "SAR11 clade"     
ASV6  "Bacteria" "Proteobacteria"   "Alphaproteobacteria" "SAR11 clade"     
ASV7  "Bacteria" "Proteobacteria"   "Alphaproteobacteria" "Rhodospirillales"
ASV8  "Bacteria" "Proteobacteria"   "Gammaproteobacteria" "Enterobacterales"
ASV9  "Bacteria" "Proteobacteria"   "Alphaproteobacteria" "SAR11 clade"     
ASV10 "Bacteria" "Proteobacteria"   "Alphaproteobacteria" "SAR11 clade"     
      Family                    Genus                  Species
ASV1  "Cyanobiaceae"            "Synechococcus CC9902" NA     
ASV2  "Pseudoalteromonadaceae"  "Pseudoalteromonas"    NA     
ASV3  "Clade I"                 "Clade Ia"             NA     
ASV4  NA                        NA                     NA     
ASV5  "Clade I"                 "Clade Ia"             NA     
ASV6  "Clade II"                NA                     NA     
ASV7  "AEGEAN-169 marine group" NA                     NA     
ASV8  "Pseudoalteromonadaceae"  "Pseudoalteromonas"    NA     
ASV9  "Clade I"                 "Clade Ia"             NA     
ASV10 "Clade I"                 "Clade Ia"             NA     </code></pre>
</div>
</div>
</section>
<section id="a-phylogenetic-tree" class="level3" data-number="2.1.4">
<h3 data-number="2.1.4" class="anchored" data-anchor-id="a-phylogenetic-tree"><span class="header-section-number">2.1.4</span> A Phylogenetic tree</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>physeq<span class="sc">@</span>phy_tree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 213 tips and 212 internal nodes.

Tip labels:
  ASV1, ASV2, ASV3, ASV4, ASV5, ASV6, ...

Rooted; includes branch lengths.</code></pre>
</div>
</div>
</section>
<section id="a-table-with-the-asv-sequences" class="level3" data-number="2.1.5">
<h3 data-number="2.1.5" class="anchored" data-anchor-id="a-table-with-the-asv-sequences"><span class="header-section-number">2.1.5</span> A table with the ASV sequences</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>physeq<span class="sc">@</span>refseq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 213:
      width seq                                             names               
  [1]   402 GGAATTTTCCGCAATGGGCGAA...CGAAAGCCAGGGGAGCGAAAGG ASV1
  [2]   425 GGAATATTGCACAATGGGCGCA...CGAAAGCGTGGGGAGCAAACAG ASV2
  [3]   400 GGAATCTTGCACAATGGAGGAA...CGAAAGCATGGGTAGCGAAGAG ASV3
  [4]   383 CGAAAACTTGACAATGCGAGCA...CGAAGCCTAGGGGCACGAACCG ASV4
  [5]   400 GGAATCTTGCACAATGGAGGAA...CGAAAGCATGGGTAGCGAAGAG ASV5
  ...   ... ...
[209]   426 GGAATTTTGCGCAATGGACGAA...CGAAAGCGTGGGGAGCGAACAG ASV209
[210]   403 GGAATATTGCACAATGGGCGCA...GGTCAACACTGACGCTCATGTA ASV210
[211]   360 CGAAAACTTCACACTGCAGGAA...GAACGGATCCGACGGTCAGGGA ASV211
[212]   400 GGAATATTGGACAATGGGCGAA...CGAAAGCGTGGGTAGCAAACAG ASV212
[213]   404 GGAATATTGCACAATGGGCGCA...GTCAACACTGACGCTCATGTAC ASV213</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="subsampling-normalization" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Subsampling normalization</h1>
<section id="rarefaction-curves" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="rarefaction-curves"><span class="header-section-number">3.1</span> Rarefaction Curves</h2>
<p>Before normalization by sub-sampling, let’s have a look at rarefaction curves, evaluate your sequencing effort and make decisions</p>
<section id="identify-your-minimum-sample-size" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="identify-your-minimum-sample-size"><span class="header-section-number">3.1.1</span> Identify your minimum sample size</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">sample_sums</span>(physeq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>S11B  S1B  S2B  S2S  S3B  S3S  S4B  S4S  S5B  S5S  S6B  S6S  S7B  S7S  S8B  S8S 
 975  837  893  983  878  889  917 1077 1018 1006 1076  937  878  936  846  958 
 S9B  S9S 
 888  991 </code></pre>
</div>
</div>
<p><strong>What is the minimum sample size?</strong></p>
</section>
<section id="run-rarefaction-curves-using-our-custom-function-ggrare-defined-in-ralpha_diversity.r" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="run-rarefaction-curves-using-our-custom-function-ggrare-defined-in-ralpha_diversity.r"><span class="header-section-number">3.1.2</span> Run rarefaction curves using our custom function <code>ggrare()</code> (defined in <code>R/alpha_diversity.R</code>)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Make rarefaction curves &amp; Add min sample size line</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggrare</span>(physeq, <span class="at">step =</span> <span class="dv">10</span>, <span class="at">color =</span> <span class="st">"Description"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fu">min</span>(<span class="fu">sample_sums</span>(physeq)), <span class="at">color =</span> <span class="st">"gray60"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 4

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 4</code></pre>
</div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><strong>Do you think is a good idea to normalize your data using this minimal sample size?</strong></p>
</section>
</section>
<section id="normalization-process-for-alpha-diversity-sub-sampling" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="normalization-process-for-alpha-diversity-sub-sampling"><span class="header-section-number">3.2</span> Normalization process for alpha diversity: sub-sampling</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>physeq_rar <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">rarefy_even_depth</span>(physeq, <span class="at">rngseed =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Check the number of sequences for each sample using sample_sums function</strong></p>
<p><strong>Did you lost a lot of ASVs?</strong></p>
</section>
<section id="run-rarefaction-curves-on-normalized-data" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="run-rarefaction-curves-on-normalized-data"><span class="header-section-number">3.3</span> Run rarefaction curves on normalized data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>p0 <span class="ot">&lt;-</span> <span class="fu">ggrare</span>(physeq_rar, <span class="at">step =</span> <span class="dv">10</span>, <span class="at">color =</span> <span class="st">"Description"</span>, <span class="at">se =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2

Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in vegan::rarefy(x[i, , drop = FALSE], n, se = se): most observed count
data have counts 1, but smallest count is 4</code></pre>
</div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="group-separation" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="group-separation"><span class="header-section-number">3.4</span> Group separation</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>p0 <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>Geo, <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="iv-alpha-diversity" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> IV-Alpha Diversity</h1>
<section id="indices" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="indices"><span class="header-section-number">4.1</span> Indices</h2>
<section id="get-taxonomy-based-diversity-indices" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1" class="anchored" data-anchor-id="get-taxonomy-based-diversity-indices"><span class="header-section-number">4.1.1</span> Get taxonomy-based diversity indices</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Get indices with alpha function (NB: index="all" if you want all the indices)</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>alpha_indices <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">alpha</span>(</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  physeq_rar,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">index =</span> <span class="fu">c</span>(<span class="st">"observed"</span>, <span class="st">"diversity_gini_simpson"</span>,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>          <span class="st">"diversity_shannon"</span>, <span class="st">"evenness_pielou"</span>,</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>          <span class="st">"dominance_relative"</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co">#save</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(alpha_indices,</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="fu">file.path</span>(output_alpha, <span class="st">"indices_alpha_resultat.txt"</span>),</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co">#which type?</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(alpha_indices)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#see</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>alpha_indices</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     observed diversity_gini_simpson diversity_shannon evenness_pielou
S11B       36              0.9447863          3.146480       0.8780420
S1B        35              0.9477924          3.177766       0.8937989
S2B        43              0.9577758          3.408022       0.9060997
S2S        36              0.9414576          3.112066       0.8684385
S3B        41              0.9503989          3.275906       0.8821441
S3S        38              0.9491313          3.246802       0.8925706
S4B        40              0.9570249          3.350694       0.9083230
S4S        40              0.9363446          3.097474       0.8396788
S5B        32              0.9271578          2.978541       0.8594252
S5S        33              0.9253250          2.993341       0.8560944
S6B        41              0.9439556          3.213508       0.8653415
S6S        28              0.8247810          2.440971       0.7325395
S7B        36              0.9435901          3.157702       0.8811735
S7S        44              0.9488744          3.278296       0.8663140
S8B        33              0.9452117          3.108474       0.8890226
S8S        39              0.9517578          3.307644       0.9028494
S9B        36              0.9443038          3.105908       0.8667202
S9S        33              0.9487545          3.180393       0.9095912
     dominance_relative
S11B         0.10274791
S1B          0.10274791
S2B          0.09677419
S2S          0.10991637
S3B          0.10991637
S3S          0.10991637
S4B          0.08721625
S4S          0.14336918
S5B          0.16606930
S5S          0.18876941
S6B          0.13620072
S6S          0.38351254
S7B          0.13022700
S7S          0.10633214
S8B          0.09438471
S8S          0.09677419
S9B          0.09438471
S9S          0.10394265</code></pre>
</div>
</div>
<p><strong>What can you notice for one sample?</strong></p>
<p><strong>How to show this graphically?</strong></p>
</section>
<section id="add-the-alpha-indices-result-to-your-metadata-sample_data-phyloseq-object" class="level3" data-number="4.1.2">
<h3 data-number="4.1.2" class="anchored" data-anchor-id="add-the-alpha-indices-result-to-your-metadata-sample_data-phyloseq-object"><span class="header-section-number">4.1.2</span> Add the alpha indices result to your metadata (sample_data) phyloseq object</h3>
<p>Important because many times you will probably want to add new variables in the phyloseq class object!!!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Turn into sample_data object : sample_data function</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>alpha_indices <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(alpha_indices)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co">#See</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(alpha_indices)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sample_data"
attr(,"package")
[1] "phyloseq"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Add alpha_indices to phyloseq sample_data object: merge_phyloseq function!</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>physeq_rar <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">merge_phyloseq</span>(physeq_rar, alpha_indices)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co">#See the result</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_data</span>(physeq_rar)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="alpha-diversity-representations" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="alpha-diversity-representations"><span class="header-section-number">4.2</span> Alpha diversity representations</h2>
<p><em>This section will show you how to plot by different ways the alpha diversity and its customization.</em> <em>Understand how it works!</em></p>
<section id="alpha-representations-using-phyloseqplot_richness" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="alpha-representations-using-phyloseqplot_richness"><span class="header-section-number">4.2.1</span> Alpha representations using phyloseq::plot_richness()</h3>
<p><em>You are limited to the indices calculated by the phyloseq::estimate_richness function</em> <em>(i.e.”Observed”, “Chao1”, “ACE”, “Shannon”, “Simpson”, “InvSimpson”, “Fisher”).</em></p>
<section id="selected-indices-sampname" class="level4" data-number="4.2.1.1">
<h4 data-number="4.2.1.1" class="anchored" data-anchor-id="selected-indices-sampname"><span class="header-section-number">4.2.1.1</span> Selected indices + SampName</h4>
<p><code>x</code> allow you to choose the column from sample_data(physeq_rar) for applying the label</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(physeq_rar, <span class="at">x =</span> <span class="st">"SampName"</span>,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"Simpson"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="color-by-group-color-geo-change-sample-name" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="color-by-group-color-geo-change-sample-name"><span class="header-section-number">4.2.2</span> Color by group: <code>color = Geo</code> &amp; change sample name</h3>
<p>For color option pass the column of <code>sample_data(physeq_rar)</code> that you want. Here different colors is applied depending on <code>Geo</code> (which is North and South, so 2 different colors)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(physeq_rar,</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">"Description"</span>,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">color=</span><span class="st">"Geo"</span>,</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures=</span><span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"Simpson"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="make-box_plot-by-adding-geom_boxplot-function" class="level3" data-number="4.2.3">
<h3 data-number="4.2.3" class="anchored" data-anchor-id="make-box_plot-by-adding-geom_boxplot-function"><span class="header-section-number">4.2.3</span> Make box_plot by adding geom_boxplot function</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(physeq_rar,</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x=</span><span class="st">"Geo"</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">color=</span><span class="st">"Geo"</span>,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures=</span><span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"Simpson"</span>)) <span class="sc">+</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="make-box_plot-geom_boxplot-fill-color-of-boxplot-fill-transparency-with-alpha" class="level3" data-number="4.2.4">
<h3 data-number="4.2.4" class="anchored" data-anchor-id="make-box_plot-geom_boxplot-fill-color-of-boxplot-fill-transparency-with-alpha"><span class="header-section-number">4.2.4</span> Make box_plot : geom_boxplot + fill color of boxplot (fill) + transparency (with alpha)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_richness</span>(physeq_rar,</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">x =</span> <span class="st">"Geo"</span>,</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">measures =</span> <span class="fu">c</span>(<span class="st">"Observed"</span>, <span class="st">"Shannon"</span>, <span class="st">"Simpson"</span>)) <span class="sc">+</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill =</span> Geo), <span class="at">alpha =</span> <span class="fl">0.4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="alpha-representations-using-microbiomeboxplot_alpha-not-shown" class="level3" data-number="4.2.5">
<h3 data-number="4.2.5" class="anchored" data-anchor-id="alpha-representations-using-microbiomeboxplot_alpha-not-shown"><span class="header-section-number">4.2.5</span> Alpha representations using <code>Microbiome::boxplot_alpha</code> (not shown)</h3>
<p>Again, you are limited to the indices calculated by the Microbiome::alpha function</p>
</section>
<section id="alpha-representations-using-ggplot2" class="level3" data-number="4.2.6">
<h3 data-number="4.2.6" class="anchored" data-anchor-id="alpha-representations-using-ggplot2"><span class="header-section-number">4.2.6</span> Alpha representations using ggplot2</h3>
<p>Interest: Freedom!! you can use ANY indices that you have calculated from different packages &amp; included in sample_data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Before : Change your phyloseq class oject sample_data as a dataframe</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">sample_data</span>(physeq_rar))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="boxplot-color-control-points-and-mean-sd-stat_summary" class="level4" data-number="4.2.6.1">
<h4 data-number="4.2.6.1" class="anchored" data-anchor-id="boxplot-color-control-points-and-mean-sd-stat_summary"><span class="header-section-number">4.2.6.1</span> Boxplot, color control, points and Mean SD: <code>stat_summary()</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> Geo, <span class="at">y =</span> observed)) <span class="sc">+</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>),</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">color=</span><span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>)) <span class="sc">+</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">colour =</span> groupe), <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="fl">0.07</span>), <span class="at">cex =</span> <span class="fl">2.2</span>) <span class="sc">+</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> mean, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">17</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">color =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun.data =</span> <span class="st">"mean_se"</span>, <span class="at">geom =</span> <span class="st">"errorbar"</span>, <span class="at">width =</span> .<span class="dv">1</span>, <span class="at">color =</span> <span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="combine-graphs-on-same-figure-patchwork" class="level4" data-number="4.2.6.2">
<h4 data-number="4.2.6.2" class="anchored" data-anchor-id="combine-graphs-on-same-figure-patchwork"><span class="header-section-number">4.2.6.2</span> Combine graphs on same figure: <code>patchwork</code></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Put your  graphs in different variables P1,P2,P3</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> Geo, <span class="at">y =</span> observed)) <span class="sc">+</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>,<span class="st">"#E7B800"</span>),</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">color=</span><span class="fu">c</span>(<span class="st">"#00AFBB"</span>,<span class="st">"#E7B800"</span>)) <span class="sc">+</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">colour =</span> groupe), <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="fl">0.07</span>), <span class="at">cex =</span> <span class="fl">2.2</span>) <span class="sc">+</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> Geo, <span class="at">y =</span> evenness_pielou)) <span class="sc">+</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>,</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>),</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>)) <span class="sc">+</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">colour =</span> groupe), <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="fl">0.07</span>), <span class="at">cex =</span> <span class="fl">2.2</span>) <span class="sc">+</span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> Geo, <span class="at">y =</span> diversity_gini_simpson)) <span class="sc">+</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>,</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>),</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>)) <span class="sc">+</span></span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">colour =</span> groupe), <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="fl">0.07</span>), <span class="at">cex =</span> <span class="fl">2.2</span>) <span class="sc">+</span></span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Put the graph of p1, p2 and p3 on same Figure</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> p3 <span class="sc">+</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_annotation</span>(<span class="at">tag_levels =</span> <span class="st">"A"</span>) <span class="sc">+</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>
</section>
<section id="statistical-hypothesis-for-alpha-diversity" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Statistical hypothesis for alpha diversity</h1>
<section id="normality-test-check-the-normal-or-not-normal-distribution-of-your-data-to-choose-the-right-test" class="level3" data-number="5.0.1">
<h3 data-number="5.0.1" class="anchored" data-anchor-id="normality-test-check-the-normal-or-not-normal-distribution-of-your-data-to-choose-the-right-test"><span class="header-section-number">5.0.1</span> Normality test: Check the Normal or not normal distribution of your data to choose the right test!</h3>
<section id="shapiro-test-h0-null-hypothesis-follows-normal-distribution" class="level4" data-number="5.0.1.1">
<h4 data-number="5.0.1.1" class="anchored" data-anchor-id="shapiro-test-h0-null-hypothesis-follows-normal-distribution"><span class="header-section-number">5.0.1.1</span> Shapiro test: H0 Null Hypothesis: follows Normal distribution!</h4>
<p>Means if p&lt;0.05 -&gt; reject the H0 (so does not follow a normal distribution)</p>
</section>
<section id="q-qplots-compare-your-distribution-with-a-theoretical-normal-distribution" class="level4" data-number="5.0.1.2">
<h4 data-number="5.0.1.2" class="anchored" data-anchor-id="q-qplots-compare-your-distribution-with-a-theoretical-normal-distribution"><span class="header-section-number">5.0.1.2</span> Q-Qplots: Compare your distribution with a theoretical normal distribution</h4>
<p>If your data follow a normal distribution, you’re expecting a linear relationship theoritical vs.&nbsp;experimental</p>
<p>Our custom function <code>indices_normality()</code> (defined in <code>R/alpha_diversity.R</code>) plots the results of Shapiro test as well as Q-Qplots.</p>
</section>
</section>
<section id="select-indices-to-test-run-normality-check" class="level3" data-number="5.0.2">
<h3 data-number="5.0.2" class="anchored" data-anchor-id="select-indices-to-test-run-normality-check"><span class="header-section-number">5.0.2</span> Select indices to test &amp; run normality check</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="sc">|&gt;</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(observed,</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>                diversity_gini_simpson,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>                diversity_shannon,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>                evenness_pielou) <span class="sc">|&gt;</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">indices_normality</span>(<span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><strong>What are your conclusions?</strong></p>
</section>
<section id="anova-parametric-follows-normal-distribution-and-at-least-3-groups" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="anova-parametric-follows-normal-distribution-and-at-least-3-groups"><span class="header-section-number">5.1</span> ANOVA: parametric (follows normal distribution) AND at least 3 groups</h2>
<section id="anova-for-observed-asv-and-4-groups" class="level4" data-number="5.1.0.1">
<h4 data-number="5.1.0.1" class="anchored" data-anchor-id="anova-for-observed-asv-and-4-groups"><span class="header-section-number">5.1.0.1</span> Anova for Observed ASV and 4 groups</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many groups used? See the column "groupe" of metadata:</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">factor</span>(metadata<span class="sc">$</span>groupe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] SGF NBF NBF NBS NBF NBS NBF NBS NBF NBS SGF SGS SGF SGS SGF SGS SGF SGS
Levels: NBF NBS SGF SGS</code></pre>
</div>
</div>
</section>
<section id="variance" class="level4" data-number="5.1.0.2">
<h4 data-number="5.1.0.2" class="anchored" data-anchor-id="variance"><span class="header-section-number">5.1.0.2</span> Variance</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check homogeneity of variance between groups</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (avoid bias in ANOVA result &amp; keep the power of the test)</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="co"># H0= equality of variances in the different populations</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>stats<span class="sc">::</span><span class="fu">bartlett.test</span>(observed <span class="sc">~</span> groupe, metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  observed by groupe
Bartlett's K-squared = 3.1798, df = 3, p-value = 0.3647</code></pre>
</div>
</div>
<p><strong>Conclusion?</strong></p>
</section>
<section id="alternative-to-bartlett-levene-test-package-car-less-sensitive-to-normality-deviation" class="level3" data-number="5.1.1">
<h3 data-number="5.1.1" class="anchored" data-anchor-id="alternative-to-bartlett-levene-test-package-car-less-sensitive-to-normality-deviation"><span class="header-section-number">5.1.1</span> Alternative to Bartlett : Levene test (package <code>car</code>), less sensitive to normality deviation</h3>
<p>Global Test: Anova tell you if that some of the group means are different, but you don’t know which pairs of groups are different!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>aov_observed <span class="ot">&lt;-</span> stats<span class="sc">::</span><span class="fu">aov</span>(observed <span class="sc">~</span> groupe, metadata)</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(aov_observed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Df Sum Sq Mean Sq F value Pr(&gt;F)
groupe       3  13.03   4.343   0.211  0.887
Residuals   14 288.75  20.625               </code></pre>
</div>
</div>
<section id="which-pairs-of-groups-are-different---post-hoc-test-tukey-multiple-pairwise-comparisons" class="level4" data-number="5.1.1.1">
<h4 data-number="5.1.1.1" class="anchored" data-anchor-id="which-pairs-of-groups-are-different---post-hoc-test-tukey-multiple-pairwise-comparisons"><span class="header-section-number">5.1.1.1</span> Which pairs of groups are different? -&gt; Post-hoc test: Tukey multiple pairwise-comparisons</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>signif_pairgroups <span class="ot">&lt;-</span> stats<span class="sc">::</span><span class="fu">TukeyHSD</span>(aov_observed, <span class="at">method =</span> <span class="st">"bh"</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>signif_pairgroups</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: stats::aov(formula = observed ~ groupe, data = metadata)

$groupe
         diff        lwr      upr     p adj
NBS-NBF -1.45 -10.304898 7.404898 0.9631679
SGF-NBF -1.80 -10.148478 6.548478 0.9217657
SGS-NBF -2.20 -11.054898 6.654898 0.8866424
SGF-NBS -0.35  -9.204898 8.504898 0.9994302
SGS-NBS -0.75 -10.083882 8.583882 0.9953019
SGS-SGF -0.40  -9.254898 8.454898 0.9991510</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="kruskal-wallis-non-parametric-at-least-three-groups" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="kruskal-wallis-non-parametric-at-least-three-groups"><span class="header-section-number">5.2</span> Kruskal-Wallis: non-parametric &amp; at least three groups</h2>
<section id="kruskal-for-diversity_shannon-and-4-groups" class="level4" data-number="5.2.0.1">
<h4 data-number="5.2.0.1" class="anchored" data-anchor-id="kruskal-for-diversity_shannon-and-4-groups"><span class="header-section-number">5.2.0.1</span> Kruskal for diversity_shannon and 4 groups</h4>
<p>Global test</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>stats<span class="sc">::</span><span class="fu">kruskal.test</span>(diversity_shannon <span class="sc">~</span> groupe, <span class="at">data =</span> metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Kruskal-Wallis rank sum test

data:  diversity_shannon by groupe
Kruskal-Wallis chi-squared = 2.9544, df = 3, p-value = 0.3987</code></pre>
</div>
</div>
</section>
<section id="post-hoc-test-dunn-test-pairwise-group-test" class="level4" data-number="5.2.0.2">
<h4 data-number="5.2.0.2" class="anchored" data-anchor-id="post-hoc-test-dunn-test-pairwise-group-test"><span class="header-section-number">5.2.0.2</span> Post hoc test: Dunn test (pairwise group test)</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>signifgroup <span class="ot">&lt;-</span> FSA<span class="sc">::</span><span class="fu">dunnTest</span>(diversity_shannon <span class="sc">~</span> groupe,</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">data =</span> metadata,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">method =</span> <span class="st">"bh"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: groupe was coerced to a factor.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#See</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>signifgroup</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Comparison          Z   P.unadj     P.adj
1  NBF - NBS  1.5218359 0.1280502 0.7683013
2  NBF - SGF  1.2439326 0.2135244 0.6405731
3  NBS - SGF -0.3490449 0.7270556 0.7270556
4  NBF - SGS  0.4048921 0.6855568 0.8226682
5  NBS - SGS -1.0596259 0.2893148 0.5786297
6  SGF - SGS -0.7678988 0.4425473 0.6638209</code></pre>
</div>
</div>
</section>
</section>
<section id="t-test-parametric-2-groups-i.e-north-vs.-sud" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="t-test-parametric-2-groups-i.e-north-vs.-sud"><span class="header-section-number">5.3</span> T-test: parametric, 2 groups (<em>i.e</em> North Vs. Sud)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>stats<span class="sc">::</span><span class="fu">bartlett.test</span>(observed <span class="sc">~</span> Geo, metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  observed by Geo
Bartlett's K-squared = 0.38191, df = 1, p-value = 0.5366</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>observed_ttest <span class="ot">&lt;-</span> stats<span class="sc">::</span><span class="fu">t.test</span>(observed <span class="sc">~</span> Geo, <span class="at">data =</span> metadata)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co">#see</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>observed_ttest</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  observed by Geo
t = 0.66008, df = 15.246, p-value = 0.5191
alternative hypothesis: true difference in means between group North and group South is not equal to 0
95 percent confidence interval:
 -2.966072  5.632739
sample estimates:
mean in group North mean in group South 
           37.55556            36.22222 </code></pre>
</div>
</div>
</section>
<section id="wilcoxon-rank-sum-non-parametric-2-groups" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="wilcoxon-rank-sum-non-parametric-2-groups"><span class="header-section-number">5.4</span> Wilcoxon rank sum: non-parametric &amp; 2 Groups</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>pairwise_test <span class="ot">&lt;-</span> ggpubr<span class="sc">::</span><span class="fu">compare_means</span>(diversity_shannon <span class="sc">~</span> Geo,</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>                                       metadata,</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">method =</span> <span class="st">"wilcox.test"</span>)</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="co">#See</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>pairwise_test</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 8
  .y.               group1 group2     p p.adj p.format p.signif method  
  &lt;chr&gt;             &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   
1 diversity_shannon South  North  0.863  0.86 0.86     ns       Wilcoxon</code></pre>
</div>
</div>
<section id="boxplot-representation-with-p-value-information" class="level3" data-number="5.4.1">
<h3 data-number="5.4.1" class="anchored" data-anchor-id="boxplot-representation-with-p-value-information"><span class="header-section-number">5.4.1</span> Boxplot representation with p-value information</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Boxplot as previously seen</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>graph_shan <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(metadata, <span class="fu">aes</span>(<span class="at">x =</span> Geo, <span class="at">y =</span> diversity_shannon)) <span class="sc">+</span> </span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha=</span><span class="fl">0.6</span>,</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>),</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="fu">c</span>(<span class="st">"#00AFBB"</span>, <span class="st">"#E7B800"</span>)) <span class="sc">+</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(<span class="at">colour =</span> groupe),</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">position =</span> <span class="fu">position_jitter</span>(<span class="fl">0.02</span>) ,</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">cex=</span><span class="fl">2.2</span>)<span class="sc">+</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> mean, <span class="at">geom =</span> <span class="st">"point"</span>,</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">shape =</span> <span class="dv">17</span>, <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="st">"white"</span>)</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Add p-value on graph</span></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>graph_shan <span class="sc">+</span> ggpubr<span class="sc">::</span><span class="fu">stat_pvalue_manual</span>(</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>  pairwise_test,</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">y.position =</span> <span class="fl">3.5</span>,</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="st">"p.adj = {p.adj}"</span>,</span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="st">"blue"</span>,</span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">linetype =</span> <span class="dv">1</span>,</span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">tip.length =</span> <span class="fl">0.01</span></span>
<span id="cb68-21"><a href="#cb68-21" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>
<section id="taxonomy-barplot-graph" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Taxonomy: barplot graph</h1>
<section id="abundance-transformation" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="abundance-transformation"><span class="header-section-number">6.1</span> Abundance Transformation</h2>
<section id="counts-in-percentage-using-phyloseqtransform_sample_counts" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="counts-in-percentage-using-phyloseqtransform_sample_counts"><span class="header-section-number">6.1.1</span> Counts in percentage using <code>phyloseq::transform_sample_counts()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>pourcentS <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">transform_sample_counts</span>(physeq_rar, <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x) <span class="sc">*</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>See plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_bar</span>(pourcentS)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>What are the separation lines?</p>
</section>
<section id="summarise-at-a-given-taxonomic-level-with-phyloseqtax_glom" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="summarise-at-a-given-taxonomic-level-with-phyloseqtax_glom"><span class="header-section-number">6.1.2</span> Summarise at a given taxonomic level with <code>phyloseq::tax_glom()</code></h3>
<p>Remember ranks can be obtained with <code>phyloseq::rank_names()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">rank_names</span>(pourcentS)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>Phylum_glom <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">tax_glom</span>(pourcentS,</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">taxrank =</span> <span class="st">"Phylum"</span>,</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">NArm =</span> <span class="cn">FALSE</span>)</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot at Phylum taxonomic rank, with color</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_bar</span>(Phylum_glom, <span class="at">fill =</span> <span class="st">"Phylum"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>NArm?</em></p>
</section>
<section id="filter-phylum-mean-of-the-line-phyloseqfilter_taxa" class="level3" data-number="6.1.3">
<h3 data-number="6.1.3" class="anchored" data-anchor-id="filter-phylum-mean-of-the-line-phyloseqfilter_taxa"><span class="header-section-number">6.1.3</span> Filter phylum (mean of the line): <code>phyloseq::filter_taxa()</code></h3>
<p>Let’s filter out the phylums with a mean relative abundance inferior to 1%</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>Phylum_1 <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">filter_taxa</span>(Phylum_glom,</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">flist =</span> <span class="cf">function</span>(x) <span class="fu">mean</span>(x) <span class="sc">&gt;=</span> <span class="dv">1</span>,</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">prune =</span> <span class="cn">TRUE</span>)</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot at Phylum taxonomic rank, with color</span></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_bar</span>(Phylum_1, <span class="at">fill =</span> <span class="st">"Phylum"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="how-to-save-a-table-into-a-file-exemple-of-phylum-taxonomic-table" class="level3" data-number="6.1.4">
<h3 data-number="6.1.4" class="anchored" data-anchor-id="how-to-save-a-table-into-a-file-exemple-of-phylum-taxonomic-table"><span class="header-section-number">6.1.4</span> How to save a table into a file: exemple of phylum taxonomic table</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">df_export</span>(<span class="fu">otu_table</span>(Phylum_glom)),</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">row.names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="fu">file.path</span>(output_alpha, <span class="st">"Phylum_pourcent.tsv"</span>),</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-black-lines" class="level3" data-number="6.1.5">
<h3 data-number="6.1.5" class="anchored" data-anchor-id="remove-black-lines"><span class="header-section-number">6.1.5</span> Remove black lines</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_bar</span>(Phylum_glom, <span class="st">"Description"</span>, <span class="at">fill =</span> <span class="st">"Phylum"</span>) <span class="sc">+</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">colour =</span> Phylum), <span class="at">stat =</span> <span class="st">"identity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="microbiome-package" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="microbiome-package"><span class="header-section-number">6.2</span> Microbiome package</h2>
<section id="microbiomeaggregate_taxa" class="level3" data-number="6.2.1">
<h3 data-number="6.2.1" class="anchored" data-anchor-id="microbiomeaggregate_taxa"><span class="header-section-number">6.2.1</span> <code>microbiome::aggregate_taxa()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Order Rank</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>Order_microb <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">aggregate_taxa</span>(pourcentS, <span class="st">"Order"</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Filter at 1%</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>Order1 <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">filter_taxa</span>(Order_microb, <span class="cf">function</span>(x) <span class="fu">mean</span>(x) <span class="sc">&gt;=</span> <span class="dv">1</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="microbiomeplot_composition" class="level3" data-number="6.2.2">
<h3 data-number="6.2.2" class="anchored" data-anchor-id="microbiomeplot_composition"><span class="header-section-number">6.2.2</span> <code>microbiome::plot_composition()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>p_order <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(Order1,</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">otu.sort =</span> <span class="st">"abundance"</span>,</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">sample.sort =</span> <span class="st">"Description"</span>,</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">x.label =</span> <span class="st">"Description"</span>,</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">plot.type =</span> <span class="st">"barplot"</span>,</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">verbose =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"Relative abundance (%)"</span>)</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="co">#see</span></span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a>p_order</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Average by group :average_by option</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>p_order_groupe <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">plot_composition</span>(Order1,</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">otu.sort =</span> <span class="st">"abundance"</span>,</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">sample.sort =</span> <span class="st">"Description"</span>,</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">x.label =</span> <span class="st">"Description"</span>,</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">plot.type =</span> <span class="st">"barplot"</span>,</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">verbose =</span> <span class="cn">FALSE</span>,</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">average_by =</span> <span class="st">"Geo"</span>) <span class="sc">+</span></span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"Relative abundance (%)"</span>)</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a><span class="co">#see</span></span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a>p_order_groupe</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="interactive-barplot-with-plotlyggplotly" class="level3" data-number="6.2.3">
<h3 data-number="6.2.3" class="anchored" data-anchor-id="interactive-barplot-with-plotlyggplotly"><span class="header-section-number">6.2.3</span> Interactive barplot with <code>plotly::ggplotly()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>plotly<span class="sc">::</span><span class="fu">ggplotly</span>(p_order)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="how-to-manage-colors-in-barplots" class="level3" data-number="6.2.4">
<h3 data-number="6.2.4" class="anchored" data-anchor-id="how-to-manage-colors-in-barplots"><span class="header-section-number">6.2.4</span> How to manage colors in barplots</h3>
<p>With the number of Phyla, Order etc a barplot can become very confusing… Need to have distinct color for each taxonomic groups.</p>
<p>Use the library <code>RColorBrewer</code> et <code>scale_fill_manual()</code> See <a href="https://www.nceas.ucsb.edu/sites/default/files/2020-04/colorPaletteCheatsheet.pdf">here</a> to understand the possibilities</p>
<p>You can visualise <code>RColorBrewer</code>’s palettes with the following command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>RColorBrewer<span class="sc">::</span><span class="fu">display.brewer.all</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="build-your-own-palette" class="level3" data-number="6.2.5">
<h3 data-number="6.2.5" class="anchored" data-anchor-id="build-your-own-palette"><span class="header-section-number">6.2.5</span> Build your own palette</h3>
<p>Let’s assemble from two RColorBrewer’s palettes a single 13 colors palette</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co">#See Set2 colors</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>(col1 <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">name =</span> <span class="st">"Set2"</span>, <span class="at">n =</span> <span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "#66C2A5" "#FC8D62" "#8DA0CB" "#E78AC3" "#A6D854" "#FFD92F" "#E5C494"
[8] "#B3B3B3"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">#See  Paired colors </span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>(col2 <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">name =</span> <span class="st">"Paired"</span>, <span class="at">n =</span> <span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "#A6CEE3" "#1F78B4" "#B2DF8A" "#33A02C" "#FB9A99"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Build your set of colors using brewer.pal or your own colors</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>mycolors <span class="ot">&lt;-</span> <span class="fu">c</span>(col1, col2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="use-your-palette-in-the-p_order-barplot" class="level3" data-number="6.2.6">
<h3 data-number="6.2.6" class="anchored" data-anchor-id="use-your-palette-in-the-p_order-barplot"><span class="header-section-number">6.2.6</span> Use your palette in the p_order barplot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Use scale_fill_manual</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>p_order <span class="sc">+</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_fill_manual</span>(<span class="st">"Order"</span>, <span class="at">values =</span> mycolors) <span class="sc">+</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="to-go-even-further-in-chosing-colors" class="level3" data-number="6.2.7">
<h3 data-number="6.2.7" class="anchored" data-anchor-id="to-go-even-further-in-chosing-colors"><span class="header-section-number">6.2.7</span> To go even further in chosing colors</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>order_color <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"SAR11 clade"</span> <span class="ot">=</span> <span class="st">"cyan3"</span>,</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Enterobacterales"</span> <span class="ot">=</span> <span class="st">"#33A02C"</span>,      <span class="co"># lime green</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Synechococcales"</span> <span class="ot">=</span> <span class="st">"#8DA0CB"</span>,       <span class="co"># light steel blue</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Rhodospirillales"</span><span class="ot">=</span><span class="st">"pink"</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Marine Group II"</span><span class="ot">=</span><span class="st">"red"</span>,</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Actinomarinales"</span><span class="ot">=</span><span class="st">"purple"</span>,</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Pseudomonadales"</span> <span class="ot">=</span> <span class="st">"#8B0000"</span>,       <span class="co"># dark red</span></span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Unknown"</span><span class="ot">=</span><span class="st">"#B3B3B3"</span>,</span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Flavobacteriales"</span> <span class="ot">=</span> <span class="st">"gold2"</span>,</span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Chloroplast"</span> <span class="ot">=</span> <span class="st">"#FDBF6F"</span>,       <span class="co"># light orange</span></span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Rhodobacterales"</span> <span class="ot">=</span> <span class="st">"darkorange1"</span>    <span class="co"># dark orange</span></span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a>p_order <span class="sc">+</span></span>
<span id="cb88-16"><a href="#cb88-16" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">scale_fill_manual</span>(<span class="st">"Order"</span>, <span class="at">values =</span> order_color) <span class="sc">+</span></span>
<span id="cb88-17"><a href="#cb88-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>,</span>
<span id="cb88-18"><a href="#cb88-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="other-data-manipulation-select-specific-taxa-merge-samples" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="other-data-manipulation-select-specific-taxa-merge-samples"><span class="header-section-number">6.3</span> Other data Manipulation : select specific taxa, merge samples</h2>
<section id="select-actinobacteria-and-bacteroidetes-phyloseqsubset_taxa" class="level3" data-number="6.3.1">
<h3 data-number="6.3.1" class="anchored" data-anchor-id="select-actinobacteria-and-bacteroidetes-phyloseqsubset_taxa"><span class="header-section-number">6.3.1</span> Select Actinobacteria AND Bacteroidetes: <code>phyloseq::subset_taxa()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>myselection1 <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_taxa</span>(Phylum_glom, Phylum <span class="sc">==</span> <span class="st">"Actinobacteriota"</span> <span class="sc">|</span> Phylum <span class="sc">==</span> <span class="st">"Bacteroidota"</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_bar</span>(myselection1, <span class="at">x =</span> <span class="st">"Description"</span>, <span class="at">fill =</span> <span class="st">"Phylum"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">plot_bar</span>(myselection1, <span class="at">x =</span> <span class="st">"Description"</span>,</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fill=</span><span class="st">"Phylum"</span>, <span class="at">facet_grid =</span> <span class="sc">~</span>Phylum) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="keep-all-with-the-exception-of-a-class-a-genus-etc-e.g.-contamination" class="level3" data-number="6.3.2">
<h3 data-number="6.3.2" class="anchored" data-anchor-id="keep-all-with-the-exception-of-a-class-a-genus-etc-e.g.-contamination"><span class="header-section-number">6.3.2</span> Keep all with the exception of a class, a genus etc (e.g.&nbsp;contamination)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>myselection2 <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_taxa</span>(physeq, Class <span class="sc">!=</span> <span class="st">"Thermoplasmata"</span> <span class="sc">|</span> <span class="fu">is.na</span>(Class))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="understand" class="level3" data-number="6.3.3">
<h3 data-number="6.3.3" class="anchored" data-anchor-id="understand"><span class="header-section-number">6.3.3</span> Understand:</h3>
<p><span style="color: red;">!</span> <span style="color: steelblue;"><strong>= means IS NOT</strong></span></p>
<p><span style="color: red;">|</span> <span style="color: steelblue;"><strong>= AND</strong></span></p>
<p><span style="color: red;">Is.na</span> <span style="color: steelblue;"><strong>= do not remove the NA (Not Assigned at the Class rank), by default it will be removed. be careful</strong>!</span></p>
</section>
<section id="merge-samples-groups-duplicates-etc" class="level3" data-number="6.3.4">
<h3 data-number="6.3.4" class="anchored" data-anchor-id="merge-samples-groups-duplicates-etc"><span class="header-section-number">6.3.4</span> Merge samples (groups, duplicates etc)</h3>
<p>Use a column from metadata to group/merge samples (North &amp; South)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>(NordSud <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">merge_samples</span>(physeq_rar, <span class="st">"Geo"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in asMethod(object): NAs introduced by coercion

Warning in asMethod(object): NAs introduced by coercion

Warning in asMethod(object): NAs introduced by coercion

Warning in asMethod(object): NAs introduced by coercion</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 209 taxa and 2 samples ]
sample_data() Sample Data:       [ 2 samples by 26 sample variables ]
tax_table()   Taxonomy Table:    [ 209 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 209 tips and 208 internal nodes ]</code></pre>
</div>
</div>
</section>
<section id="sample-selection-phyloseqsubset_samples" class="level3" data-number="6.3.5">
<h3 data-number="6.3.5" class="anchored" data-anchor-id="sample-selection-phyloseqsubset_samples"><span class="header-section-number">6.3.5</span> Sample selection: <code>phyloseq::subset_samples()</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>(sub_North <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pourcentS, Geo <span class="sc">==</span> <span class="st">"North"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 209 taxa and 9 samples ]
sample_data() Sample Data:       [ 9 samples by 26 sample variables ]
tax_table()   Taxonomy Table:    [ 209 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 209 tips and 208 internal nodes ]
refseq()      DNAStringSet:      [ 209 reference sequences ]</code></pre>
</div>
</div>
</section>
<section id="alternative-way-phyloseqprune_samples" class="level3" data-number="6.3.6">
<h3 data-number="6.3.6" class="anchored" data-anchor-id="alternative-way-phyloseqprune_samples"><span class="header-section-number">6.3.6</span> Alternative way: <code>phyloseq::prune_samples</code></h3>
<p>Define what you want to keep</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"S1B"</span>, <span class="st">"S2S"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then extract these samples from pourcentS phyloseq object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>keep2samples <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">prune_samples</span>(keep, pourcentS)</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sample_names</span>(keep2samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "S1B" "S2S"</code></pre>
</div>
</div>
</section>
</section>
<section id="retrieve-sequences-from-a-phyloseq-object" class="level2" data-number="6.4">
<h2 data-number="6.4" class="anchored" data-anchor-id="retrieve-sequences-from-a-phyloseq-object"><span class="header-section-number">6.4</span> Retrieve sequences from a phyloseq object</h2>
<section id="one-sequence" class="level3" data-number="6.4.1">
<h3 data-number="6.4.1" class="anchored" data-anchor-id="one-sequence"><span class="header-section-number">6.4.1</span> One sequence:</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">writeXStringSet</span>(physeq_rar<span class="sc">@</span>refseq[<span class="st">"ASV1"</span>],</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">filepath =</span> <span class="fu">file.path</span>(output_alpha,<span class="st">"ASV1.fasta"</span>),</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">format =</span> <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="by-name" class="level3" data-number="6.4.2">
<h3 data-number="6.4.2" class="anchored" data-anchor-id="by-name"><span class="header-section-number">6.4.2</span> By name</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>listASV <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ASV2"</span>, <span class="st">"ASV8"</span>, <span class="st">"ASV32"</span>, <span class="st">"ASV58"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">writeXStringSet</span>(physeq_rar<span class="sc">@</span>refseq[listASV],</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">filepath =</span> <span class="fu">file.path</span>(output_alpha,<span class="st">"several_asvs.fasta"</span>),</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">format =</span> <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="from-a-selection" class="level3" data-number="6.4.3">
<h3 data-number="6.4.3" class="anchored" data-anchor-id="from-a-selection"><span class="header-section-number">6.4.3</span> From a selection</h3>
<p>Let’s export a fasta files of all ASVs with a maximum relative abundance superior to 10% in North samples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pourcentS, Geo <span class="sc">==</span> <span class="st">"North"</span>) <span class="sc">|&gt;</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  phyloseq<span class="sc">::</span><span class="fu">filter_taxa</span>(<span class="at">flist =</span> <span class="cf">function</span>(x) <span class="fu">max</span>(x) <span class="sc">&gt;=</span> <span class="dv">10</span>, <span class="at">prune =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  phyloseq<span class="sc">::</span><span class="fu">refseq</span>() <span class="sc">|&gt;</span></span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>  Biostrings<span class="sc">::</span><span class="fu">writeXStringSet</span>(</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">filepath =</span> <span class="fu">file.path</span>(output_alpha, <span class="st">"fancy_selection_asvs.fasta"</span>),</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">format =</span> <span class="st">"fasta"</span></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="retrieve-all-sequences" class="level3" data-number="6.4.4">
<h3 data-number="6.4.4" class="anchored" data-anchor-id="retrieve-all-sequences"><span class="header-section-number">6.4.4</span> Retrieve all sequences</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">writeXStringSet</span>(physeq_rar<span class="sc">@</span>refseq,</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">filepath =</span> <span class="fu">file.path</span>(output_alpha,<span class="st">"all_asvs.fasta"</span>),</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">format =</span> <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="core-microbiota-analysis" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Core microbiota analysis</h1>
<p><strong>Identify the taxa names of the core microbiota</strong></p>
<section id="which-core-compare-north-south-core-microbiota" class="level3" data-number="7.0.1">
<h3 data-number="7.0.1" class="anchored" data-anchor-id="which-core-compare-north-south-core-microbiota"><span class="header-section-number">7.0.1</span> Which core? Compare North &amp; South core microbiota</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create 2 phyloseq objects for North and South sample groups</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>sub_North <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pourcentS, Geo <span class="sc">==</span> <span class="st">"North"</span>)</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>sub_South <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">subset_samples</span>(pourcentS, Geo <span class="sc">==</span> <span class="st">"South"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Check group North ok</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>sub_North<span class="sc">@</span>sam_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    SampName   Geo Description groupe Pres PicoEuk Synec Prochloro NanoEuk
S1B      S1B North     North1B    NBF   52     660 32195     10675     955
S2B      S2B North     North2B    NBF   59     890 25480     16595     670
S2S      S2S North     North2S    NBS    0     890 25480     16595     670
S3B      S3B North     North3B    NBF   74     835 13340     25115    1115
S3S      S3S North     North3S    NBS    0     715 26725     16860     890
S4B      S4B North     North4B    NBF   78    2220  3130     29835    2120
S4S      S4S North     North4S    NBS   78    2220  3130     29835    2120
S5B      S5B North     North5B    NBF   42    1620 55780     23795    2555
S5S      S5S North     North5S    NBS    0    1620 56555     22835    2560
    Crypto SiOH4   NO2   NO3   NH4   PO4    NT    PT   Chla       T       S
S1B    115 1.813 0.256 0.889 0.324 0.132 9.946 3.565 0.0000 22.7338 37.6204
S2B    395 2.592 0.105 1.125 0.328 0.067 9.378 3.391 0.0000 22.6824 37.6627
S2S    395 3.381 0.231 0.706 0.450 0.109 8.817 3.345 0.0000 22.6854 37.6176
S3B    165 1.438 0.057 1.159 0.369 0.174 8.989 2.568 0.0000 21.5296 37.5549
S3S    200 1.656 0.098 0.794 0.367 0.095 7.847 2.520 0.0000 22.5610 37.5960
S4B    235 2.457 0.099 1.087 0.349 0.137 8.689 3.129 0.0000 18.8515 37.4542
S4S    235 2.457 0.099 1.087 0.349 0.137 8.689 3.129 0.0000 18.8515 37.4542
S5B   1355 2.028 0.103 1.135 0.216 0.128 8.623 3.137 0.0102 24.1905 38.3192
S5S    945 2.669 0.136 0.785 0.267 0.114 9.146 3.062 0.0000 24.1789 38.3213
    Sigma_t observed diversity_gini_simpson diversity_shannon evenness_pielou
S1B 26.0046       35              0.9477924          3.177766       0.8937989
S2B 26.0521       43              0.9577758          3.408022       0.9060997
S2S 26.0137       36              0.9414576          3.112066       0.8684385
S3B 26.2987       41              0.9503989          3.275906       0.8821441
S3S 26.0332       38              0.9491313          3.246802       0.8925706
S4B 26.9415       40              0.9570249          3.350694       0.9083230
S4S 26.9415       40              0.9363446          3.097474       0.8396788
S5B 26.1037       32              0.9271578          2.978541       0.8594252
S5S 26.1065       33              0.9253250          2.993341       0.8560944
    dominance_relative
S1B         0.10274791
S2B         0.09677419
S2S         0.10991637
S3B         0.10991637
S3S         0.10991637
S4B         0.08721625
S4S         0.14336918
S5B         0.16606930
S5S         0.18876941</code></pre>
</div>
</div>
</section>
<section id="change-first-column-name-of-taxonomy-rank" class="level3" data-number="7.0.2">
<h3 data-number="7.0.2" class="anchored" data-anchor-id="change-first-column-name-of-taxonomy-rank"><span class="header-section-number">7.0.2</span> Change first column name of taxonomy rank</h3>
<p>Replace “Kingdom” by “Domain”, needed for the use of add_best function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Before</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sub_North<span class="sc">@</span>tax_table)[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Kingdom"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Apply change for North</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sub_North<span class="sc">@</span>tax_table)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"Domain"</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="co">#See </span></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sub_North<span class="sc">@</span>tax_table)[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Domain"</code></pre>
</div>
</div>
</section>
<section id="add-the-lowest-taxonomy-classification" class="level3" data-number="7.0.3">
<h3 data-number="7.0.3" class="anchored" data-anchor-id="add-the-lowest-taxonomy-classification"><span class="header-section-number">7.0.3</span> Add the lowest taxonomy classification</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>sub_North <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">add_besthit</span>(sub_North, <span class="at">sep =</span> <span class="st">":"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="see-the-transformation-of-tax_table" class="level3" data-number="7.0.4">
<h3 data-number="7.0.4" class="anchored" data-anchor-id="see-the-transformation-of-tax_table"><span class="header-section-number">7.0.4</span> See the transformation of tax_table</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sub_North<span class="sc">@</span>tax_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
                          Domain     Phylum             Class                
ASV1:Synechococcus CC9902 "Bacteria" "Cyanobacteria"    "Cyanobacteriia"     
ASV2:Pseudoalteromonas    "Bacteria" "Proteobacteria"   "Gammaproteobacteria"
ASV3:Clade Ia             "Bacteria" "Proteobacteria"   "Alphaproteobacteria"
ASV4:Marine Group II      "Archaea"  "Thermoplasmatota" "Thermoplasmata"     
ASV5:Clade Ia             "Bacteria" "Proteobacteria"   "Alphaproteobacteria"
ASV6:Clade II             "Bacteria" "Proteobacteria"   "Alphaproteobacteria"
                          Order              Family                  
ASV1:Synechococcus CC9902 "Synechococcales"  "Cyanobiaceae"          
ASV2:Pseudoalteromonas    "Enterobacterales" "Pseudoalteromonadaceae"
ASV3:Clade Ia             "SAR11 clade"      "Clade I"               
ASV4:Marine Group II      "Marine Group II"  NA                      
ASV5:Clade Ia             "SAR11 clade"      "Clade I"               
ASV6:Clade II             "SAR11 clade"      "Clade II"              
                          Genus                  Species
ASV1:Synechococcus CC9902 "Synechococcus CC9902" NA     
ASV2:Pseudoalteromonas    "Pseudoalteromonas"    NA     
ASV3:Clade Ia             "Clade Ia"             NA     
ASV4:Marine Group II      NA                     NA     
ASV5:Clade Ia             "Clade Ia"             NA     
ASV6:Clade II             NA                     NA     </code></pre>
</div>
</div>
</section>
<section id="identify-core-microbiota" class="level3" data-number="7.0.5">
<h3 data-number="7.0.5" class="anchored" data-anchor-id="identify-core-microbiota"><span class="header-section-number">7.0.5</span> Identify Core microbiota</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="co">#North</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>(core_taxa_north <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">core_members</span>(sub_North,</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">detection =</span> <span class="fl">0.0001</span>,</span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">prevalence =</span> <span class="dv">50</span><span class="sc">/</span><span class="dv">100</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ASV1:Synechococcus CC9902"            
 [2] "ASV2:Pseudoalteromonas"               
 [3] "ASV3:Clade Ia"                        
 [4] "ASV4:Marine Group II"                 
 [5] "ASV5:Clade Ia"                        
 [6] "ASV6:Clade II"                        
 [7] "ASV9:Clade Ia"                        
 [8] "ASV10:Clade Ia"                       
 [9] "ASV11:AEGEAN-169 marine group"        
[10] "ASV12:Prochlorococcus MIT9313.marinus"
[11] "ASV16:AEGEAN-169 marine group"        
[12] "ASV18:Clade Ib"                       
[13] "ASV22:Clade Ia"                       
[14] "ASV23:Clade Ia"                       
[15] "ASV26:Chloroplast"                    
[16] "ASV30:Marine Group II"                
[17] "ASV32:Dadabacteriales"                
[18] "ASV33:SAR324 clade(Marine group B)"   
[19] "ASV35:Clade IV"                       
[20] "ASV37:Marine Group III"               
[21] "ASV49:SAR202 clade"                   
[22] "ASV53:AEGEAN-169 marine group"        </code></pre>
</div>
</div>
</section>
<section id="get-core-microbiota-phyloseq-object" class="level3" data-number="7.0.6">
<h3 data-number="7.0.6" class="anchored" data-anchor-id="get-core-microbiota-phyloseq-object"><span class="header-section-number">7.0.6</span> Get core microbiota phyloseq object</h3>
<p>Get the phyloseq object with also sequences, phylo tree etc.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>(phyloseq_core_north <span class="ot">&lt;-</span> microbiome<span class="sc">::</span><span class="fu">core</span>(sub_North,</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">detection =</span> <span class="fl">0.0001</span>,</span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">prevalence =</span> .<span class="dv">5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 22 taxa and 9 samples ]
sample_data() Sample Data:       [ 9 samples by 26 sample variables ]
tax_table()   Taxonomy Table:    [ 22 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 22 tips and 21 internal nodes ]
refseq()      DNAStringSet:      [ 22 reference sequences ]</code></pre>
</div>
</div>
<p>See full taxanomy of core members</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>(tax_mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(phyloseq<span class="sc">::</span><span class="fu">tax_table</span>(phyloseq_core_north)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                        Domain                       Phylum
ASV1:Synechococcus CC9902             Bacteria                Cyanobacteria
ASV2:Pseudoalteromonas                Bacteria               Proteobacteria
ASV3:Clade Ia                         Bacteria               Proteobacteria
ASV4:Marine Group II                   Archaea             Thermoplasmatota
ASV5:Clade Ia                         Bacteria               Proteobacteria
ASV6:Clade II                         Bacteria               Proteobacteria
ASV9:Clade Ia                         Bacteria               Proteobacteria
ASV10:Clade Ia                        Bacteria               Proteobacteria
ASV11:AEGEAN-169 marine group         Bacteria               Proteobacteria
ASV12:Prochlorococcus MIT9313.marinus Bacteria                Cyanobacteria
ASV16:AEGEAN-169 marine group         Bacteria               Proteobacteria
ASV18:Clade Ib                        Bacteria               Proteobacteria
ASV22:Clade Ia                        Bacteria               Proteobacteria
ASV23:Clade Ia                        Bacteria               Proteobacteria
ASV26:Chloroplast                     Bacteria                Cyanobacteria
ASV30:Marine Group II                  Archaea             Thermoplasmatota
ASV32:Dadabacteriales                 Bacteria                 Dadabacteria
ASV33:SAR324 clade(Marine group B)    Bacteria SAR324 clade(Marine group B)
ASV35:Clade IV                        Bacteria               Proteobacteria
ASV37:Marine Group III                 Archaea             Thermoplasmatota
ASV49:SAR202 clade                    Bacteria                  Chloroflexi
ASV53:AEGEAN-169 marine group         Bacteria               Proteobacteria
                                                    Class            Order
ASV1:Synechococcus CC9902                  Cyanobacteriia  Synechococcales
ASV2:Pseudoalteromonas                Gammaproteobacteria Enterobacterales
ASV3:Clade Ia                         Alphaproteobacteria      SAR11 clade
ASV4:Marine Group II                       Thermoplasmata  Marine Group II
ASV5:Clade Ia                         Alphaproteobacteria      SAR11 clade
ASV6:Clade II                         Alphaproteobacteria      SAR11 clade
ASV9:Clade Ia                         Alphaproteobacteria      SAR11 clade
ASV10:Clade Ia                        Alphaproteobacteria      SAR11 clade
ASV11:AEGEAN-169 marine group         Alphaproteobacteria Rhodospirillales
ASV12:Prochlorococcus MIT9313.marinus      Cyanobacteriia  Synechococcales
ASV16:AEGEAN-169 marine group         Alphaproteobacteria Rhodospirillales
ASV18:Clade Ib                        Alphaproteobacteria      SAR11 clade
ASV22:Clade Ia                        Alphaproteobacteria      SAR11 clade
ASV23:Clade Ia                        Alphaproteobacteria      SAR11 clade
ASV26:Chloroplast                          Cyanobacteriia      Chloroplast
ASV30:Marine Group II                      Thermoplasmata  Marine Group II
ASV32:Dadabacteriales                       Dadabacteriia  Dadabacteriales
ASV33:SAR324 clade(Marine group B)                   &lt;NA&gt;             &lt;NA&gt;
ASV35:Clade IV                        Alphaproteobacteria      SAR11 clade
ASV37:Marine Group III                     Thermoplasmata Marine Group III
ASV49:SAR202 clade                        Dehalococcoidia     SAR202 clade
ASV53:AEGEAN-169 marine group         Alphaproteobacteria Rhodospirillales
                                                       Family
ASV1:Synechococcus CC9902                        Cyanobiaceae
ASV2:Pseudoalteromonas                 Pseudoalteromonadaceae
ASV3:Clade Ia                                         Clade I
ASV4:Marine Group II                                     &lt;NA&gt;
ASV5:Clade Ia                                         Clade I
ASV6:Clade II                                        Clade II
ASV9:Clade Ia                                         Clade I
ASV10:Clade Ia                                        Clade I
ASV11:AEGEAN-169 marine group         AEGEAN-169 marine group
ASV12:Prochlorococcus MIT9313.marinus            Cyanobiaceae
ASV16:AEGEAN-169 marine group         AEGEAN-169 marine group
ASV18:Clade Ib                                        Clade I
ASV22:Clade Ia                                        Clade I
ASV23:Clade Ia                                        Clade I
ASV26:Chloroplast                                        &lt;NA&gt;
ASV30:Marine Group II                                    &lt;NA&gt;
ASV32:Dadabacteriales                                    &lt;NA&gt;
ASV33:SAR324 clade(Marine group B)                       &lt;NA&gt;
ASV35:Clade IV                                       Clade IV
ASV37:Marine Group III                                   &lt;NA&gt;
ASV49:SAR202 clade                                       &lt;NA&gt;
ASV53:AEGEAN-169 marine group         AEGEAN-169 marine group
                                                        Genus Species
ASV1:Synechococcus CC9902                Synechococcus CC9902    &lt;NA&gt;
ASV2:Pseudoalteromonas                      Pseudoalteromonas    &lt;NA&gt;
ASV3:Clade Ia                                        Clade Ia    &lt;NA&gt;
ASV4:Marine Group II                                     &lt;NA&gt;    &lt;NA&gt;
ASV5:Clade Ia                                        Clade Ia    &lt;NA&gt;
ASV6:Clade II                                            &lt;NA&gt;    &lt;NA&gt;
ASV9:Clade Ia                                        Clade Ia    &lt;NA&gt;
ASV10:Clade Ia                                       Clade Ia    &lt;NA&gt;
ASV11:AEGEAN-169 marine group                            &lt;NA&gt;    &lt;NA&gt;
ASV12:Prochlorococcus MIT9313.marinus Prochlorococcus MIT9313 marinus
ASV16:AEGEAN-169 marine group                            &lt;NA&gt;    &lt;NA&gt;
ASV18:Clade Ib                                       Clade Ib    &lt;NA&gt;
ASV22:Clade Ia                                       Clade Ia    &lt;NA&gt;
ASV23:Clade Ia                                       Clade Ia    &lt;NA&gt;
ASV26:Chloroplast                                        &lt;NA&gt;    &lt;NA&gt;
ASV30:Marine Group II                                    &lt;NA&gt;    &lt;NA&gt;
ASV32:Dadabacteriales                                    &lt;NA&gt;    &lt;NA&gt;
ASV33:SAR324 clade(Marine group B)                       &lt;NA&gt;    &lt;NA&gt;
ASV35:Clade IV                                           &lt;NA&gt;    &lt;NA&gt;
ASV37:Marine Group III                                   &lt;NA&gt;    &lt;NA&gt;
ASV49:SAR202 clade                                       &lt;NA&gt;    &lt;NA&gt;
ASV53:AEGEAN-169 marine group                            &lt;NA&gt;    &lt;NA&gt;</code></pre>
</div>
</div>
</section>
<section id="visualise-core-microbiome-with-microbiomeplot_core" class="level3" data-number="7.0.7">
<h3 data-number="7.0.7" class="anchored" data-anchor-id="visualise-core-microbiome-with-microbiomeplot_core"><span class="header-section-number">7.0.7</span> Visualise core microbiome with <code>microbiome::plot_core()</code></h3>
<p>Visualise the core microbiome of North samples</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>microbiome<span class="sc">::</span><span class="fu">plot_core</span>(phyloseq_core_north,</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">plot.type =</span> <span class="st">"heatmap"</span>,</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">colours =</span> <span class="fu">rev</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">"RdBu"</span>)),</span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">prevalences =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">0</span>, <span class="at">to =</span> <span class="dv">1</span>, <span class="at">by =</span> .<span class="dv">1</span>),</span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">detections =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="fl">0.1</span>, <span class="at">to =</span> <span class="dv">5</span>, <span class="at">by =</span> <span class="fl">0.2</span>)) <span class="sc">+</span></span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">guide =</span> <span class="fu">guide_axis</span>(<span class="at">n.dodge =</span> <span class="dv">2</span>))<span class="sc">+</span></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Detection Threshold (Relative Abundance (%))"</span>) <span class="sc">+</span></span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"ASVs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in microbiome::plot_core(phyloseq_core_north, plot.type = "heatmap", : The plot_core function is typically used with compositional 
                data. The data is not compositional. Make sure that you
                intend to operate on non-compositional data.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="alpha_diversity_files/figure-html/unnamed-chunk-75-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>Do the same for the South samples .. please!</p>
<p>What are your conclusions about the comparison between North &amp; South core micobiota at the ASV level?</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/ANF-MetaBioDiv\.github\.io\/course-material\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center"><a href="http://abims.sb-roscoff.fr/"><img src="../img/logo_abims.png" alt="ABiMS" height="35"></a> <a href="https://www.umr-beep.fr/"><img src="../img/logo_beep.png" alt="beep" height="40"></a> <a href="https://www.france-bioinformatique.fr/"><img src="../img/logo_ifb.png" alt="IFB" height="40"></a> <a href="https://www.cnrs.fr//"><img src="../img/logo_cnrs.png" alt="CNRS" height="40"></a> <a href="https://www.ird.fr/"><img src="../img/logo_ird.png" alt="IRD" height="40"></a> <a href="https://www.mio.osupytheas.fr/"><img src="../img/logo_mio.png" alt="MIO" height="40"></a> <a href="https://umr-marbec.fr/"><img src="../img/logo_marbec.jpeg" alt="marbec" height="40"></a></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>