<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.339">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2022-08-09">

<title>ANF MetaBioDiv - PARTIII_BETA_Diversity_ANF</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ANF MetaBioDiv</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">    
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/rforbeginer.html">
 <span class="dropdown-text">R for beginners</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/introduction.pdf">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/dada2.html" target="_blank">
 <span class="dropdown-text">Analysing metabarcoding data with dada2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/phyloseq.html" target="_blank">
 <span class="dropdown-text">Introduction to phyloseq</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/preprocessing_dada2.html">
 <span class="dropdown-text">Preprocessing: dada2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals/preprocessing_phyloseq.html">
 <span class="dropdown-text">Preprocessing: phyloseq</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/alpha_diversity.pdf" target="_blank">
 <span class="dropdown-text">Alpha diversity</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/stat_mean_comparison.pdf" target="_blank">
 <span class="dropdown-text">Introduction to univariate statistics</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/alpha_diversity.html">
 <span class="dropdown-text">Alpha diversity</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-4" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Day 4</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-4">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/beta_diversity.pdf" target="_blank">
 <span class="dropdown-text">Beta diversity</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/beta_diversity.html">
 <span class="dropdown-text">Beta diversity</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ANF-MetaBioDiv/course-material" target="_blank"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#preparing-the-r-session" id="toc-preparing-the-r-session" class="nav-link active" data-scroll-target="#preparing-the-r-session"><span class="header-section-number">1</span> Preparing the R session</a>
  <ul class="collapse">
  <li><a href="#load-some-libraries" id="toc-load-some-libraries" class="nav-link" data-scroll-target="#load-some-libraries"><span class="header-section-number">1.1</span> Load some libraries</a></li>
  <li><a href="#load-the-data-in-r" id="toc-load-the-data-in-r" class="nav-link" data-scroll-target="#load-the-data-in-r"><span class="header-section-number">1.2</span> Load the data in R</a></li>
  </ul></li>
  <li><a href="#preparation-of-the-data" id="toc-preparation-of-the-data" class="nav-link" data-scroll-target="#preparation-of-the-data"><span class="header-section-number">2</span> Preparation of the data</a>
  <ul class="collapse">
  <li><a href="#normalisation" id="toc-normalisation" class="nav-link" data-scroll-target="#normalisation"><span class="header-section-number">2.1</span> Normalisation</a>
  <ul class="collapse">
  <li><a href="#rarefaction" id="toc-rarefaction" class="nav-link" data-scroll-target="#rarefaction"><span class="header-section-number">2.1.1</span> Rarefaction</a></li>
  <li><a href="#centered-log-ratio-clr-transformation" id="toc-centered-log-ratio-clr-transformation" class="nav-link" data-scroll-target="#centered-log-ratio-clr-transformation"><span class="header-section-number">2.1.2</span> Centered log-ratio (CLR) transformation</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#visualisation-of-the-meta-community-composition" id="toc-visualisation-of-the-meta-community-composition" class="nav-link" data-scroll-target="#visualisation-of-the-meta-community-composition"><span class="header-section-number">3</span> Visualisation of the meta-community composition</a>
  <ul class="collapse">
  <li><a href="#treemaps" id="toc-treemaps" class="nav-link" data-scroll-target="#treemaps"><span class="header-section-number">3.1</span> Treemaps</a>
  <ul class="collapse">
  <li><a href="#what-for" id="toc-what-for" class="nav-link" data-scroll-target="#what-for"><span class="header-section-number">3.1.1</span> What for?</a></li>
  <li><a href="#using-the-package-treemap" id="toc-using-the-package-treemap" class="nav-link" data-scroll-target="#using-the-package-treemap"><span class="header-section-number">3.1.2</span> Using the package <code>treemap</code></a></li>
  <li><a href="#using-the-package-treemapify" id="toc-using-the-package-treemapify" class="nav-link" data-scroll-target="#using-the-package-treemapify"><span class="header-section-number">3.1.3</span> Using the package <code>treemapify</code></a></li>
  <li><a href="#observations" id="toc-observations" class="nav-link" data-scroll-target="#observations"><span class="header-section-number">3.1.4</span> Observations</a></li>
  </ul></li>
  <li><a href="#stacked-barplots" id="toc-stacked-barplots" class="nav-link" data-scroll-target="#stacked-barplots"><span class="header-section-number">3.2</span> Stacked barplots</a></li>
  </ul></li>
  <li><a href="#distance-or-dissimilarity-matrices" id="toc-distance-or-dissimilarity-matrices" class="nav-link" data-scroll-target="#distance-or-dissimilarity-matrices"><span class="header-section-number">4</span> Distance or dissimilarity matrices</a>
  <ul class="collapse">
  <li><a href="#calculation" id="toc-calculation" class="nav-link" data-scroll-target="#calculation"><span class="header-section-number">4.1</span> Calculation</a>
  <ul class="collapse">
  <li><a href="#compositional-taxonomic" id="toc-compositional-taxonomic" class="nav-link" data-scroll-target="#compositional-taxonomic"><span class="header-section-number">4.1.1</span> Compositional taxonomic</a></li>
  <li><a href="#compositional-phylogenetic-unweighted-unifrac" id="toc-compositional-phylogenetic-unweighted-unifrac" class="nav-link" data-scroll-target="#compositional-phylogenetic-unweighted-unifrac"><span class="header-section-number">4.1.2</span> Compositional phylogenetic (Unweighted unifrac)</a></li>
  <li><a href="#structural-taxonomic-bray-curtis" id="toc-structural-taxonomic-bray-curtis" class="nav-link" data-scroll-target="#structural-taxonomic-bray-curtis"><span class="header-section-number">4.1.3</span> Structural taxonomic (Bray-Curtis)</a></li>
  <li><a href="#structural-phylogenetic-weighted-unifrac" id="toc-structural-phylogenetic-weighted-unifrac" class="nav-link" data-scroll-target="#structural-phylogenetic-weighted-unifrac"><span class="header-section-number">4.1.4</span> Structural phylogenetic (Weighted UniFrac)</a></li>
  </ul></li>
  <li><a href="#visualisation" id="toc-visualisation" class="nav-link" data-scroll-target="#visualisation"><span class="header-section-number">4.2</span> Visualisation</a></li>
  </ul></li>
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering"><span class="header-section-number">5</span> Hierarchical clustering</a>
  <ul class="collapse">
  <li><a href="#hierarchical-ascendant-classification-hac" id="toc-hierarchical-ascendant-classification-hac" class="nav-link" data-scroll-target="#hierarchical-ascendant-classification-hac"><span class="header-section-number">5.1</span> Hierarchical ascendant classification (HAC)</a></li>
  <li><a href="#cophenetic-correlation" id="toc-cophenetic-correlation" class="nav-link" data-scroll-target="#cophenetic-correlation"><span class="header-section-number">5.2</span> Cophenetic Correlation</a></li>
  <li><a href="#looking-for-interpretable-clusters" id="toc-looking-for-interpretable-clusters" class="nav-link" data-scroll-target="#looking-for-interpretable-clusters"><span class="header-section-number">5.3</span> Looking for Interpretable Clusters</a></li>
  <li><a href="#combining-clustering-and-z-score-heatmap" id="toc-combining-clustering-and-z-score-heatmap" class="nav-link" data-scroll-target="#combining-clustering-and-z-score-heatmap"><span class="header-section-number">5.4</span> Combining clustering and Z-score Heatmap</a>
  <ul class="collapse">
  <li><a href="#select-the-top-30-asv" id="toc-select-the-top-30-asv" class="nav-link" data-scroll-target="#select-the-top-30-asv"><span class="header-section-number">5.4.1</span> Select the top 30 ASV</a></li>
  <li><a href="#get-the-otu_table-z-score-transformation" id="toc-get-the-otu_table-z-score-transformation" class="nav-link" data-scroll-target="#get-the-otu_table-z-score-transformation"><span class="header-section-number">5.4.2</span> Get the otu_table &amp; Z-score transformation</a></li>
  <li><a href="#heat-map-z-score" id="toc-heat-map-z-score" class="nav-link" data-scroll-target="#heat-map-z-score"><span class="header-section-number">5.4.3</span> Heat map Z-score</a></li>
  <li><a href="#add-the-taxonomy-for-asv-names" id="toc-add-the-taxonomy-for-asv-names" class="nav-link" data-scroll-target="#add-the-taxonomy-for-asv-names"><span class="header-section-number">5.4.4</span> Add the Taxonomy for ASV names</a></li>
  <li><a href="#apply-to-the-heatmap" id="toc-apply-to-the-heatmap" class="nav-link" data-scroll-target="#apply-to-the-heatmap"><span class="header-section-number">5.4.5</span> Apply to the Heatmap</a></li>
  <li><a href="#add-a-boxplot-of-asv-abundance-distribution-within-sample" id="toc-add-a-boxplot-of-asv-abundance-distribution-within-sample" class="nav-link" data-scroll-target="#add-a-boxplot-of-asv-abundance-distribution-within-sample"><span class="header-section-number">5.4.6</span> Add a boxplot of ASV Abundance distribution within sample</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#indirect-gradient-analysis" id="toc-indirect-gradient-analysis" class="nav-link" data-scroll-target="#indirect-gradient-analysis"><span class="header-section-number">6</span> Indirect gradient analysis</a>
  <ul class="collapse">
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca"><span class="header-section-number">6.1</span> Principal component analysis (PCA)</a>
  <ul class="collapse">
  <li><a href="#number-of-pcs-to-retain" id="toc-number-of-pcs-to-retain" class="nav-link" data-scroll-target="#number-of-pcs-to-retain"><span class="header-section-number">6.1.1</span> Number of PCs to retain</a></li>
  <li><a href="#plotting-the-ordination" id="toc-plotting-the-ordination" class="nav-link" data-scroll-target="#plotting-the-ordination"><span class="header-section-number">6.1.2</span> Plotting the ordination</a></li>
  <li><a href="#determine-the-variables-that-drive-variation-among-each-pc" id="toc-determine-the-variables-that-drive-variation-among-each-pc" class="nav-link" data-scroll-target="#determine-the-variables-that-drive-variation-among-each-pc"><span class="header-section-number">6.1.3</span> Determine the variables that drive variation among each PC</a></li>
  <li><a href="#correlate-the-principal-components-back-to-environmental-data" id="toc-correlate-the-principal-components-back-to-environmental-data" class="nav-link" data-scroll-target="#correlate-the-principal-components-back-to-environmental-data"><span class="header-section-number">6.1.4</span> Correlate the principal components back to environmental data</a></li>
  </ul></li>
  <li><a href="#principal-component-analysis-pcoa" id="toc-principal-component-analysis-pcoa" class="nav-link" data-scroll-target="#principal-component-analysis-pcoa"><span class="header-section-number">6.2</span> Principal component analysis (PCoA)</a></li>
  <li><a href="#non-metric-multidimensional-scaling-nmds" id="toc-non-metric-multidimensional-scaling-nmds" class="nav-link" data-scroll-target="#non-metric-multidimensional-scaling-nmds"><span class="header-section-number">6.3</span> Non Metric Multidimensional Scaling (NMDS)</a></li>
  </ul></li>
  <li><a href="#hypothese-testing-analyses" id="toc-hypothese-testing-analyses" class="nav-link" data-scroll-target="#hypothese-testing-analyses"><span class="header-section-number">7</span> Hypothese testing analyses</a>
  <ul class="collapse">
  <li><a href="#permutational-multiple-analysis-of-variance-permanova" id="toc-permutational-multiple-analysis-of-variance-permanova" class="nav-link" data-scroll-target="#permutational-multiple-analysis-of-variance-permanova"><span class="header-section-number">7.1</span> PERmutational Multiple ANalysis Of VAriance (PERMANOVA)</a></li>
  <li><a href="#analysis-of-similarity-anosim" id="toc-analysis-of-similarity-anosim" class="nav-link" data-scroll-target="#analysis-of-similarity-anosim"><span class="header-section-number">7.2</span> ANalysis Of SIMilarity (ANOSIM)</a></li>
  </ul></li>
  <li><a href="#direct-gradient-analysis" id="toc-direct-gradient-analysis" class="nav-link" data-scroll-target="#direct-gradient-analysis"><span class="header-section-number">8</span> Direct gradient analysis</a>
  <ul class="collapse">
  <li><a href="#redundant-analysis-rda" id="toc-redundant-analysis-rda" class="nav-link" data-scroll-target="#redundant-analysis-rda"><span class="header-section-number">8.1</span> Redundant analysis (RDA)</a>
  <ul class="collapse">
  <li><a href="#running-the-rda" id="toc-running-the-rda" class="nav-link" data-scroll-target="#running-the-rda"><span class="header-section-number">8.1.1</span> Running the RDA</a></li>
  <li><a href="#significance-testing" id="toc-significance-testing" class="nav-link" data-scroll-target="#significance-testing"><span class="header-section-number">8.1.2</span> Significance testing</a></li>
  <li><a href="#selecting-variables" id="toc-selecting-variables" class="nav-link" data-scroll-target="#selecting-variables"><span class="header-section-number">8.1.3</span> Selecting variables</a></li>
  <li><a href="#rda-plot" id="toc-rda-plot" class="nav-link" data-scroll-target="#rda-plot"><span class="header-section-number">8.1.4</span> RDA plot</a></li>
  </ul></li>
  <li><a href="#multiple-regression-on-dissimilarity-matrices-mrm" id="toc-multiple-regression-on-dissimilarity-matrices-mrm" class="nav-link" data-scroll-target="#multiple-regression-on-dissimilarity-matrices-mrm"><span class="header-section-number">8.2</span> Multiple Regression on dissimilarity Matrices (MRM)</a></li>
  </ul></li>
  <li><a href="#diffential-abundance-analysis-daa" id="toc-diffential-abundance-analysis-daa" class="nav-link" data-scroll-target="#diffential-abundance-analysis-daa"><span class="header-section-number">9</span> Diffential abundance analysis (DAA)</a>
  <ul class="collapse">
  <li><a href="#linear-discriminant-analysis-effect-size-lefse" id="toc-linear-discriminant-analysis-effect-size-lefse" class="nav-link" data-scroll-target="#linear-discriminant-analysis-effect-size-lefse"><span class="header-section-number">9.1</span> Linear discriminant analysis Effect Size (LEFse)</a></li>
  <li><a href="#differential-analysis-of-compositions-of-microbiomes-with-bias-correction-ancom-bc" id="toc-differential-analysis-of-compositions-of-microbiomes-with-bias-correction-ancom-bc" class="nav-link" data-scroll-target="#differential-analysis-of-compositions-of-microbiomes-with-bias-correction-ancom-bc"><span class="header-section-number">9.2</span> Differential analysis of compositions of microbiomes with bias correction (ANCOM-BC)</a></li>
  <li><a href="#anova-like-differential-expression-aldex2" id="toc-anova-like-differential-expression-aldex2" class="nav-link" data-scroll-target="#anova-like-differential-expression-aldex2"><span class="header-section-number">9.3</span> ANOVA-like differential expression (ALDEx2)</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/ANF-MetaBioDiv/course-material/edit/main/practicals/beta_diversity.Rmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/ANF-MetaBioDiv/course-material/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">PARTIII_BETA_Diversity_ANF</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jean-Christophe Auguet </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 9, 2022</p>
    </div>
  </div>
  
    
  </div>
  


</header>

<section id="preparing-the-r-session" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Preparing the R session</h1>
<section id="load-some-libraries" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="load-some-libraries"><span class="header-section-number">1.1</span> Load some libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">load_all</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>output_beta <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>, <span class="st">"beta_diversity"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(output_beta)) <span class="fu">dir.create</span>(output_beta, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-the-data-in-r" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="load-the-data-in-r"><span class="header-section-number">1.2</span> Load the data in R</h2>
<p>Load the data and inspect the phyloseq object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>physeq <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"asv_table"</span>,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"phyloseq_object_alpha_beta_div.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="preparation-of-the-data" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Preparation of the data</h1>
<section id="normalisation" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="normalisation"><span class="header-section-number">2.1</span> Normalisation</h2>
<p>Here we will consider two approaches for library size normalization. The first will involve simply subsampling the data without replacement; however, this approach comes with limitations that are well described <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531" target="_blank">here</a>.</p>
<p>The second will employ a compositional data analysis approach and involves working with log-ratios.</p>
<section id="rarefaction" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="rarefaction"><span class="header-section-number">2.1.1</span> Rarefaction</h3>
<p>We will subsample reads from each sample without replacement to a constant depth. Let see first how many reads we have per sample:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowSums</span>(physeq<span class="sc">@</span>otu_table<span class="sc">@</span>.Data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>S11B  S1B  S2B  S2S  S3B  S3S  S4B  S4S  S5B  S5S  S6B  S6S  S7B  S7S  S8B  S8S 
 975  837  893  983  878  889  917 1077 1018 1006 1076  937  878  936  846  958 
 S9B  S9S 
 888  991 </code></pre>
</div>
</div>
<p>We will plot these results and look at the rank abudance of the reads</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>readsumsdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">nreads =</span> <span class="fu">sort</span>(<span class="fu">taxa_sums</span>(physeq), <span class="at">decreasing =</span> <span class="cn">TRUE</span>),</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">sorted =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ntaxa</span>(physeq),</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">type =</span> <span class="st">"OTUs"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">nreads =</span> <span class="fu">sort</span>(<span class="fu">sample_sums</span>(physeq), <span class="at">decreasing =</span> <span class="cn">TRUE</span>), </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">sorted =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nsamples</span>(physeq),</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">type =</span> <span class="st">"Samples"</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>readsumsdf <span class="ot">&lt;-</span> <span class="fu">rbind</span>(readsumsdf, tmp)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(readsumsdf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     nreads sorted type
ASV1   1558      1 OTUs
ASV2    973      2 OTUs
ASV3    899      3 OTUs
ASV4    833      4 OTUs
ASV5    767      5 OTUs
ASV6    654      6 OTUs</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(readsumsdf, <span class="fu">aes</span>(<span class="at">x =</span> sorted, <span class="at">y =</span> nreads)) <span class="sc">+</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Total number of reads"</span>) <span class="sc">+</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>type, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We will now transform to equal sample sum (Only if the number of reads is not the same between samples) in order to be sure that the sampling effort is the same between samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set the seed for random sampling</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># it allows reproductibility</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">10000</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># minimum reads in a sample</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(<span class="fu">rowSums</span>(physeq<span class="sc">@</span>otu_table<span class="sc">@</span>.Data))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 837</code></pre>
</div>
</div>
<p>The minimun number of reads in a sample is 837. Lets do the randomization at 800 reads per sample in order to apply the process also in the sample having this minimum of reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>physeq_rar <span class="ot">&lt;-</span> <span class="fu">rarefy_even_depth</span>(physeq, <span class="at">sample.size =</span> <span class="dv">800</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowSums</span>(physeq_rar<span class="sc">@</span>otu_table<span class="sc">@</span>.Data) <span class="co">#how many reads per sample</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>S11B  S1B  S2B  S2S  S3B  S3S  S4B  S4S  S5B  S5S  S6B  S6S  S7B  S7S  S8B  S8S 
 800  800  800  800  800  800  800  800  800  800  800  800  800  800  800  800 
 S9B  S9S 
 800  800 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>physeq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 213 taxa and 18 samples ]
sample_data() Sample Data:       [ 18 samples by 21 sample variables ]
tax_table()   Taxonomy Table:    [ 213 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 213 tips and 212 internal nodes ]
refseq()      DNAStringSet:      [ 213 reference sequences ]</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>physeq_rar</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 208 taxa and 18 samples ]
sample_data() Sample Data:       [ 18 samples by 21 sample variables ]
tax_table()   Taxonomy Table:    [ 208 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 208 tips and 207 internal nodes ]
refseq()      DNAStringSet:      [ 208 reference sequences ]</code></pre>
</div>
</div>
</section>
<section id="centered-log-ratio-clr-transformation" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="centered-log-ratio-clr-transformation"><span class="header-section-number">2.1.2</span> Centered log-ratio (CLR) transformation</h3>
<p>A detailed discussion of compositional data analysis (CoDA) is beyond the scope of this session but just know that microbiome data is compositional since reads total is constrained by the sequencing depth. Relative abundances (proportions) are obviously constraint by a sum equal to one. This total constraint induces strong dependencies among the observed abundances of the different taxa. In fact, nor the absolute abundance (read counts) nor the relative abundance (proportion) of one taxon alone are informative of the real abundance of the taxon in the environment. Instead, they provide information on the relative measure of abundance when compared to the abundance of other taxa in the same sample. For this reason, these data fail to meet many of the assumptions of our favorite statistical methods developed for unconstrained random variables. Working with ratios of compositional elements lets us transform these data to the Euclidian space and apply our favorite methods. There are different types of log-ratio “transformations” including the additive log-ratio, centered log-ratio, and isometric log-ratio transforms. Find more information <a href="https://academic.oup.com/bioinformatics/article/34/16/2870/4956011?login=false">here</a>, <a href="https://www.springer.com/us/book/9789811315336?gclid=Cj0KCQjw3uboBRDCARIsAO2XcYAphJ23am-AoIBh18HoW-WpAd8TwbQUEhc_DJV9gM-zWYtXe0-6l8saAkNHEALw_wcB">here</a> and <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full" target="_blank">here</a></p>
<p>Let’s perform the CLR transformation</p>
<p>tmp</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a> <span class="co"># we first replace the zeros using</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a> <span class="co"># the Count Zero Multiplicative approach</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> zCompositions<span class="sc">::</span><span class="fu">cmultRepl</span>(physeq<span class="sc">@</span>otu_table,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">method =</span> <span class="st">"CZM"</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>                                <span class="at">label =</span> <span class="dv">0</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># generate the centered log-ratio transformed. ASVs are in rows!!!!!</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>physeq_clr_asv <span class="ot">&lt;-</span> <span class="fu">apply</span>(tmp, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">log</span>(x) <span class="sc">-</span> <span class="fu">mean</span>(<span class="fu">log</span>(x)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a new phyloseq object with CLR tranformed counts</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>physeq_clr <span class="ot">&lt;-</span> physeq</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">otu_table</span>(physeq_clr) <span class="ot">&lt;-</span> <span class="fu">otu_table</span>(<span class="fu">t</span>(physeq_clr_asv),</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">taxa_are_rows =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(physeq_clr<span class="sc">@</span>otu_table<span class="sc">@</span>.Data[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         ASV1      ASV2     ASV3      ASV4     ASV5     ASV6      ASV7
S11B 3.893295  2.349996 3.573772  3.379616 3.597029 2.820000  3.174172
S1B  4.051237 -1.253233 2.982039 -1.253233 3.778370 3.717746 -1.253233
S2B  3.041890 -1.827389 2.836038  1.988740 3.018360 3.231934 -1.827389
S2S  3.705729  3.536908 3.401733  1.555907 3.666120 2.832200  2.654519
S3B  3.635430 -1.548237 3.023628 -1.548237 3.449713 3.848352 -1.548237
          ASV8      ASV9     ASV10
S11B  2.657481  2.844693 -1.981602
S1B  -1.253233  3.142381  3.906988
S2B  -1.827389 -1.827389  3.041890
S2S   3.347666  2.899641 -2.042622
S3B  -1.548237 -1.548237  3.600944</code></pre>
</div>
</div>
<p>We can see that the values are now no longer counts, but rather the dominance (or lack thereof) for each taxa relative to the geometric mean of all taxa on the logarithmic scale. This centered log-ratio (CLR) transformed table can be used directly in a PCA or RDA to generate a beta diveristy ordination using the Aitchison distance.</p>
<p><em>Tips: This chunk can be coded in one line using the <a href="https://microbiome.github.io/tutorials/" target="_blank">microbiome package</a>: physeq_clr &lt;- microbiome::transform(physeq, “clr”)</em></p>
</section>
</section>
</section>
<section id="visualisation-of-the-meta-community-composition" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Visualisation of the meta-community composition</h1>
<p>Often an early step in many microbiome projects is to visualize the relative abundance of organisms at specific taxonomic ranks. Treemaps and stacked barplots are two ways of doing this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>physeq_phylum <span class="ot">&lt;-</span> physeq_rar <span class="sc">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tax_glom</span>(<span class="at">taxrank =</span> <span class="st">"Family"</span>) <span class="sc">%&gt;%</span>                     <span class="co"># agglomerate at the Family level</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">transform_sample_counts</span>(<span class="cf">function</span>(x) {x<span class="sc">/</span><span class="fu">sum</span>(x)} ) <span class="sc">%&gt;%</span> <span class="co"># Transform to rel. abundance</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">psmelt</span>() <span class="sc">%&gt;%</span>                                         <span class="co"># Melt to long format</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(Abundance <span class="sc">&gt;</span> <span class="fl">0.02</span>) <span class="sc">%&gt;%</span>                         <span class="co"># Filter out low abundance taxa</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(Family)                                      <span class="co"># Sort data frame alphabetically by phylum</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(physeq_phylum)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   OTU Sample Abundance SampName   Geo Description groupe Pres PicoEuk Synec
1 ASV7    S4S 0.2483311      S4S North     North4S    NBS   78    2220  3130
2 ASV7    S4B 0.2129784      S4B North     North4B    NBF   78    2220  3130
3 ASV7    S5S 0.1833085      S5S North     North5S    NBS    0    1620 56555
4 ASV7    S2B 0.1776119      S2B North     North2B    NBF   59     890 25480
5 ASV7    S3S 0.1745283      S3S North     North3S    NBS    0     715 26725
6 ASV7    S5B 0.1741214      S5B North     North5B    NBF   42    1620 55780
  Prochloro NanoEuk Crypto SiOH4   NO2   NO3   NH4   PO4    NT    PT   Chla
1     29835    2120    235 2.457 0.099 1.087 0.349 0.137 8.689 3.129 0.0000
2     29835    2120    235 2.457 0.099 1.087 0.349 0.137 8.689 3.129 0.0000
3     22835    2560    945 2.669 0.136 0.785 0.267 0.114 9.146 3.062 0.0000
4     16595     670    395 2.592 0.105 1.125 0.328 0.067 9.378 3.391 0.0000
5     16860     890    200 1.656 0.098 0.794 0.367 0.095 7.847 2.520 0.0000
6     23795    2555   1355 2.028 0.103 1.135 0.216 0.128 8.623 3.137 0.0102
        T       S Sigma_t  Kingdom         Phylum               Class
1 18.8515 37.4542 26.9415 Bacteria Proteobacteria Alphaproteobacteria
2 18.8515 37.4542 26.9415 Bacteria Proteobacteria Alphaproteobacteria
3 24.1789 38.3213 26.1065 Bacteria Proteobacteria Alphaproteobacteria
4 22.6824 37.6627 26.0521 Bacteria Proteobacteria Alphaproteobacteria
5 22.5610 37.5960 26.0332 Bacteria Proteobacteria Alphaproteobacteria
6 24.1905 38.3192 26.1037 Bacteria Proteobacteria Alphaproteobacteria
             Order                  Family
1 Rhodospirillales AEGEAN-169 marine group
2 Rhodospirillales AEGEAN-169 marine group
3 Rhodospirillales AEGEAN-169 marine group
4 Rhodospirillales AEGEAN-169 marine group
5 Rhodospirillales AEGEAN-169 marine group
6 Rhodospirillales AEGEAN-169 marine group</code></pre>
</div>
</div>
<section id="treemaps" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="treemaps"><span class="header-section-number">3.1</span> Treemaps</h2>
<section id="what-for" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="what-for"><span class="header-section-number">3.1.1</span> What for?</h3>
<p>Looking at the composition of the meta community with a treemap allow roughly to detect if some taxa should not be present (contaminants) and observe if the dominanting taxa correspond to the habitat studied.</p>
</section>
<section id="using-the-package-treemap" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="using-the-package-treemap"><span class="header-section-number">3.1.2</span> Using the package <code>treemap</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#pdf(file="treemap.pdf", wi = 7, he = 7)</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>treemap<span class="sc">::</span><span class="fu">treemap</span>(physeq_phylum, <span class="at">index=</span><span class="fu">c</span>(<span class="st">"Class"</span>, <span class="st">"Family"</span>), <span class="at">vSize=</span><span class="st">"Abundance"</span>, <span class="at">type=</span><span class="st">"index"</span>,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">fontsize.labels=</span><span class="fu">c</span>(<span class="dv">15</span>,<span class="dv">12</span>),                <span class="co"># size of labels. Give the size per level of aggregation: size for group, size for subgroup, sub-subgroups...</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">fontcolor.labels=</span><span class="fu">c</span>(<span class="st">"white"</span>,<span class="st">"black"</span>),    <span class="co"># Color of labels</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">fontface.labels=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>),                  <span class="co"># Font of labels: 1,2,3,4 for normal, bold, italic, bold-italic...</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">align.labels=</span><span class="fu">list</span>(</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>          <span class="fu">c</span>(<span class="st">"center"</span>, <span class="st">"center"</span>), </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>          <span class="fu">c</span>(<span class="st">"left"</span>, <span class="st">"bottom"</span>)),                 <span class="co"># Where to place labels in the rectangle?</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">overlap.labels=</span><span class="fl">0.5</span>,                      <span class="co"># number between 0 and 1 that determines the tolerance of the overlap between labels. 0 means that labels of lower levels are not printed if higher level labels overlap, 1  means that labels are always printed. In-between values, for instance the default value .5, means that lower level labels are printed if other labels do not overlap with more than .5  times their area size.</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">inflate.labels=</span>F, <span class="co"># If true, labels are bigger when rectangle is bigger.</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">border.col=</span><span class="fu">c</span>(<span class="st">"black"</span>,<span class="st">"white"</span>),          <span class="co">#Color of the boders separating the taxonomic levels</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">border.lwds=</span><span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">2</span>),</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>        <span class="co">#palette = "Set3",                        # Select your color palette from the RColorBrewer presets or make your own.</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">fontsize.title=</span><span class="dv">12</span></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="using-the-package-treemapify" class="level3" data-number="3.1.3">
<h3 data-number="3.1.3" class="anchored" data-anchor-id="using-the-package-treemapify"><span class="header-section-number">3.1.3</span> Using the package <code>treemapify</code></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(physeq,<span class="cf">function</span>(x) {x<span class="sc">/</span><span class="fu">sum</span>(x)} ) <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">psmelt</span>() <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Family, Class) <span class="sc">%&gt;%</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">abundance =</span> <span class="fu">sum</span>(Abundance)) <span class="sc">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(tmp,<span class="fu">aes</span>(<span class="at">area=</span>abundance,<span class="at">label=</span>Family,<span class="at">fill=</span>Class,<span class="at">subgroup=</span>Class))<span class="sc">+</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  treemapify<span class="sc">::</span><span class="fu">geom_treemap</span>()<span class="sc">+</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  treemapify<span class="sc">::</span><span class="fu">geom_treemap_subgroup_border</span>() <span class="sc">+</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  treemapify<span class="sc">::</span><span class="fu">geom_treemap_subgroup_text</span>(<span class="at">place =</span> <span class="st">"centre"</span>,</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">grow =</span> T,</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">alpha =</span> <span class="fl">0.5</span>,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">colour =</span> <span class="st">"black"</span>,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">fontface =</span> <span class="st">"italic"</span>,</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">min.size =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  treemapify<span class="sc">::</span><span class="fu">geom_treemap_text</span>(<span class="at">colour =</span> <span class="st">"white"</span>,</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>                                <span class="at">place =</span> <span class="st">"topleft"</span>,</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>                                <span class="at">reflow =</span> <span class="cn">TRUE</span>)<span class="sc">+</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(here<span class="sc">::</span><span class="fu">here</span>(output_beta,<span class="st">"treemap_treemapify.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="observations" class="level3" data-number="3.1.4">
<h3 data-number="3.1.4" class="anchored" data-anchor-id="observations"><span class="header-section-number">3.1.4</span> Observations</h3>
<p>Here we can observe that the meta-community is dominated by typical marine clades such as the AEGEAN marine group in Alphaproteobacteria or the SAR86 clade in Gammaproteobacteria. So everything is good so far.</p>
</section>
</section>
<section id="stacked-barplots" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="stacked-barplots"><span class="header-section-number">3.2</span> Stacked barplots</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(physeq_phylum, <span class="fu">aes</span>(<span class="at">x =</span> Sample, <span class="at">y =</span> Abundance, <span class="at">fill =</span> Family)) <span class="sc">+</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># facet_wrap(~Treatment, nrow=1, scales = "free_x") +</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Relative Abundance (Family &gt; 2%)"</span>) <span class="sc">+</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="sc">+</span> <span class="co">#remove the space below the 0 of the y axis in the graph</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Community composition"</span>) <span class="sc">+</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">size =</span> <span class="dv">10</span>,</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">vjust =</span> <span class="fl">0.8</span>),</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),  <span class="co">#remove major-grid labels</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>())  <span class="co">#remove minor-grid labels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(here<span class="sc">::</span><span class="fu">here</span>(output_beta, <span class="st">"asv_composition.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we can already see some differences in the composition at the Family level with an enrichment in Pseudoalteromonadaceae in some samples and Cyanobiaceae. Note that we are limited by our ability to discernate between than more than 9-12 colors in this kind of graphic.</p>
</section>
</section>
<section id="distance-or-dissimilarity-matrices" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Distance or dissimilarity matrices</h1>
<section id="calculation" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="calculation"><span class="header-section-number">4.1</span> Calculation</h2>
<p>Over the years, ecologists have invented numerous ways of quantifying dissimilarity between pairs of ecosystems.Four components of species community beta-diveristy can be assessed using different distances or dissimilarities. Compostional distances or dissimilarities do not consider the relative abundance of taxa, only their presence (detection) or absence, which can make it (overly) sensitive to rare taxa, sequencing artefacts, and abundance filtering choices. Conversely, structural distances or dissimilarities do put (perhaps too much) more importance on highly abundant taxa, when determining dissimilarities. Phylogentic distances or dissimilarities take into account the phylogenetic relatedness of the taxa / sequences in your samples when calculating dissimilarity while taxnomic distances or dissimilarities do not.</p>
<section id="compositional-taxonomic" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1" class="anchored" data-anchor-id="compositional-taxonomic"><span class="header-section-number">4.1.1</span> Compositional taxonomic</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>physeq_rar_jaccard <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(physeq_rar,</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">method =</span> <span class="st">"jaccard"</span>,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">binary =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># trick to avoid negative egein values in PCoA</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="co"># it recreates what ade4::dist.binary() does</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>physeq_rar_jaccard <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(physeq_rar_jaccard)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="compositional-phylogenetic-unweighted-unifrac" class="level3" data-number="4.1.2">
<h3 data-number="4.1.2" class="anchored" data-anchor-id="compositional-phylogenetic-unweighted-unifrac"><span class="header-section-number">4.1.2</span> Compositional phylogenetic (Unweighted unifrac)</h3>
<p>The GUniFrac package requires a rooted tree as input data. We can use the function midpoint() from the phangorn package to obtain the rooted tree.</p>
<p>To check if your tree is rooted, you may use this function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ape<span class="sc">::</span><span class="fu">is.rooted</span>(physeq_rar<span class="sc">@</span>phy_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>If not, you can use the function midpoint() from the package phangorn</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">phy_tree</span>(physeq_rar) <span class="ot">&lt;-</span> phangorn<span class="sc">::</span><span class="fu">midpoint</span>(physeq_rar<span class="sc">@</span>phy_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, Unifrac distances can be calculated</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>unifracs <span class="ot">&lt;-</span> GUniFrac<span class="sc">::</span><span class="fu">GUniFrac</span>(physeq_rar<span class="sc">@</span>otu_table<span class="sc">@</span>.Data, physeq_rar<span class="sc">@</span>phy_tree, <span class="at">alpha=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>))<span class="sc">$</span>unifracs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The unifracs object is a list containing 5 distance matrices correspoonding to the weighted UniFrac (d_1), the unweighted UniFrac (d_UW), Variance adjusted UniFrac (d_VAW), GUniFrac with alpha = 0, GUniFrac with alpha = 0.5</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>physeq_rar_du <span class="ot">&lt;-</span> unifracs[, , <span class="st">"d_UW"</span>]   <span class="co"># Unweighted UniFrac</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="structural-taxonomic-bray-curtis" class="level3" data-number="4.1.3">
<h3 data-number="4.1.3" class="anchored" data-anchor-id="structural-taxonomic-bray-curtis"><span class="header-section-number">4.1.3</span> Structural taxonomic (Bray-Curtis)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># physeq_rar_bray &lt;- vegan::vegdist(physeq_rar@otu_table@.Data, method = "bray")</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">transform_sample_counts</span>(physeq,<span class="cf">function</span>(x) {x<span class="sc">/</span><span class="fu">sum</span>(x)} )</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>physeq_rar_bray <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(tmp, <span class="at">method =</span> <span class="st">"bray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="structural-phylogenetic-weighted-unifrac" class="level3" data-number="4.1.4">
<h3 data-number="4.1.4" class="anchored" data-anchor-id="structural-phylogenetic-weighted-unifrac"><span class="header-section-number">4.1.4</span> Structural phylogenetic (Weighted UniFrac)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>physeq_rar_dw <span class="ot">&lt;-</span> unifracs[, , <span class="st">"d_1"</span>]   <span class="co"># Weighted UniFrac</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="visualisation" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="visualisation"><span class="header-section-number">4.2</span> Visualisation</h2>
<p>You can actually calculate all these distances directly in phyloseq using <em>dist.calc()</em>. There are currently 44 explicitly supported method options in the phyloseq package, see <a href="https://joey711.github.io/phyloseq/distance.html" target="_blank">here</a> for more informations. Here, we will loop through each distance method, save each plot to a list and then plot these results in a combined graphic using ggplot2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>dist_methods <span class="ot">&lt;-</span> <span class="fu">unlist</span>(distanceMethodList)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">position =</span> <span class="fu">seq_along</span>(dist_methods),</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>           dist_methods)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            position dist_methods
UniFrac1           1      unifrac
UniFrac2           2     wunifrac
DPCoA              3        dpcoa
JSD                4          jsd
vegdist1           5    manhattan
vegdist2           6    euclidean
vegdist3           7     canberra
vegdist4           8         bray
vegdist5           9   kulczynski
vegdist6          10      jaccard
vegdist7          11        gower
vegdist8          12     altGower
vegdist9          13     morisita
vegdist10         14         horn
vegdist11         15    mountford
vegdist12         16         raup
vegdist13         17     binomial
vegdist14         18         chao
vegdist15         19          cao
betadiver1        20            w
betadiver2        21           -1
betadiver3        22            c
betadiver4        23           wb
betadiver5        24            r
betadiver6        25            I
betadiver7        26            e
betadiver8        27            t
betadiver9        28           me
betadiver10       29            j
betadiver11       30          sor
betadiver12       31            m
betadiver13       32           -2
betadiver14       33           co
betadiver15       34           cc
betadiver16       35            g
betadiver17       36           -3
betadiver18       37            l
betadiver19       38           19
betadiver20       39           hk
betadiver21       40          rlb
betadiver22       41          sim
betadiver23       42           gl
betadiver24       43            z
dist1             44      maximum
dist2             45       binary
dist3             46    minkowski
designdist        47          ANY</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Select the distances of interest</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>dist_methods <span class="ot">&lt;-</span> dist_methods[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">10</span>, <span class="dv">8</span>)]</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>dist_methods</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  UniFrac1   UniFrac2   vegdist6   vegdist4 
 "unifrac" "wunifrac"  "jaccard"     "bray" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Loop through each distance method, save each plot to a list, called plist.</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">"list"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> dist_methods){</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate distance matrix</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  iDist <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(physeq_rar, <span class="at">method =</span> i)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Calculate PCoA ordination</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  iMDS <span class="ot">&lt;-</span> <span class="fu">ordinate</span>(physeq_rar, <span class="st">"MDS"</span>, <span class="at">distance =</span> iDist)</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Make plot. Don't carry over previous plot (if error, p will be blank)</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create plot, store as temp variable, p</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">plot_ordination</span>(physeq_rar, iMDS, <span class="at">color=</span> <span class="st">"Geo"</span>)</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add title to each plot</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"MDS using distance method "</span>, i, <span class="at">sep=</span><span class="st">""</span>))</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save the graphic to list</span></span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>  plist[[i]] <span class="ot">=</span> p </span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Combine results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">ldply</span>(plist, <span class="cf">function</span>(x) x<span class="sc">$</span>data)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      .id      Axis.1      Axis.2 SampName   Geo Description groupe Pres
1 unifrac  0.09023445  0.06150644     S11B South     South5B    SGF   35
2 unifrac -0.21048836 -0.19946687      S1B North     North1B    NBF   52
3 unifrac -0.21001002 -0.08655455      S2B North     North2B    NBF   59
4 unifrac  0.12583068  0.07022248      S2S North     North2S    NBS    0
5 unifrac -0.31465014 -0.06077941      S3B North     North3B    NBF   74
6 unifrac -0.16616937  0.01827175      S3S North     North3S    NBS    0
  PicoEuk Synec Prochloro NanoEuk Crypto SiOH4   NO2   NO3   NH4   PO4    NT
1    5370 46830       580    6010   1690 3.324 0.083 0.756 0.467 0.115 9.539
2     660 32195     10675     955    115 1.813 0.256 0.889 0.324 0.132 9.946
3     890 25480     16595     670    395 2.592 0.105 1.125 0.328 0.067 9.378
4     890 25480     16595     670    395 3.381 0.231 0.706 0.450 0.109 8.817
5     835 13340     25115    1115    165 1.438 0.057 1.159 0.369 0.174 8.989
6     715 26725     16860     890    200 1.656 0.098 0.794 0.367 0.095 7.847
     PT   Chla       T       S Sigma_t
1 4.138 0.0182 23.0308 38.9967 26.9631
2 3.565 0.0000 22.7338 37.6204 26.0046
3 3.391 0.0000 22.6824 37.6627 26.0521
4 3.345 0.0000 22.6854 37.6176 26.0137
5 2.568 0.0000 21.5296 37.5549 26.2987
6 2.520 0.0000 22.5610 37.5960 26.0332</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(df)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"distance"</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(Axis<span class="fl">.1</span>, Axis<span class="fl">.2</span>, <span class="at">color =</span> Geo)) <span class="sc">+</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">3</span>, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>distance, <span class="at">scales=</span><span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"PCoA (MDS) on various distance metrics"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can observe that there is a fairly good separation between North and South samples except for the Weighted UniFrac distance which tends to give too much weight to the most abundant ASVs that are also the most frequent.</p>
</section>
</section>
<section id="hierarchical-clustering" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Hierarchical clustering</h1>
<section id="hierarchical-ascendant-classification-hac" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="hierarchical-ascendant-classification-hac"><span class="header-section-number">5.1</span> Hierarchical ascendant classification (HAC)</h2>
<p>A first step in many microbiome projects is to examine how samples cluster on some measure of (dis)similarity. There are many ways to do perform such clustering, see <a href="https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/cluster-analysis" target="_blank">here</a>. Since microbiome data is compositional, we will perform here a hierarchical ascendant classification (HAC) of samples based on Aitchison distance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">#distance matrix calculation</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>physeq_clr_dist <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(physeq_clr, <span class="at">method =</span> <span class="st">"euclidean"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s see the different clustering obtained with the four aggregation criterion</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Simple aggregation criterion</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>spe_single <span class="ot">&lt;-</span> <span class="fu">hclust</span>(physeq_clr_dist, <span class="at">method =</span> <span class="st">"single"</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Complete aggregation criterion</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>spe_complete <span class="ot">&lt;-</span> <span class="fu">hclust</span>(physeq_clr_dist, <span class="at">method =</span> <span class="st">"complete"</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Unweighted pair group method with arithmetic mean</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>spe_upgma <span class="ot">&lt;-</span> <span class="fu">hclust</span>(physeq_clr_dist, <span class="at">method =</span> <span class="st">"average"</span>)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Ward criterion</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>spe_ward <span class="ot">&lt;-</span> <span class="fu">hclust</span>(physeq_clr_dist, <span class="at">method =</span> <span class="st">"ward.D"</span>)</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spe_single, <span class="at">main =</span> <span class="st">"single"</span>)</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spe_complete, <span class="at">main =</span> <span class="st">"complete"</span>)</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spe_upgma, <span class="at">main =</span> <span class="st">"UPGMA"</span>)</span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spe_ward, <span class="at">main =</span> <span class="st">"ward"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Remember that clustering is a heuristic procedure, not a statistical test. The choices of an association coefficient and a clustering method influence the result. This stresses the importance of choosing a method that is consistent with the aims of the analysis.</p>
</section>
<section id="cophenetic-correlation" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="cophenetic-correlation"><span class="header-section-number">5.2</span> Cophenetic Correlation</h2>
<p>A cophenetic matrix is a matrix representing the cophenetic distances among all pairs of objects. A Pearson’s r correlation, called the cophenetic correlation in this context, can be computed between the original dissimilarity matrix and the cophenetic matrix. The method with the highest cophenetic correlation may be seen as the one that produced the best clustering model for the distance matrix.</p>
<p>Let us compute the cophenetic matrix and correlation of four clustering results presented above, by means of the function <em>cophenetic()</em> of package stats.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Cophenetic corrélation</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>spe_single_coph <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>(spe_single)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(physeq_clr_dist, spe_single_coph)</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>spe_complete_coph <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>(spe_complete)</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(physeq_clr_dist, spe_complete_coph)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>spe_upgma_coph <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>(spe_upgma)</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(physeq_clr_dist, spe_upgma_coph)</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>spe_ward_coph <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>(spe_ward)</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(physeq_clr_dist, spe_ward_coph)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Which dendrogram retains the closest relationship to the Aitchinson distance matrix?</em></p>
<p>To illustrate the relationship between a distance matrix and a set of cophenetic matrices obtained from various methods, one can draw Shepard-like diagrams (Legendre and Legendre 1998, p.&nbsp;377) by plotting the original distances against the cophenetic distances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>plot_coph_cor <span class="ot">&lt;-</span> <span class="cf">function</span>(cophenetic_distance, hclust_type){</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># first calculate the correlation between</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the cophenetic distance and the observed distance</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  cor_res <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">cor</span>(physeq_clr_dist, cophenetic_distance),<span class="dv">3</span>)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># generate a scatter plot to visualise</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># the relationship</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(<span class="at">x =</span> physeq_clr_dist,</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">y =</span> cophenetic_distance,</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Aitchison distance"</span>,</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Cophenetic distance"</span>,</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">35</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">35</span>),</span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="fu">c</span>(hclust_type, <span class="fu">paste</span>(<span class="st">"Cophenetic correlation "</span>, cor_res)))</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_coph_cor</span>(<span class="at">cophenetic_distance =</span> spe_complete_coph,</span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>              <span class="at">hclust_type =</span> <span class="st">"Single linkage"</span>)</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_coph_cor</span>(<span class="at">cophenetic_distance =</span> spe_complete_coph,</span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>              <span class="at">hclust_type =</span> <span class="st">"Complete linkage"</span>)</span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_coph_cor</span>(<span class="at">cophenetic_distance =</span> spe_upgma_coph,</span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a>              <span class="at">hclust_type =</span> <span class="st">"Average linkage"</span>)</span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_coph_cor</span>(<span class="at">cophenetic_distance =</span> spe_ward_coph,</span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a>              <span class="at">hclust_type =</span> <span class="st">"Ward linkage"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>It seems clear that the UPGMA method give the most faithful representation of original distances.</p>
</section>
<section id="looking-for-interpretable-clusters" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="looking-for-interpretable-clusters"><span class="header-section-number">5.3</span> Looking for Interpretable Clusters</h2>
<p>To interpret and compare clustering results, users generally look for interpretable clusters. This means that a decision must be made: at what level should the dendrogram be cut? Many indices (more than 30) has been published in the literature for finding the right number of clusters in a dataset. The process has been covered <a href="http://www.sthda.com/english/wiki/determining-the-optimal-number-of-clusters-3-must-known-methods-unsupervised-machine-learning">here</a>. The fusion level values of a dendrogram are the dissimilarity values where a fusion between two branches of a dendrogram occurs. Plotting the fusion level values may help define cutting levels. Let us plot the fusion level values for for the UPGMA dendrogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Fusion level plot</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> spe_upgma<span class="sc">$</span>height,</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">y =</span> phyloseq<span class="sc">::</span><span class="fu">nsamples</span>(physeq_clr)<span class="sc">:</span><span class="dv">2</span>,</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">"S"</span>,</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Fusion levels - Aitchison - Average"</span>,</span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"k (number of cluster)"</span>,</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"h (node height)"</span>)</span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> spe_upgma<span class="sc">$</span>height,</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">y =</span> phyloseq<span class="sc">::</span><span class="fu">nsamples</span>(physeq_clr)<span class="sc">:</span><span class="dv">2</span>,</span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">labels =</span> phyloseq<span class="sc">::</span><span class="fu">nsamples</span>(physeq_clr)<span class="sc">:</span><span class="dv">2</span>,</span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="st">"red"</span>,</span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex =</span> <span class="fl">0.8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>From right to left, this first graph shows clear jumps after each fusion between 2 groups. We’ll use the package NbClust which will compute, with a single function call, 24 indices for confirming the right number of clusters in the dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"NbClust"</span>, <span class="at">lib =</span> <span class="st">"."</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"NbClust"</span>, <span class="at">lib.loc =</span> <span class="st">"."</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>nclust <span class="ot">&lt;-</span> <span class="fu">nb_clust_all</span>(<span class="at">data =</span> <span class="fu">t</span>(physeq_clr_asv), <span class="at">seed =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Trying kl index..."
[1] "Trying ch index..."
[1] "Trying hartigan index..."
[1] "Trying scott index..."
[1] "Trying cindex index..."
[1] "Trying db index..."
[1] "Trying silhouette index..."
[1] "Trying duda index..."
[1] "Trying pseudot2 index..."
[1] "Trying beale index..."</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in pf(beale, pp, df2): Production de NaN</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Trying ratkowsky index..."
[1] "Trying ball index..."
[1] "Trying ptbiserial index..."
[1] "Trying gap index..."
[1] "Trying frey index..."
[1] "Trying mcclain index..."
[1] "Trying gamma index..."
[1] "Trying gplus index..."
[1] "Trying tau index..."
[1] "Trying dunn index..."
[1] "Trying hubert index..."</code></pre>
</div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>*** : The Hubert index is a graphical method of determining the number of clusters.
                In the plot of Hubert index, we seek a significant knee that corresponds to a 
                significant increase of the value of the measure i.e the significant peak in Hubert
                index second differences plot. 
 
[1] "Trying sdindex index..."
[1] "Trying dindex index..."</code></pre>
</div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-32-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>*** : The D index is a graphical method of determining the number of clusters. 
                In the plot of D index, we seek a significant knee (the significant peak in Dindex
                second differences plot) that corresponds to a significant increase of the value of
                the measure. 
 
[1] "Trying sdbw index..."
Based on a number of criteria, we will select 3 clusters.</code></pre>
</div>
</div>
<p>NbClust confirm the identification of two clusters of samples. We will go back to the dendrogram and cut it at the corresponding distances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Cut the dendrogram in order to obtain K groups and compare their compositionC</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="co"># Number of groups given by the fusion level plot</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Cut the dendrogram</span></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>spe_upgma_clust <span class="ot">&lt;-</span> <span class="fu">cutree</span>(<span class="at">tree =</span> spe_upgma, <span class="at">k =</span> k)</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(spe_upgma_clust)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spe_upgma_clust
 1  2 
12  6 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>spe_upgma_clust2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">UPGMA_clusters =</span> spe_upgma_clust)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot dendrogram with group labels</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(spe_upgma,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">hang =</span> <span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Height"</span>,</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Aitchison distance - UPGMA"</span>)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rect.hclust</span>(spe_upgma,</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">k =</span> k,</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">border =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>,</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">cluster =</span> spe_upgma_clust)</span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>,</span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>       <span class="fu">paste</span>(<span class="st">"Cluster"</span>, <span class="dv">1</span><span class="sc">:</span>k),</span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">pch =</span> <span class="dv">22</span>,</span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="dv">2</span><span class="sc">:</span>(k <span class="sc">+</span> <span class="dv">1</span>),</span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">bty =</span> <span class="st">"n"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>Do the groups obtained make sense? Do you obtain enough groups containing a substantial number of sites?</em></p>
<p>There are several ways to measure the robustness of a clustering algorithm. Three commonly used metrics are the Dunn index, Davis-Bouldin index and Silhoutte index. The Dunn index is calculated as a ratio of the smallest inter-cluster distance to the largest intra-cluster distance. A high DI means better clustering since observations in each cluster are closer together, while clusters themselves are further away from each other. We’ll use the function <em>cluster.stats()</em> in <strong>fpc</strong> package for computing the Dunn index which can be used for cluster validation,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>cs <span class="ot">&lt;-</span> fpc<span class="sc">::</span><span class="fu">cluster.stats</span>(<span class="at">d =</span> physeq_clr_dist,</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">clustering =</span> spe_upgma_clust)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>cs<span class="sc">$</span>dunn</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8369982</code></pre>
</div>
</div>
<p>The Dunn index is high indicating a good clustering of samples. Now that we identified two groups of samples based on their microbial community composition, we may want to look at which microbial clades or ASVs are enriched in each of the groups.</p>
</section>
<section id="combining-clustering-and-z-score-heatmap" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="combining-clustering-and-z-score-heatmap"><span class="header-section-number">5.4</span> Combining clustering and Z-score Heatmap</h2>
<p><strong>Z-score</strong> heatmap are normalized (centered around the mean (by line!!) &amp; reduced (Standard deviation= SD). It’s the comparison on an observed value of a sample to the mean of the population. So, it answers to the question, how far from the population mean is a score for a given sample. The scores are given in SD to the population mean.</p>
<section id="select-the-top-30-asv" class="level3" data-number="5.4.1">
<h3 data-number="5.4.1" class="anchored" data-anchor-id="select-the-top-30-asv"><span class="header-section-number">5.4.1</span> Select the top 30 ASV</h3>
<p>Here we used ASV but of course you can do it on Family or genus taxonomic level!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Transform Row/normalized counts in percentage: transform_sample_counts</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>pourcentS <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">transform_sample_counts</span>(physeq_rar, <span class="cf">function</span>(x) x<span class="sc">/</span><span class="fu">sum</span>(x) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Selection of top 30 taxa </span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>mytop30 <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">sort</span>(phyloseq<span class="sc">::</span><span class="fu">taxa_sums</span>(pourcentS), <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>])</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Extraction of taxa from the object pourcentS</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>selection30 <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">prune_taxa</span>(mytop30, pourcentS)</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="co">#See new object with only the top 30 ASV</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>selection30</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="get-the-otu_table-z-score-transformation" class="level3" data-number="5.4.2">
<h3 data-number="5.4.2" class="anchored" data-anchor-id="get-the-otu_table-z-score-transformation"><span class="header-section-number">5.4.2</span> Get the otu_table &amp; Z-score transformation</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Retrieve abundance of ASV (otu_table) as table &amp; put in data.prop variable</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>selection30_asv <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">otu_table</span>(selection30)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>selection30_sample <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">sample_data</span>(selection30)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Change the rownames</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="co">#See</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(selection30_asv)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "S11B" "S1B"  "S2B"  "S2S"  "S3B"  "S3S"  "S4B"  "S4S"  "S5B"  "S5S" 
[11] "S6B"  "S6S"  "S7B"  "S7S"  "S8B"  "S8S"  "S9B"  "S9S" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Change... Why?</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="co"># rownames(data.prop)&lt;-c("S11B_South5B","S1B_North1B","S2B_North2B","S2S_North2S","S3B_North3B","S3S_North3S","S4B_North4B","S4S_North4S","S5B_North5B","S5S_North5S","S6B_South1B","S6S_South1S","S7B_South2B","S7S_South2S","S8B_South3B","S8S_South3S","S9B_South4B","S9S_South4S")</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>sample_new_names <span class="ot">&lt;-</span> <span class="fu">paste</span>(selection30_sample<span class="sc">$</span>SampName,</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>                          selection30_sample<span class="sc">$</span>Description,</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Z-score transformation (with scale)</span></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>heat <span class="ot">&lt;-</span> <span class="fu">t</span>(base<span class="sc">::</span><span class="fu">scale</span>(selection30_asv))</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a><span class="co">#See</span></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">data.frame</span>(heat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           S11B         S1B        S2B        S2S        S3B         S3S
ASV1  1.0670101 -0.36085474 -0.8368097  0.5070631 -0.6688256  0.08710287
ASV2 -0.3822681 -0.72549212 -0.7254921  0.5166521 -0.7254921 -0.59474010
ASV3  1.4657223 -1.12279860 -0.5254476  0.6692543 -0.3761099 -2.16816282
ASV4  0.2776466 -1.18129127 -0.9019202 -0.8087965 -1.1812913 -0.77775527
ASV5  1.1642633  0.31674810  0.2397013  1.3954038  0.3552715  0.20117785
ASV6  0.4514863 -0.01289961  0.7417276  0.3934381  1.4963548 -1.81239520
            S4B        S4S         S5B        S5S         S6B         S6S
ASV1 -1.7327249 -0.3608547  1.48697039  2.2149015  1.48697039 -0.47284414
ASV2 -0.6110841 -0.6764601 -0.49667608 -0.7254921  0.38590007  3.34416457
ASV3 -0.6747854 -0.7245646  1.06748833 -1.0232401 -0.02765514  0.02212411
ASV4 -0.3121368 -1.1812913  1.67450193  1.1778422 -0.68463158 -0.56046666
ASV5 -0.3766734  0.5864120 -1.30123544  0.2782247  1.43392721 -1.30123544
ASV6  0.7997758 -1.8123952  0.04514863 -1.8123952 -0.59338206 -0.59338206
            S7B         S7S        S8B        S8S        S9B          S9S
ASV1 -0.8928044  0.03110817 -0.6128309 -0.3888521 -0.5568362  0.003110817
ASV2  0.9742842 -0.18614003  0.4349321 -0.5293641  0.4022441  0.320524054
ASV3  0.6194751 -0.22677213  0.5696958  1.8639563 -0.1769929  0.768812836
ASV4  0.8363887  1.02263609  1.2088835  1.1778422  0.8053475 -0.591507891
ASV5 -1.3012354 -1.30123544 -1.3012354  0.3552715  1.2798335 -0.723384173
ASV6 -0.1870443  0.10319688  0.7417276  0.2192934  0.5095346  1.322210022</code></pre>
</div>
</div>
</section>
<section id="heat-map-z-score" class="level3" data-number="5.4.3">
<h3 data-number="5.4.3" class="anchored" data-anchor-id="heat-map-z-score"><span class="header-section-number">5.4.3</span> Heat map Z-score</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  heat,</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_names_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">6</span>),</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">direction =</span> <span class="st">"vertical"</span>,</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>                              <span class="at">title =</span> <span class="st">"Z-scores"</span>, </span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>                              <span class="at">grid_width =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>),</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">legend_height =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"cm"</span>))</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="add-the-taxonomy-for-asv-names" class="level3" data-number="5.4.4">
<h3 data-number="5.4.4" class="anchored" data-anchor-id="add-the-taxonomy-for-asv-names"><span class="header-section-number">5.4.4</span> Add the Taxonomy for ASV names</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co">#get taxnomic table</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>taxon <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">tax_table</span>(selection30) <span class="sc">|&gt;</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co">#concatene ASV with Phylum &amp; Family names</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>myname <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">rownames</span>(taxon), taxon<span class="sc">$</span>Phylum, taxon<span class="sc">$</span>Family, <span class="at">sep=</span><span class="st">"_"</span>)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="co">#apply</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(selection30_asv) <span class="ot">&lt;-</span> myname</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="apply-to-the-heatmap" class="level3" data-number="5.4.5">
<h3 data-number="5.4.5" class="anchored" data-anchor-id="apply-to-the-heatmap"><span class="header-section-number">5.4.5</span> Apply to the Heatmap</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">#re-run Z-score to take into account the colnames change</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>heat <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">scale</span>(selection30_asv))</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>my_top_annotation <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">anno_block</span>(<span class="at">gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fill =</span><span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>)),</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">labels =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>),</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>                                               <span class="at">labels_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">"white"</span>,</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>                                                                      <span class="at">fontsize =</span> <span class="dv">10</span>))</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>  heat,</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_names_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">6</span>),</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_columns =</span><span class="cn">TRUE</span>,</span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">direction =</span> <span class="st">"vertical"</span>,</span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>   <span class="at">title =</span><span class="st">"Z-scores"</span>,</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>   <span class="at">grid_width =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>),</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>   <span class="at">legend_height =</span> <span class="fu">unit</span>(<span class="dv">4</span>, <span class="st">"cm"</span>)),</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">top_annotation =</span> ComplexHeatmap<span class="sc">::</span><span class="fu">HeatmapAnnotation</span>(<span class="at">foo =</span> my_top_annotation),</span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">column_km =</span> <span class="dv">2</span>,</span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">column_names_gp=</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">6</span>)</span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="add-a-boxplot-of-asv-abundance-distribution-within-sample" class="level3" data-number="5.4.6">
<h3 data-number="5.4.6" class="anchored" data-anchor-id="add-a-boxplot-of-asv-abundance-distribution-within-sample"><span class="header-section-number">5.4.6</span> Add a boxplot of ASV Abundance distribution within sample</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>boxplot <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">anno_boxplot</span>(<span class="fu">t</span>(selection30_asv), </span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">which =</span> <span class="st">"row"</span>,</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fill =</span> <span class="st">"turquoise3"</span>))</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>my_boxplot_left_anno <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">HeatmapAnnotation</span>(<span class="at">Abund =</span> boxplot,</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>                                                          <span class="at">which =</span> <span class="st">"row"</span>,</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>                                                          <span class="at">width =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"cm"</span>))</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>my_top_anno <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">anno_block</span>(<span class="at">gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fill =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">6</span>)),</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"South"</span>, <span class="st">"North"</span>),</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">labels_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">col =</span> <span class="st">"white"</span>,</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>                                                                <span class="at">fontsize =</span> <span class="dv">10</span>))</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>my_top_anno <span class="ot">&lt;-</span> ComplexHeatmap<span class="sc">::</span><span class="fu">HeatmapAnnotation</span>(<span class="at">foo =</span> my_top_anno)</span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(</span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a>  heat,</span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">row_names_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">7</span>),</span>
<span id="cb69-19"><a href="#cb69-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">left_annotation =</span> my_boxplot_left_anno, </span>
<span id="cb69-20"><a href="#cb69-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">direction =</span> <span class="st">"vertical"</span>,</span>
<span id="cb69-21"><a href="#cb69-21" aria-hidden="true" tabindex="-1"></a>                              <span class="at">title =</span><span class="st">"Z-scores"</span>,</span>
<span id="cb69-22"><a href="#cb69-22" aria-hidden="true" tabindex="-1"></a>                              <span class="at">grid_width =</span> <span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"cm"</span>),</span>
<span id="cb69-23"><a href="#cb69-23" aria-hidden="true" tabindex="-1"></a>                              <span class="at">legend_height =</span> <span class="fu">unit</span>(<span class="dv">3</span>, <span class="st">"cm"</span>)),</span>
<span id="cb69-24"><a href="#cb69-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">top_annotation =</span> my_top_anno,</span>
<span id="cb69-25"><a href="#cb69-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">column_km =</span> <span class="dv">2</span>,</span>
<span id="cb69-26"><a href="#cb69-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_columns =</span> <span class="cn">TRUE</span>,</span>
<span id="cb69-27"><a href="#cb69-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">column_dend_side =</span> <span class="st">"bottom"</span>,</span>
<span id="cb69-28"><a href="#cb69-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">column_names_gp =</span> grid<span class="sc">::</span><span class="fu">gpar</span>(<span class="at">fontsize =</span> <span class="dv">7</span>)</span>
<span id="cb69-29"><a href="#cb69-29" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can now observe that microbial communities in samples from the south differ in their microbial composition from sample from the north. The significant effect of treatment (North/south) remains to be tested statistically, we’ll see how it is done in the hypothese testing section. This difference in community composition is due to the apparent differential abundance of many top ASV of the dataset. The identification of significant biomarkers in North and South samples will be covered in the differential abundance testing section.</p>
</section>
</section>
</section>
<section id="indirect-gradient-analysis" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Indirect gradient analysis</h1>
<p>While cluster analysis looks for discontinuities in a dataset, ordination extracts the main trends in the form of continuous axes. It is therefore particularly well adapted to analyse data from natural ecological communities, which are generally structured in gradients. That is why these type of analysis are called gradient analysis. The aim of ordination methods is to represent the data along a reduced number of orthogonal axes, constructed in such a way that they represent, in decreasing order, the main trends of the data. We’ll see four types of analysis widely used in ecology: PCA, PCoA, NMDS. All these methods are descriptive: no statistical test is provided to assess the significance of the structures detected. That is the role of constrained ordination or hypothese testing analysis that are presented in the following chapters.</p>
<section id="principal-component-analysis-pca" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="principal-component-analysis-pca"><span class="header-section-number">6.1</span> Principal component analysis (PCA)</h2>
<p>Principal components analysis (PCA) is a method to summarise, in a low-dimensional space, the variance in a multivariate scatter of points. In doing so, it provides an overview of linear relationships between your objects and variables. This can often act as a good starting point in multivariate data analysis by allowing you to note trends, groupings, key variables, and potential outliers. Again, because of the compositional nature of microbiome data, we will use the Aitchinson distance here. Be aware that this is the CLR transformed ASV table which is used directly not the Aitchinson distance matrix. The function will calculate a euclidean distance on this CLR transformed table to get the Aitchison matrix. There are many packages allowing PCA analysis. We will use the recent <a href="https://bioconductor.org/packages/devel/bioc/vignettes/PCAtools/inst/doc/PCAtools.html" target="_blank">PCAtools</a> package wich provides functions for data exploration via PCA, and allows the user to generate publication-ready figures.</p>
<section id="number-of-pcs-to-retain" class="level3" data-number="6.1.1">
<h3 data-number="6.1.1" class="anchored" data-anchor-id="number-of-pcs-to-retain"><span class="header-section-number">6.1.1</span> Number of PCs to retain</h3>
<p>First, we will use a scree plot to examine the proportion of total variation explained by each PC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co">#prepare the ASV table to add taxonomy</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>tax_CLR <span class="ot">&lt;-</span>  <span class="fu">as.data.frame</span>(<span class="fu">tax_table</span>(physeq_clr)) <span class="co">#get taxnomic table</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="co">#concatene ASV with Family &amp; Genus names</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>ASVname <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">rownames</span>(tax_CLR), tax_CLR<span class="sc">$</span>Family, tax_CLR<span class="sc">$</span>Genus,<span class="at">sep=</span><span class="st">"_"</span>)</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a><span class="co">#apply </span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(physeq_clr_asv) <span class="ot">&lt;-</span> ASVname</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> PCAtools<span class="sc">::</span><span class="fu">pca</span>(physeq_clr_asv,</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">metadata =</span> <span class="fu">data.frame</span>(<span class="fu">sample_data</span>(physeq_clr)))</span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>PCAtools<span class="sc">::</span><span class="fu">screeplot</span>(p, <span class="at">axisLabSize =</span> <span class="dv">18</span>, <span class="at">titleLabSize =</span> <span class="dv">22</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co">#variance explained by each PC</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>variance</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         PC1          PC2          PC3          PC4          PC5          PC6 
4.426841e+01 1.107721e+01 7.288920e+00 5.906096e+00 5.371601e+00 4.700098e+00 
         PC7          PC8          PC9         PC10         PC11         PC12 
3.653510e+00 3.487138e+00 2.890673e+00 2.445228e+00 2.020781e+00 1.681659e+00 
        PC13         PC14         PC15         PC16         PC17         PC18 
1.525155e+00 1.333803e+00 1.026759e+00 7.296512e-01 5.933018e-01 5.408574e-31 </code></pre>
</div>
</div>
<p>Here we see that the first PC really stands out with 31% of the variance explained and then we have a gradual decline for the remaining components. A scree plot on its own just shows the accumulative proportion of explained variation, but we want to determine the optimum number of PCs to retain.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Horn’s parallel analysis (Horn 1965) (Buja and Eyuboglu 1992)</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>horn <span class="ot">&lt;-</span> PCAtools<span class="sc">::</span><span class="fu">parallelPCA</span>(physeq_clr_asv)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>horn<span class="sc">$</span>n</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co">#elbow method</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>elbow <span class="ot">&lt;-</span> PCAtools<span class="sc">::</span><span class="fu">findElbowPoint</span>(p<span class="sc">$</span>variance)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>elbow</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PC3 
  3 </code></pre>
</div>
</div>
<p>The two methods indicate that we should retain the first 2 or 3 PCs. The reason for this discrepancy is because finding the correct number of PCs is a difficult task and is akin to finding the ‘correct’ number of clusters in a dataset - there is no correct answer. Most studies take into account only the two first PCs.</p>
</section>
<section id="plotting-the-ordination" class="level3" data-number="6.1.2">
<h3 data-number="6.1.2" class="anchored" data-anchor-id="plotting-the-ordination"><span class="header-section-number">6.1.2</span> Plotting the ordination</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Plotting the PCA</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>PCAtools<span class="sc">::</span><span class="fu">biplot</span>(</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>  p,</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">lab =</span> p<span class="sc">$</span>metadata<span class="sc">$</span>SampName,</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">colby =</span> <span class="st">"Geo"</span>,</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pointSize =</span> <span class="dv">5</span>,</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">hline =</span> <span class="dv">0</span>, <span class="at">vline =</span> <span class="dv">0</span>,</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">legendPosition =</span> <span class="st">"right"</span></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Each point is a sample, and samples that appear closer together are typically more similar to each other than samples which are further apart. So by colouring the points by treatment you can see that the microbiota from the North are often, but not always, highly distinct from sample from the south.</p>
</section>
<section id="determine-the-variables-that-drive-variation-among-each-pc" class="level3" data-number="6.1.3">
<h3 data-number="6.1.3" class="anchored" data-anchor-id="determine-the-variables-that-drive-variation-among-each-pc"><span class="header-section-number">6.1.3</span> Determine the variables that drive variation among each PC</h3>
<p>One benefit of not using a distance matrix, is that you can plot taxa “loadings” onto your PCA axes, using the <em>showLoadings = TRUE</em> argument. PCAtools allow you to plots the number of taxa loading vectors you want beginning by those having the more weight on each PCs. The relative length of each loading vector indicates its contribution to each PCA axis shown, and allows you to roughly estimate which samples will contain more of that taxon.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>PCAtools<span class="sc">::</span><span class="fu">biplot</span>(</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>  p, </span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># loadings parameters</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">showLoadings =</span> <span class="cn">TRUE</span>,</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">lengthLoadingsArrowsFactor =</span> <span class="fl">1.5</span>,</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">sizeLoadingsNames =</span> <span class="dv">3</span>,</span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">colLoadingsNames =</span> <span class="st">'red4'</span>,</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ntopLoadings =</span> <span class="dv">3</span>,</span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># other parameters</span></span>
<span id="cb78-10"><a href="#cb78-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">lab =</span> p<span class="sc">$</span>metadata<span class="sc">$</span>X.SampleID,</span>
<span id="cb78-11"><a href="#cb78-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">colby =</span> <span class="st">"Geo"</span>,</span>
<span id="cb78-12"><a href="#cb78-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">hline =</span> <span class="dv">0</span>, <span class="at">vline =</span> <span class="dv">0</span>,</span>
<span id="cb78-13"><a href="#cb78-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">legendPosition =</span> <span class="st">"right"</span></span>
<span id="cb78-14"><a href="#cb78-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>ASVs 7, 11 and 12 have a high contribution to PC1 while ASVs 38, 40, and 47 have a high contribution on the second PC. These ASVs belong to only two families. Samples from the South seems to be enriched in ASV7 while North samples contain higher abundances of ASV11 and 12. The two Noth sample outliers at the top of the plot are caracterized by a higher abundance of ASVs 38, 40, and 47.</p>
</section>
<section id="correlate-the-principal-components-back-to-environmental-data" class="level3" data-number="6.1.4">
<h3 data-number="6.1.4" class="anchored" data-anchor-id="correlate-the-principal-components-back-to-environmental-data"><span class="header-section-number">6.1.4</span> Correlate the principal components back to environmental data</h3>
<p>Further exploration of the PCs can come through correlations with environmental data. Here, we will correlate the two first PCs with environmental data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>PCAtools<span class="sc">::</span><span class="fu">eigencorplot</span>(</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  p,</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">components =</span> PCAtools<span class="sc">::</span><span class="fu">getComponents</span>(p, <span class="dv">1</span><span class="sc">:</span>horn<span class="sc">$</span>n),</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">metavars =</span> <span class="fu">c</span>(<span class="st">'SiOH4'</span>,<span class="st">'NO2'</span>,<span class="st">'NO3'</span>,<span class="st">'NH4'</span>,<span class="st">'PO4'</span>,</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>              <span class="st">'NT'</span>,<span class="st">'PT'</span>,<span class="st">'Chla'</span>,<span class="st">"T"</span>, <span class="st">"S"</span>, <span class="st">"Sigma_t"</span>),</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">'white'</span>, <span class="st">'cornsilk1'</span>, <span class="st">'gold'</span>,</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>          <span class="st">'forestgreen'</span>, <span class="st">'darkgreen'</span>),</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cexCorval =</span> <span class="fl">1.2</span>,</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">fontCorval =</span> <span class="dv">2</span>,</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">posLab =</span> <span class="st">"all"</span>,</span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">rotLabX =</span> <span class="dv">45</span>,</span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="cn">TRUE</span>,</span>
<span id="cb79-13"><a href="#cb79-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="fu">bquote</span>(PC <span class="sc">~</span> Spearman <span class="sc">~</span> r<span class="sc">^</span><span class="dv">2</span> <span class="sc">~</span> environmental <span class="sc">~</span> correlates),</span>
<span id="cb79-14"><a href="#cb79-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">plotRsquared =</span> <span class="cn">TRUE</span>,</span>
<span id="cb79-15"><a href="#cb79-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">corFUN =</span> <span class="st">"spearman"</span>,</span>
<span id="cb79-16"><a href="#cb79-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">corUSE =</span> <span class="st">"pairwise.complete.obs"</span>,</span>
<span id="cb79-17"><a href="#cb79-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">corMultipleTestCorrection =</span> <span class="st">'BH'</span>,</span>
<span id="cb79-18"><a href="#cb79-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">signifSymbols =</span> <span class="fu">c</span>(<span class="st">"****"</span>, <span class="st">"***"</span>, <span class="st">"**"</span>, <span class="st">"*"</span>, <span class="st">""</span>),</span>
<span id="cb79-19"><a href="#cb79-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">signifCutpoints =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.0001</span>, <span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="dv">1</span>)</span>
<span id="cb79-20"><a href="#cb79-20" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The only signficant correlation found is between the first PC1 explaining the separation between South and North samples and salinity. This is unteresting but correlation between variables does not automatically means that the change in one variable is the cause of the change in the values of the other variable. We’ll check later if there is a causal relationship between the salinity gradient and the difference observed in southern and northern bacterial communities.</p>
</section>
</section>
<section id="principal-component-analysis-pcoa" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="principal-component-analysis-pcoa"><span class="header-section-number">6.2</span> Principal component analysis (PCoA)</h2>
<p>Principal coordinates analysis (PCoA, also known as metric multidimensional scaling, MDS) attempts to represent the distances between samples in a low-dimensional, Euclidean space. In particular, it maximizes the linear correlation between the distances in the distance matrix, and the distances in a space of low dimension (typically, 2 or 3 axes are selected). As always, the choice of (dis)similarity measure is critical and must be suitable to the data in question. Here we will use the Bray-Curtis distance. When the distance metric is Euclidean, PCoA is equivalent to Principal Components Analysis. The interpretation of the results is the same as with PCA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co">#BPCoA on Bray-Curtis dissimilarity</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>pcoa_asv <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">pcoa</span>(physeq_rar_bray)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>pcoa_coord <span class="ot">&lt;-</span> pcoa_asv<span class="sc">$</span>vectors[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Data frame for hull</span></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>hull <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">"Axis.1"</span> <span class="ot">=</span> pcoa_coord[, <span class="dv">1</span>],</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"Axis.2"</span> <span class="ot">=</span> pcoa_coord[, <span class="dv">2</span>],</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"sample"</span> <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">sample_data</span>(physeq_rar<span class="sc">@</span>sam_data)))</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a><span class="co"># North &lt;- hull[hull$sample.Geo  == "North", ][chull(hull[hull$sample.Geo ==  "North", c("Axis.1", "Axis.2")]), ]  # hull values for North</span></span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a><span class="co"># South &lt;- hull[hull$sample.Geo == "South", ][chull(hull[hull$sample.Geo == </span></span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                          "South", c("Axis.1", "Axis.2")]), ]  # hull values for Jellyfishes  </span></span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a><span class="co"># hull_data &lt;- rbind(North, South)</span></span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-17"><a href="#cb80-17" aria-hidden="true" tabindex="-1"></a><span class="co">#Vector of color for hulls</span></span>
<span id="cb80-18"><a href="#cb80-18" aria-hidden="true" tabindex="-1"></a><span class="co"># color &lt;- rep("#a65628", length(hull_data$sample.Geo))</span></span>
<span id="cb80-19"><a href="#cb80-19" aria-hidden="true" tabindex="-1"></a><span class="co"># color[hull_data$sample.Geo == "North"] &lt;- "#1919ff"</span></span>
<span id="cb80-20"><a href="#cb80-20" aria-hidden="true" tabindex="-1"></a><span class="co"># hull_data &lt;- cbind(hull_data, color)</span></span>
<span id="cb80-21"><a href="#cb80-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-22"><a href="#cb80-22" aria-hidden="true" tabindex="-1"></a>hull_col <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#a65628"</span>,<span class="st">"#1919ff"</span>)</span>
<span id="cb80-23"><a href="#cb80-23" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(hull_col) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"North"</span>,<span class="st">"South"</span>)</span>
<span id="cb80-24"><a href="#cb80-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-25"><a href="#cb80-25" aria-hidden="true" tabindex="-1"></a>hull_data <span class="ot">&lt;-</span> hull <span class="sc">%&gt;%</span></span>
<span id="cb80-26"><a href="#cb80-26" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(sample.Geo) <span class="sc">%&gt;%</span></span>
<span id="cb80-27"><a href="#cb80-27" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="fu">chull</span>(Axis<span class="fl">.1</span>,Axis<span class="fl">.2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb80-28"><a href="#cb80-28" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">color =</span> hull_col[sample.Geo])</span>
<span id="cb80-29"><a href="#cb80-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-30"><a href="#cb80-30" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hull_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 24
# Groups:   sample.Geo [1]
    Axis.1  Axis.2 sample.SampName sample.Geo sample.Description sample.groupe
     &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;      &lt;chr&gt;              &lt;chr&gt;        
1  0.242   -0.0992 S2S             North      North2S            NBS          
2 -0.403   -0.130  S2B             North      North2B            NBF          
3 -0.455   -0.0922 S3B             North      North3B            NBF          
4 -0.471    0.0176 S3S             North      North3S            NBS          
5  0.00454  0.407  S5S             North      North5S            NBS          
6  0.102    0.327  S5B             North      North5B            NBF          
# ℹ 18 more variables: sample.Pres &lt;int&gt;, sample.PicoEuk &lt;int&gt;,
#   sample.Synec &lt;int&gt;, sample.Prochloro &lt;int&gt;, sample.NanoEuk &lt;int&gt;,
#   sample.Crypto &lt;int&gt;, sample.SiOH4 &lt;dbl&gt;, sample.NO2 &lt;dbl&gt;,
#   sample.NO3 &lt;dbl&gt;, sample.NH4 &lt;dbl&gt;, sample.PO4 &lt;dbl&gt;, sample.NT &lt;dbl&gt;,
#   sample.PT &lt;dbl&gt;, sample.Chla &lt;dbl&gt;, sample.T &lt;dbl&gt;, sample.S &lt;dbl&gt;,
#   sample.Sigma_t &lt;dbl&gt;, color &lt;chr&gt;</code></pre>
</div>
</div>
<p>Now that we prepared the data, lets plot the PCoA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> hull, <span class="fu">aes</span>(<span class="at">x =</span> Axis<span class="fl">.1</span>, <span class="at">y =</span> Axis<span class="fl">.2</span>)) <span class="sc">+</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"lightgrey"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"lightgrey"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> hull_data,</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">group =</span> sample.Geo,</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fill =</span> sample.Geo),</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span> <span class="co"># add the convex hulls)</span></span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Darkgrey"</span>, <span class="st">"#1919ff"</span>)) <span class="sc">+</span></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> hull,</span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">color =</span> sample.Geo,</span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>                 <span class="at">size =</span> sample.S),</span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Darkgrey"</span>, <span class="st">"#1919ff"</span>)) <span class="sc">+</span></span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste</span>(<span class="st">"PCo1 ("</span>, <span class="fu">round</span>(pcoa_asv<span class="sc">$</span>values<span class="sc">$</span>Relative_eig[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste</span>(<span class="st">"PCo2 ("</span>, <span class="fu">round</span>(pcoa_asv<span class="sc">$</span>values<span class="sc">$</span>Relative_eig[<span class="dv">2</span>]<span class="sc">*</span><span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb82-17"><a href="#cb82-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb82-18"><a href="#cb82-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>), <span class="co"># remove x-axis labels</span></span>
<span id="cb82-19"><a href="#cb82-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>), <span class="co"># remove y-axis labels</span></span>
<span id="cb82-20"><a href="#cb82-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb82-21"><a href="#cb82-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),  <span class="co">#remove major-grid labels</span></span>
<span id="cb82-22"><a href="#cb82-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),  <span class="co">#remove minor-grid labels</span></span>
<span id="cb82-23"><a href="#cb82-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.background =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The ordination of the samples in the PCoA is very similar to the one observed in the PCA with a clear segregation of North and South bacterial communities. This segregation may result from the increasing salinity gradient from North to South but it relaims to be tested.</p>
<p><em>Do you see what is missing?</em></p>
<p>Indeed, there are no species plotted on this ordination. That’s because we used a dissimilarity matrix (sites x sites) as input for the PCoA function. Hence, no species scores could be calculated. However, we could work around this problem with the function <code>biplot.pcoa()</code> from the <code>ape</code> package.</p>
<p>PCoA suffers from a number of flaws, in particular the arch effect (see PCA for more information). These flaws stem, in part, from the fact that PCoA maximizes a linear correlation. Non-metric Multidimensional Scaling (NMDS) rectifies this by maximizing the rank order correlation.</p>
</section>
<section id="non-metric-multidimensional-scaling-nmds" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="non-metric-multidimensional-scaling-nmds"><span class="header-section-number">6.3</span> Non Metric Multidimensional Scaling (NMDS)</h2>
<p>NMDS attempts to represent the pairwise dissimilarity between objects in a low-dimensional space. Any dissimilarity coefficient or distance measure may be used to build the distance matrix used as input. <strong>NMDS is a rank-based approach.</strong> This means that the original distance data is substituted with ranks. While information about the magnitude of distances is lost, rank-based methods are generally more robust to data which do not have an identifiable distribution.</p>
<p>NMDS is an iterative algorithm. NMDS routines often begin by random placement of data objects in ordination space. The algorithm then begins to refine this placement by an iterative process, attempting to find an ordination in which ordinated object distances closely match the order of object dissimilarities in the original distance matrix. The stress value reflects how well the ordination summarizes the observed distances among the samples.</p>
<p>NMDS is not an eigenanalysis. This has three important consequences:</p>
<ul>
<li>There is no unique ordination result</li>
<li>The axes of the ordination are not ordered according to the variance they explain</li>
<li>The number of dimensions of the low-dimensional space must be specified before running the analysis</li>
</ul>
<p>Axes are not ordered in NMDS. <code>vegan::metaMDS()</code> automatically rotates the final result of the NMDS using PCA to make axis 1 correspond to the greatest variance among the NMDS sample points.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co">#NMDS plot on Aitchison distance</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>physeq_clr_nmds <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">metaMDS</span>(physeq_clr_dist, <span class="at">k=</span><span class="dv">2</span>, <span class="at">trymax=</span><span class="dv">100</span>) <span class="co">#Aitchison distance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A useful way to assess the appropriateness of an NMDS result is to compare, in a Shepard diagram, the distances among objects in the ordination plot with the original distances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">stressplot</span>(physeq_clr_nmds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>There is a good non-metric fit between observed dissimilarities (in our distance matrix) and the distances in ordination space. Also the stress of our final result was good.</p>
<p><em>do you know how much the stress is?</em></p>
<p>The stress value can be used as an indicator of the goodness-of-fit. Stress values &gt;0.2 are generally poor and potentially not interpretable, whereas values &lt;0.1 are good and &lt;0.05 are excellent, leaving little danger of misinterpretation.</p>
<p>So we can go further and plot the results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>nmds_coord <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(physeq_clr_nmds<span class="sc">$</span>points)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Data frame for hull</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>hull <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">"Axis.1"</span> <span class="ot">=</span> nmds_coord[,<span class="dv">1</span>],</span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"Axis.2"</span> <span class="ot">=</span> nmds_coord[,<span class="dv">2</span>],</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"sample"</span> <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">sample_data</span>(physeq_clr<span class="sc">@</span>sam_data)))</span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a><span class="co"># North &lt;- hull[hull$sample.Geo  == "North", ][chull(hull[hull$sample.Geo == </span></span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                                 "North", c("Axis.1", "Axis.2")]), ]  # hull values for North</span></span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a><span class="co"># South &lt;- hull[hull$sample.Geo == "South", ][chull(hull[hull$sample.Geo == </span></span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                                "South", c("Axis.1", "Axis.2")]), ]  # hull values for Jellyfishes  </span></span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a><span class="co"># hull_data &lt;- rbind(North, South)</span></span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-15"><a href="#cb85-15" aria-hidden="true" tabindex="-1"></a><span class="co"># #Vector of color for hulls</span></span>
<span id="cb85-16"><a href="#cb85-16" aria-hidden="true" tabindex="-1"></a><span class="co"># color &lt;- rep("#a65628", length(hull_data$sample.Geo))</span></span>
<span id="cb85-17"><a href="#cb85-17" aria-hidden="true" tabindex="-1"></a><span class="co"># color[hull_data$sample.Geo == "North"] &lt;- "#1919ff"</span></span>
<span id="cb85-18"><a href="#cb85-18" aria-hidden="true" tabindex="-1"></a><span class="co"># hull_data &lt;- cbind(hull_data, color)</span></span>
<span id="cb85-19"><a href="#cb85-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-20"><a href="#cb85-20" aria-hidden="true" tabindex="-1"></a>hull_col <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#a65628"</span>,<span class="st">"#1919ff"</span>)</span>
<span id="cb85-21"><a href="#cb85-21" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(hull_col) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"North"</span>,<span class="st">"South"</span>)</span>
<span id="cb85-22"><a href="#cb85-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-23"><a href="#cb85-23" aria-hidden="true" tabindex="-1"></a>hull_data <span class="ot">&lt;-</span> hull <span class="sc">%&gt;%</span></span>
<span id="cb85-24"><a href="#cb85-24" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(sample.Geo) <span class="sc">%&gt;%</span></span>
<span id="cb85-25"><a href="#cb85-25" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">slice</span>(<span class="fu">chull</span>(Axis<span class="fl">.1</span>,Axis<span class="fl">.2</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb85-26"><a href="#cb85-26" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">color =</span> hull_col[sample.Geo])</span>
<span id="cb85-27"><a href="#cb85-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-28"><a href="#cb85-28" aria-hidden="true" tabindex="-1"></a><span class="co">#pdf(file="NMDS_Aitchison.pdf", wi = 7, he = 7)</span></span>
<span id="cb85-29"><a href="#cb85-29" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(hull,<span class="fu">aes</span>(<span class="at">x =</span> Axis<span class="fl">.1</span>, <span class="at">y =</span> Axis<span class="fl">.2</span>)) <span class="sc">+</span></span>
<span id="cb85-30"><a href="#cb85-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"lightgrey"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb85-31"><a href="#cb85-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"lightgrey"</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb85-32"><a href="#cb85-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_polygon</span>(<span class="at">data =</span> hull_data,</span>
<span id="cb85-33"><a href="#cb85-33" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">group =</span> sample.Geo,</span>
<span id="cb85-34"><a href="#cb85-34" aria-hidden="true" tabindex="-1"></a>                   <span class="at">fill =</span> sample.Geo),</span>
<span id="cb85-35"><a href="#cb85-35" aria-hidden="true" tabindex="-1"></a>               <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span> <span class="co"># add the convex hulls)</span></span>
<span id="cb85-36"><a href="#cb85-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Darkgrey"</span>, <span class="st">"#1919ff"</span>)) <span class="sc">+</span></span>
<span id="cb85-37"><a href="#cb85-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> hull,</span>
<span id="cb85-38"><a href="#cb85-38" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">color =</span> sample.Geo,</span>
<span id="cb85-39"><a href="#cb85-39" aria-hidden="true" tabindex="-1"></a>                 <span class="at">size =</span> sample.S),</span>
<span id="cb85-40"><a href="#cb85-40" aria-hidden="true" tabindex="-1"></a>             <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb85-41"><a href="#cb85-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"Darkgrey"</span>, <span class="st">"#1919ff"</span>)) <span class="sc">+</span></span>
<span id="cb85-42"><a href="#cb85-42" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> hull_data,</span>
<span id="cb85-43"><a href="#cb85-43" aria-hidden="true" tabindex="-1"></a>            <span class="at">x =</span> <span class="sc">-</span><span class="dv">0</span>, <span class="at">y =</span> <span class="sc">-</span><span class="dv">9</span>,</span>
<span id="cb85-44"><a href="#cb85-44" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="fu">paste</span>(<span class="st">"Stress ="</span>, <span class="fu">round</span>(physeq_clr_nmds<span class="sc">$</span>stress, <span class="dv">2</span>)),</span>
<span id="cb85-45"><a href="#cb85-45" aria-hidden="true" tabindex="-1"></a>            <span class="at">colour =</span> <span class="st">"Black"</span>,</span>
<span id="cb85-46"><a href="#cb85-46" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="dv">5</span>)  <span class="sc">+</span></span>
<span id="cb85-47"><a href="#cb85-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste</span>(<span class="st">"MDS1"</span>)) <span class="sc">+</span></span>
<span id="cb85-48"><a href="#cb85-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste</span>(<span class="st">"MDS2"</span>)) <span class="sc">+</span></span>
<span id="cb85-49"><a href="#cb85-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb85-50"><a href="#cb85-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb85-51"><a href="#cb85-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="co"># remove x-axis labels</span></span>
<span id="cb85-52"><a href="#cb85-52" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="co"># remove y-axis labels</span></span>
<span id="cb85-53"><a href="#cb85-53" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.background =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb85-54"><a href="#cb85-54" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),  <span class="co">#remove major-grid labels</span></span>
<span id="cb85-55"><a href="#cb85-55" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),  <span class="co">#remove minor-grid labels</span></span>
<span id="cb85-56"><a href="#cb85-56" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.background =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co">#dev.off()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We observe the same ordiantion pattern of the samples as in the PCA and PCoA. There are no species scores (same problem as we encountered with PCoA). We can work around this problem, by using the <em>wascores</em> function giving metaMDS the original community matrix as input and specifying the distance measure.</p>
<p>The next question is: Which environmental variable is driving the observed differences in species composition? Similarly to what we have done with PCA, we can correlate environmental variables with our ordination axes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Correlation with environmental data</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="fu">names</span>(hull))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          names.hull.
1              Axis.1
2              Axis.2
3     sample.SampName
4          sample.Geo
5  sample.Description
6       sample.groupe
7         sample.Pres
8      sample.PicoEuk
9        sample.Synec
10   sample.Prochloro
11     sample.NanoEuk
12      sample.Crypto
13       sample.SiOH4
14         sample.NO2
15         sample.NO3
16         sample.NH4
17         sample.PO4
18          sample.NT
19          sample.PT
20        sample.Chla
21           sample.T
22           sample.S
23     sample.Sigma_t</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>env <span class="ot">&lt;-</span> hull[, <span class="dv">13</span><span class="sc">:</span><span class="dv">23</span>]</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The function envfit will add the environmental variables as vectors to the ordination plot</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>ef <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">envfit</span>(physeq_clr_nmds, env, <span class="at">permu =</span> <span class="dv">1000</span>)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>ef</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
***VECTORS

                  NMDS1    NMDS2     r2   Pr(&gt;r)    
sample.SiOH4   -0.73545 -0.67758 0.3142 0.062937 .  
sample.NO2     -0.70256 -0.71162 0.2340 0.149850    
sample.NO3      0.91015  0.41429 0.3105 0.064935 .  
sample.NH4     -0.48733 -0.87322 0.4995 0.006993 ** 
sample.PO4     -0.96817 -0.25030 0.1932 0.203796    
sample.NT      -0.02763 -0.99962 0.0094 0.939061    
sample.PT      -0.59514 -0.80362 0.4299 0.020979 *  
sample.Chla    -0.98935 -0.14553 0.2240 0.142857    
sample.T       -0.86808  0.49643 0.2997 0.077922 .  
sample.S       -0.99976  0.02200 0.7272 0.000999 ***
sample.Sigma_t -0.85483 -0.51890 0.2334 0.163836    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
Permutation: free
Number of permutations: 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The two last columns are of interest: the squared correlation coefficient and the associated p-value</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the vectors of the significant correlations and interpret the plot</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(physeq_clr_nmds, <span class="at">type =</span> <span class="st">"t"</span>, <span class="at">display =</span> <span class="st">"sites"</span>)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ef, <span class="at">p.max =</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here again, we can see that the salinity is strongly correlated with the first axis separating samples from the South and the North. To a lesser extent, new environmental variables related with the trophic conditions of the habitat (NH4 and PT) were correlated with the second axis of the NMDS. The detection of these new relations between microbial communities and the environment may be related to the fact that NMDS is best suited to detect the non linear response of microbes to environmental gradients.</p>
<p>Many different types of indirect gradient analysis are available outthere. In the following graph, we offer suggestions of some of the appropriate choices based on data input structure and expected relationships among variables.</p>
<p><img src="../img/Indirect_gradient_methods.png" class="img-fluid"></p>
</section>
</section>
<section id="hypothese-testing-analyses" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Hypothese testing analyses</h1>
<p>We saw some segregation between Northern and Southern samples suggesting some differences in the bacterial communities according to sample type. While indirect gradient or classification analyses are exploratory data visualization tool, we can test whether the samples cluster beyond that expected using hypotheses testing methods such as multivariate analysis of variance with permutation (PERMANOVA), and analysis of group similarities (ANOSIM), multi-response permutation procedures (MRPP), and Mantel’s test (MANTEL) and more recently Dirichlet-multinomial models.</p>
<section id="permutational-multiple-analysis-of-variance-permanova" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="permutational-multiple-analysis-of-variance-permanova"><span class="header-section-number">7.1</span> PERmutational Multiple ANalysis Of VAriance (PERMANOVA)</h2>
<p>PERMANOVA was proposed by Anderson and McArdle to apply the powerful ANOVA to multivariate ecological datasets. PERMANOVA is one of most widely used nonparametric methods to fit multivariate models to microbiome data. It is a multivariate analysis of variance based on distance matrices and permutation. It does this by partitioning the sums of squares for the within- and between-cluster components using the concept of centroids. Many permutations of the data (i.e.&nbsp;random shuffling) are used to generate the null distribution. Find more informations on PERMANOVA <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/9781118445112.stat07841" target="_blank">here</a> and on the <code>adonis2()</code> function <a href="https://www.researchgate.net/topic/Adonis" target="_blank">here</a> Now let us evaluate whether the group (North vs.&nbsp;South) has a significant effect on overall bacterial community composition.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co">#PERMANOVA</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">sample_data</span>(physeq_clr))</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>results_permanova <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(physeq_clr_dist <span class="sc">~</span> Geo,</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">data =</span> metadata,</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">perm =</span> <span class="dv">1000</span>)</span>
<span id="cb92-6"><a href="#cb92-6" aria-hidden="true" tabindex="-1"></a>results_permanova</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 1000

vegan::adonis2(formula = physeq_clr_dist ~ Geo, data = metadata, permutations = 1000)
         Df SumOfSqs      R2      F   Pr(&gt;F)   
Geo       1    982.0 0.28553 6.3942 0.001998 **
Residual 16   2457.3 0.71447                   
Total    17   3439.3 1.00000                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Here we can see that the North/South grouping explain significantly (p &lt; 0.001) 20% of the variance in the ASV Aitchison matrix. In other words Nothern and Southern bacterial differ significatively in their bacterial composition. The test from ADONIS can be confounded by differences in dispersion (or spread) so we want to check this as well..</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Testing the assumption of similar multivariate spread among the groups (ie. analogous to variance homogeneity)</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(vegan<span class="sc">::</span><span class="fu">betadisper</span>(physeq_clr_dist, metadata<span class="sc">$</span>Geo))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: Distances
          Df  Sum Sq Mean Sq F value  Pr(&gt;F)  
Groups     1  28.854 28.8543  3.3556 0.08566 .
Residuals 16 137.580  8.5987                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Here the groups have significant different spreads and permanova result may be impacted by that although PERMANOVA is very robust to difference in group dispersion. We can also check which taxa contribute most to the community differences using the ancient adonis() function and the ASV table of CLR transformed counts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Show coefficients for the top taxa separating the groups</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>permanova <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis</span>(<span class="fu">t</span>(physeq_clr_asv) <span class="sc">~</span> Geo,</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">data =</span> metadata,</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">permutations =</span> <span class="dv">1000</span>,</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">"euclidean"</span>)</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>coef <span class="ot">&lt;-</span> <span class="fu">coefficients</span>(permanova)[<span class="st">"Geo1"</span>,]</span>
<span id="cb96-9"><a href="#cb96-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-10"><a href="#cb96-10" aria-hidden="true" tabindex="-1"></a>top.coef <span class="ot">&lt;-</span> coef[<span class="fu">rev</span>(<span class="fu">order</span>(<span class="fu">abs</span>(coef)))[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]]</span>
<span id="cb96-11"><a href="#cb96-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-12"><a href="#cb96-12" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">14</span>, <span class="dv">2</span>, <span class="dv">1</span>))</span>
<span id="cb96-13"><a href="#cb96-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-14"><a href="#cb96-14" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">sort</span>(top.coef),</span>
<span id="cb96-15"><a href="#cb96-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">horiz =</span> <span class="cn">TRUE</span>,</span>
<span id="cb96-16"><a href="#cb96-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">1</span>,</span>
<span id="cb96-17"><a href="#cb96-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Top taxa"</span>,</span>
<span id="cb96-18"><a href="#cb96-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">cex.names =</span> <span class="fl">0.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><em>Are these ASVs the same or different compared to ASV contributing the most to PC axes?</em></p>
<p><code>adonis()</code> and <code>adonis2()</code> allow us to explore the effect of categorical or <a href="https://stats.stackexchange.com/questions/449176/the-way-permanova-handling-continuous-explanatory-variable" target="_blank">continuous</a> variables.</p>
<p><strong>NB</strong>: An important difference between adonis et adonis2: in adonis terms are tested sequentially and this is the only option. This means that the order you enter your variables is important (if design is unbalanced). This is because the first explanatory variable is added to the model. Then the next one is added to see if it explains significantly more variation not explained by the previous variables. This is equivalent to using by=“terms” in adonis2. If you don’t want the order to matter you can use adonis2 with by=“margin”, or if you want to check if the model as a whole is significant you can use by=NULL. Order does not matter when by=“margin” because the significance is tested against a model that includes all other variables not just the ones preceding it in the formula.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Permanova on continuous variables</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>permanova_S <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(physeq_clr_dist <span class="sc">~</span> S,</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">data =</span> metadata,</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">perm =</span> <span class="dv">1000</span>)</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>permanova_S</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 1000

vegan::adonis2(formula = physeq_clr_dist ~ S, data = metadata, permutations = 1000)
         Df SumOfSqs      R2      F   Pr(&gt;F)    
S         1   1126.5 0.32753 7.7929 0.000999 ***
Residual 16   2312.8 0.67247                    
Total    17   3439.3 1.00000                    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>permanova_NH4 <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(physeq_clr_dist <span class="sc">~</span> NH4,</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">data =</span> metadata,</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">perm =</span> <span class="dv">1000</span>)</span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>permanova_NH4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 1000

vegan::adonis2(formula = physeq_clr_dist ~ NH4, data = metadata, permutations = 1000)
         Df SumOfSqs     R2      F  Pr(&gt;F)  
NH4       1    620.4 0.1804 3.5216 0.01698 *
Residual 16   2818.9 0.8196                 
Total    17   3439.3 1.0000                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>permanova_PT <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(physeq_clr_dist <span class="sc">~</span> PT,</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">data =</span> metadata,</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">perm =</span> <span class="dv">1000</span>)</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>permanova_PT</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 1000

vegan::adonis2(formula = physeq_clr_dist ~ PT, data = metadata, permutations = 1000)
         Df SumOfSqs      R2      F  Pr(&gt;F)  
PT        1    567.9 0.16512 3.1645 0.01698 *
Residual 16   2871.4 0.83488                 
Total    17   3439.3 1.00000                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>The result confirm that salinity and to a lesser extent NH4 and PT are important factors shaping microbial communities but what about the other variables? Lets construct a model with all the co-variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Inspecting co-variables</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>permanova_all <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(physeq_clr_dist <span class="sc">~</span> SiOH4 <span class="sc">+</span> NO2 <span class="sc">+</span> NO3 <span class="sc">+</span> NH4 <span class="sc">+</span> PO4 <span class="sc">+</span> NT <span class="sc">+</span> PT <span class="sc">+</span> Chla <span class="sc">+</span> T <span class="sc">+</span> S <span class="sc">+</span> Sigma_t,</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">by=</span><span class="st">"margin"</span>,</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">data=</span>metadata,</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>                                <span class="at">perm=</span><span class="dv">1000</span>)</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>permanova_all</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Marginal effects of terms
Permutation: free
Number of permutations: 1000

vegan::adonis2(formula = physeq_clr_dist ~ SiOH4 + NO2 + NO3 + NH4 + PO4 + NT + PT + Chla + T + S + Sigma_t, data = metadata, permutations = 1000, by = "margin")
         Df SumOfSqs      R2      F Pr(&gt;F)
SiOH4     1    166.7 0.04848 1.0898 0.3257
NO2       1    119.2 0.03467 0.7793 0.5754
NO3       1    126.9 0.03691 0.8297 0.5255
NH4       1    142.8 0.04152 0.9333 0.4466
PO4       1    113.4 0.03299 0.7415 0.6124
NT        1    147.0 0.04273 0.9606 0.4336
PT        1    121.3 0.03527 0.7929 0.5574
Chla      1    104.1 0.03026 0.6802 0.6893
T         1    170.3 0.04950 1.1128 0.3427
S         1    170.0 0.04944 1.1115 0.3427
Sigma_t   1    170.5 0.04958 1.1147 0.3467
Residual  6    917.9 0.26689              
Total    17   3439.3 1.00000              </code></pre>
</div>
</div>
<p><em>What happened? Why none of the variables has no significant effect any more ?</em></p>
<p>The temptation to build an ecological model using all available information (i.e., all variables) is hard to resist. Lots of time and money are exhausted gathering data and supporting information. We also hope to identify every significant variable to more accurately characterize relationships with biological relevance. Collinearity, or excessive correlation among explanatory variables, can complicate or prevent the identification of an optimal set of explanatory variables for a statistical model. Let’s take a look at which explanatory variables are correlated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co"># inpecting autocorrélation</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a><span class="co"># compute the correlation matrix</span></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>cor_metadadata <span class="ot">&lt;-</span> <span class="fu">cor</span>(metadata[, <span class="dv">11</span><span class="sc">:</span><span class="dv">21</span>], <span class="at">method =</span> <span class="st">"spearman"</span>)</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>cor_mtest <span class="ot">&lt;-</span> <span class="cf">function</span>(mat, ...) {</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>  mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(mat)</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(mat)</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>  p_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, n, n)</span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">diag</span>(p_mat) <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb105-10"><a href="#cb105-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(n <span class="sc">-</span> <span class="dv">1</span>)) {</span>
<span id="cb105-11"><a href="#cb105-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> (i <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">:</span>n) {</span>
<span id="cb105-12"><a href="#cb105-12" aria-hidden="true" tabindex="-1"></a>      tmp <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(mat[, i], mat[, j], <span class="at">method =</span> <span class="st">"spearman"</span>, ...)</span>
<span id="cb105-13"><a href="#cb105-13" aria-hidden="true" tabindex="-1"></a>      p_mat[i, j] <span class="ot">&lt;-</span> p_mat[j, i] <span class="ot">&lt;-</span> tmp<span class="sc">$</span>p.value</span>
<span id="cb105-14"><a href="#cb105-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb105-15"><a href="#cb105-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb105-16"><a href="#cb105-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(p_mat) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(p_mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(mat)</span>
<span id="cb105-17"><a href="#cb105-17" aria-hidden="true" tabindex="-1"></a>  p_mat</span>
<span id="cb105-18"><a href="#cb105-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb105-19"><a href="#cb105-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-20"><a href="#cb105-20" aria-hidden="true" tabindex="-1"></a><span class="co"># matrix of the p-value of the correlation</span></span>
<span id="cb105-21"><a href="#cb105-21" aria-hidden="true" tabindex="-1"></a>p_mat <span class="ot">&lt;-</span> <span class="fu">cor_mtest</span>(metadata[, <span class="dv">11</span><span class="sc">:</span><span class="dv">21</span>])</span>
<span id="cb105-22"><a href="#cb105-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-23"><a href="#cb105-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Leave blank on no significant coefficient</span></span>
<span id="cb105-24"><a href="#cb105-24" aria-hidden="true" tabindex="-1"></a>corrplot<span class="sc">::</span><span class="fu">corrplot</span>(cor_metadadata,</span>
<span id="cb105-25"><a href="#cb105-25" aria-hidden="true" tabindex="-1"></a>                   <span class="at">type =</span> <span class="st">"upper"</span>,</span>
<span id="cb105-26"><a href="#cb105-26" aria-hidden="true" tabindex="-1"></a>                   <span class="at">order =</span> <span class="st">"hclust"</span>,</span>
<span id="cb105-27"><a href="#cb105-27" aria-hidden="true" tabindex="-1"></a>                   <span class="at">p.mat =</span> p_mat,</span>
<span id="cb105-28"><a href="#cb105-28" aria-hidden="true" tabindex="-1"></a>                   <span class="at">sig.level =</span> <span class="fl">0.05</span>,</span>
<span id="cb105-29"><a href="#cb105-29" aria-hidden="true" tabindex="-1"></a>                   <span class="at">insig =</span> <span class="st">"blank"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-58-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can see that many explanatory variables are correlated. Lets remove some of them. We’ll keep S as a proxy of PO4, Sigma-t, NH4 and NO2. NO3 as a proxy of SiOH4. Chla as a proxy of PT.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>permanova_cor_pars <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">adonis2</span>(physeq_clr_dist <span class="sc">~</span> S <span class="sc">+</span> NO3 <span class="sc">+</span> NT <span class="sc">+</span> Chla <span class="sc">+</span> T,</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">by =</span> <span class="st">"margin"</span>,</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">data =</span> metadata,</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">perm =</span> <span class="dv">1000</span>)</span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>permanova_cor_pars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for adonis under reduced model
Marginal effects of terms
Permutation: free
Number of permutations: 1000

vegan::adonis2(formula = physeq_clr_dist ~ S + NO3 + NT + Chla + T, data = metadata, permutations = 1000, by = "margin")
         Df SumOfSqs      R2      F  Pr(&gt;F)  
S         1    449.5 0.13071 2.9676 0.02597 *
NO3       1    120.0 0.03489 0.7921 0.52847  
NT        1    145.4 0.04227 0.9597 0.41958  
Chla      1     82.7 0.02403 0.5457 0.83117  
T         1    152.8 0.04442 1.0085 0.39461  
Residual 12   1817.8 0.52854                 
Total    17   3439.3 1.00000                 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>The updated PERMANOVA model is improved over the original. We see that salinity is again significantly related to the response variable. However, the model with only salinity is even much better. The take home message is that true relationships among variables will be masked if explanatory variables are collinear. This creates problems in model creation which lead to complications in model inference. Taking the extra time to evaluate collinearity is a critical first step to creating more robust ecological models.</p>
</section>
<section id="analysis-of-similarity-anosim" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="analysis-of-similarity-anosim"><span class="header-section-number">7.2</span> ANalysis Of SIMilarity (ANOSIM)</h2>
<p><a href="https://sites.google.com/site/mb3gustame/hypothesis-tests/anosim" target="_blank">ANOSIM</a> tests for significant difference between two or more classes of objects based on any (dis)similarity measure (Clarke 1993). It compares the ranks of distances between objects of different classes with ranks of object distances within classes. The basis of this approach is similar to the NMDS ordination technique described above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co">#ANOSIM</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">anosim</span>(physeq_clr_dist, metadata<span class="sc">$</span>Geo, <span class="at">permutations =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
vegan::anosim(x = physeq_clr_dist, grouping = metadata$Geo, permutations = 1000) 
Dissimilarity: euclidean 

ANOSIM statistic R: 0.6104 
      Significance: 0.001998 

Permutation: free
Number of permutations: 1000</code></pre>
</div>
</div>
<p>Similarly to PERMANOVA, the result of ANOSIM indicates a significant effect of the Norther or Southern origin of the sample on bacterial communities.</p>
<p>A more formal approach to hypotheses testing can be done using redundancy analysis or canonical correspondence analysis that directly uses information on metadata fields when generating the ordinations and conducting testing. These approaches directly test hypotheses about environmental variables.</p>
</section>
</section>
<section id="direct-gradient-analysis" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Direct gradient analysis</h1>
<p>Simple (unconstrained) ordination analyses one data matrix and reveals its major structure in a graph constructed from a reduced set of orthogonal axes. It is therefore a passive form of analysis, and the user interprets the ordination results <em>a posteriori</em>. On the contrary, direct gradient analyses (called also canonical ordination) associates two or more data sets in the ordination process itself. Consequently, if one wishes to extract structures of a data set that are related to structures in other data sets, and/or formally test statistical hypotheses about the significance of these relationships, canonical ordination is the way to go. Here we will perform a RDA on our dataset but many other types exit: distance-based redundancy analysis (db-RDA), canonical correspondence analysis (CCA), linear discriminant analysis (LDA), canonical correlation analysis (CCorA), co-inertia analysis (CoIA) and multiple factor analysis (MFA). For more informations see this excellent <a href="http://www.ievbras.ru/ecostat/Kiril/R/Biblio_N/R_Eng/Borcard2011.pdf">book</a>.</p>
<section id="redundant-analysis-rda" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="redundant-analysis-rda"><span class="header-section-number">8.1</span> Redundant analysis (RDA)</h2>
<section id="running-the-rda" class="level3" data-number="8.1.1">
<h3 data-number="8.1.1" class="anchored" data-anchor-id="running-the-rda"><span class="header-section-number">8.1.1</span> Running the RDA</h3>
<p>RDA is a method combining regression and principal component analysis (PCA). RDA computes axes that are linear combinations of the explanatory variables. In RDA, one can truly say that the axes explain or model (in the statistical sense) the variation of the dependent matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># RDA of the Aitchinson distance</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="co"># constrained by all the environmental variables</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="co"># contained in metadata</span></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Observe the shortcut formula</span></span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a>spe_rda <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">rda</span>(<span class="fu">t</span>(physeq_clr_asv) <span class="sc">~</span> .,</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a>                      metadata[, <span class="dv">11</span><span class="sc">:</span><span class="dv">21</span>])</span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">summary</span>(spe_rda))  <span class="co"># Scaling 2 (default)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
rda(formula = t(physeq_clr_asv) ~ SiOH4 + NO2 + NO3 + NH4 + PO4 +      NT + PT + Chla + T + S + Sigma_t, data = metadata[, 11:21]) 

Partitioning of variance:
              Inertia Proportion
Total           202.3     1.0000
Constrained     148.3     0.7331
Unconstrained    54.0     0.2669

Eigenvalues, and their contribution to the variance 

Importance of components:
                         RDA1     RDA2     RDA3   RDA4   RDA5    RDA6    RDA7
Eigenvalue            74.7345 18.84629 11.90007 8.9010 7.8296 7.16839 5.44784
Proportion Explained   0.3694  0.09315  0.05882 0.0440 0.0387 0.03543 0.02693
Cumulative Proportion  0.3694  0.46256  0.52138 0.5654 0.6041 0.63951 0.66644
                        RDA8    RDA9   RDA10   RDA11     PC1     PC2     PC3
Eigenvalue            4.3302 3.95180 2.75245 2.45334 21.3087 9.45652 8.78665
Proportion Explained  0.0214 0.01953 0.01361 0.01213  0.1053 0.04674 0.04343
Cumulative Proportion 0.6878 0.70737 0.72098 0.73311  0.8384 0.88517 0.92861
                          PC4     PC5     PC6
Eigenvalue            6.36428 4.58077 3.49894
Proportion Explained  0.03146 0.02264 0.01729
Cumulative Proportion 0.96006 0.98271 1.00000

Accumulated constrained eigenvalues
Importance of components:
                         RDA1    RDA2     RDA3    RDA4    RDA5    RDA6    RDA7
Eigenvalue            74.7345 18.8463 11.90007 8.90099 7.82956 7.16839 5.44784
Proportion Explained   0.5039  0.1271  0.08023 0.06001 0.05279 0.04833 0.03673
Cumulative Proportion  0.5039  0.6310  0.71119 0.77121 0.82400 0.87233 0.90906
                        RDA8    RDA9   RDA10   RDA11
Eigenvalue            4.3302 3.95180 2.75245 2.45334
Proportion Explained  0.0292 0.02664 0.01856 0.01654
Cumulative Proportion 0.9383 0.96490 0.98346 1.00000

Scaling 2 for species and site scores
* Species are scaled proportional to eigenvalues
* Sites are unscaled: weighted dispersion equal on all dimensions
* General scaling constant of scores:  7.658034 


Species scores

                                                 RDA1     RDA2     RDA3
ASV1_Cyanobiaceae_Synechococcus CC9902        -0.0596  0.12522  0.02270
ASV2_Pseudoalteromonadaceae_Pseudoalteromonas -0.8395 -0.23273 -0.38777
ASV3_Clade I_Clade Ia                         -0.2279 -0.03634 -0.23369
ASV4_NA_NA                                    -0.7359  0.28976  0.08346
ASV5_Clade I_Clade Ia                          0.6602 -0.26225  0.42525
ASV6_Clade II_NA                              -0.2461 -0.44561 -0.26803
....                                                                   
                                                  RDA4     RDA5     RDA6
ASV1_Cyanobiaceae_Synechococcus CC9902        -0.13407  0.06996  0.01007
ASV2_Pseudoalteromonadaceae_Pseudoalteromonas  0.37976 -0.05935  0.17884
ASV3_Clade I_Clade Ia                         -0.24822  0.06073 -0.15396
ASV4_NA_NA                                     0.05702 -0.26488  0.05920
ASV5_Clade I_Clade Ia                          0.24486  0.37488 -0.64496
ASV6_Clade II_NA                              -0.47052  0.10169 -0.00963
....                                                                    


Site scores (weighted sums of species scores)

       RDA1    RDA2     RDA3    RDA4    RDA5     RDA6
S11B -1.609 -1.0001  3.67959  0.5451  0.5306 -1.53550
S1B   2.266 -0.5546  0.70614 -2.7479  1.6011 -0.40114
S2B   2.669 -2.5998  0.23205 -2.8774 -0.5574  1.06560
S2S  -1.685 -1.7292  2.25029 -0.5908 -1.8130 -1.68862
S3B   3.098 -1.7722 -0.06663 -1.7587  0.1693  0.01803
S3S   2.826  0.6370  1.23085  3.6259 -1.6220  1.72689
....                                                 


Site constraints (linear combinations of constraining variables)

       RDA1     RDA2    RDA3    RDA4    RDA5    RDA6
S11B -1.121 -0.45070  3.5093  0.5717  2.0025 -0.8636
S1B   1.518 -0.07096  0.8763 -2.9969  2.5079 -0.8466
S2B   1.711 -2.96205  0.9517 -3.0001 -1.1186  1.1216
S2S   0.508 -1.78491  1.1014 -0.0840 -2.6836 -0.9359
S3B   3.653 -1.25732 -0.9666 -1.2833  0.5731  0.1402
S3S   1.750  0.43947  2.4113  2.9230 -1.0502  1.4855
....                                                


Biplot scores for constraining variables

            RDA1     RDA2      RDA3     RDA4      RDA5     RDA6
SiOH4   -0.56344 -0.26628 -0.178900  0.18029 -0.125839 -0.50694
NO2     -0.51205 -0.10526  0.035978 -0.14724  0.173850  0.07568
NO3      0.59137  0.07856 -0.117821 -0.27492 -0.177285  0.43003
NH4     -0.63902 -0.45015  0.106009  0.24241  0.165463  0.07606
PO4     -0.48798 -0.05357 -0.216039  0.03295  0.277721  0.11559
NT       0.03505 -0.17889  0.001032 -0.53669  0.480684 -0.60184
PT      -0.61839 -0.30505  0.117144 -0.12527  0.078368 -0.51912
Chla    -0.48042 -0.08602  0.083560 -0.01575 -0.002962 -0.14361
T       -0.57615  0.24071  0.272349 -0.55212 -0.132381 -0.03980
S       -0.93639  0.02844 -0.035563 -0.02284  0.223631  0.03176
Sigma_t -0.52342 -0.20072 -0.280063  0.46375  0.374483  0.07385</code></pre>
</div>
</div>
<p>The included environmental variables explain 70.44% of the variation in bacterial community composition across sites. 29.56 % of the variance is unexplained. However, we’ll see that the propotion of variance explained is much lower. The R2 from the summary measures the strength of the canonical relationship between the response variables (Y matrix, ASVs) and the explanatory variables (X matrix) by calculating the proportion of the variation of Y explained by the variables in X. However, this R2 is biased. We calculate an Adjusted R2, which also measures the strength of the relationship between Y and X, but applies a correction of the R2 to take into account the number of explanatory variables. This is the statistic that should be reported.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Unadjusted R^2 retrieved from the rda object</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>R2 <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">RsquareAdj</span>(spe_rda)<span class="sc">$</span>r.squared</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>R2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.7331051</code></pre>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Adjusted R^2 retrieved from the rda object</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>R2adj <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">RsquareAdj</span>(spe_rda)<span class="sc">$</span>adj.r.squared</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>R2adj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2437978</code></pre>
</div>
</div>
<p>In reality, the proportion of variance explained dropped to 16.25 %. The numerical output shows that the first two canonical axes explain together 35.1% of the total variance of the data, the first axis alone explaining 25.9%. These are unadjusted values, however. Since R2 adj = 16.2 %, the percentages of accumulated constrained adj eigenvalues show that the first axis alone explains 0.162 * 0.368 = 0.059 or 5.9% variance. Because ecological data are generally quite noisy, one should never expect to obtain a very high value of R2 . Furthermore, the first unconstrained eigenvalue (PC1), the first unconstrained axe for the residuals, is comparatively high, which means that it does display an important residual structure of the response data that is not explain by the environmental parameters measure here.</p>
</section>
<section id="significance-testing" class="level3" data-number="8.1.2">
<h3 data-number="8.1.2" class="anchored" data-anchor-id="significance-testing"><span class="header-section-number">8.1.2</span> Significance testing</h3>
<p>The interpretation of the constrained ordination must be preceded by a test of statistical significance (see below). As in multiple regression, a non-significant result must not be interpreted and must be discarded.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Global test of the RDA result</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(spe_rda, <span class="at">step =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = t(physeq_clr_asv) ~ SiOH4 + NO2 + NO3 + NH4 + PO4 + NT + PT + Chla + T + S + Sigma_t, data = metadata[, 11:21])
         Df Variance      F Pr(&gt;F)
Model    11  148.315 1.4983  0.103
Residual  6   53.996              </code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tests of all canonical axes</span></span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(spe_rda, <span class="at">by =</span> <span class="st">"axis"</span>, <span class="at">step =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Forward tests for axes
Permutation: free
Number of permutations: 999

Model: rda(formula = t(physeq_clr_asv) ~ SiOH4 + NO2 + NO3 + NH4 + PO4 + NT + PT + Chla + T + S + Sigma_t, data = metadata[, 11:21])
         Df Variance      F Pr(&gt;F)
RDA1      1   74.735 8.3045  0.108
RDA2      1   18.846 2.0942  0.785
RDA3      1   11.900 1.3223  0.995
RDA4      1    8.901 0.9891  1.000
RDA5      1    7.830 0.8700  1.000
RDA6      1    7.168 0.7965  1.000
RDA7      1    5.448 0.6054  1.000
RDA8      1    4.330 0.4812  1.000
RDA9      1    3.952 0.4391  1.000
RDA10     1    2.752 0.3059  1.000
RDA11     1    2.453 0.2726  0.962
Residual  6   53.996              </code></pre>
</div>
</div>
<p>Here we can see that ur full model is statistically non significant (p = 0.08), and every canonical axis resulting from the RDA are not either statistically significant (p &gt; 0.05). This RDA model is not interpretable.</p>
<p><em>Can you tell why?</em></p>
</section>
<section id="selecting-variables" class="level3" data-number="8.1.3">
<h3 data-number="8.1.3" class="anchored" data-anchor-id="selecting-variables"><span class="header-section-number">8.1.3</span> Selecting variables</h3>
<p>It happens sometimes that one wishes to reduce the number of explanatory variables. The reasons vary: search for parsimony, rich data set but poor a priori hypotheses and possible strong linear dependencies (correlations) among the explanatory variables in the RDA model, which could render the regression coefficients of the explanatory variables in the model unstable.</p>
<p>A simple approach to identify collinearity among explanatory variables is the use of variance inflation factors (VIF). VIF calculations are straightforward and easily comprehensible; the higher the value, the higher the collinearity. VIF measure the proportion by which the variance of a regression coefficient is inflated in the presence of other explanatory variables. VIFs above 20 indicate strong collinearity. Ideally, VIFs above 10 should be at least examined, and avoided if possible.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Variance inflation factors (VIF)</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">vif.cca</span>(spe_rda)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      SiOH4         NO2         NO3         NH4         PO4          NT 
   4.066588    3.489186    3.634643   16.867288    8.819736    4.908553 
         PT        Chla           T           S     Sigma_t 
   6.835572    2.264012 5417.455601 8388.550079 6878.896122 </code></pre>
</div>
</div>
<p>Salinity, Temperature and Sigma.t have very hight VIFs wich confirm the collinearities observed earlier between explanatory variables (see the PERMANOVA section). A reduction of the number of explanatory variables is justified. In order to simplify this model, we can perform a forward selection (or backwards or stepwise). These types of selections help us select variables that are statistically important. However, it is important to note that selecting variables ecologically is much more important than performing selection in this way. If a variable of ecological interest is not selected, this does not mean it has to be removed from the RDA. Here, we will be performing forward selection on our 11 environmental variables. To do this, we can use the <em>ordiR2step()</em> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Forward selection of explanatory variables using vegan's ordiR2step()</span></span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>step_forward <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">ordiR2step</span>(vegan<span class="sc">::</span><span class="fu">rda</span>(<span class="fu">t</span>(physeq_clr_asv) <span class="sc">~</span> <span class="dv">1</span>,</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">data =</span> metadata[, <span class="dv">11</span><span class="sc">:</span><span class="dv">21</span>]),</span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">scope =</span> <span class="fu">formula</span>(spe_rda),</span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">direction =</span> <span class="st">"forward"</span>,</span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">pstep =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Step: R2.adj= 0 
Call: t(physeq_clr_asv) ~ 1 
 
                R2.adjusted
+ S              0.28550226
&lt;All variables&gt;  0.24379778
+ NH4            0.12916980
+ PT             0.11294232
+ T              0.09661417
+ NO3            0.09312322
+ SiOH4          0.08807489
+ Sigma_t        0.07454079
+ NO2            0.05626169
+ PO4            0.05157147
+ Chla           0.04721846
&lt;none&gt;           0.00000000
+ NT            -0.02048917</code></pre>
</div>
</div>
<p>Here, we are essentially adding one variable at a time, and retaining it if it significantly increases the model’s adjusted R2. The forward selection show us that a model with only salinity has higher R2 adjust than with all variable and explain 18.4 % of the variance. Lets calculate this most parsimonious RDA and check its significance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Parsimonious RDA</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>spe_rda_pars <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">rda</span>(<span class="fu">t</span>(physeq_clr_asv) <span class="sc">~</span> S, <span class="at">data =</span> metadata[, <span class="dv">11</span><span class="sc">:</span><span class="dv">21</span>])</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(spe_rda_pars, <span class="at">step =</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = t(physeq_clr_asv) ~ S, data = metadata[, 11:21])
         Df Variance      F Pr(&gt;F)    
Model     1   66.263 7.7929  0.001 ***
Residual 16  136.048                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(spe_rda_pars, <span class="at">step =</span> <span class="dv">1000</span>, <span class="at">by =</span> <span class="st">"axis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Forward tests for axes
Permutation: free
Number of permutations: 999

Model: rda(formula = t(physeq_clr_asv) ~ S, data = metadata[, 11:21])
         Df Variance      F Pr(&gt;F)    
RDA1      1   66.263 7.7929  0.001 ***
Residual 16  136.048                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>R2adj_pars <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">RsquareAdj</span>(spe_rda_pars)<span class="sc">$</span>adj.r.squared</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare variance inflation factors</span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">vif.cca</span>(spe_rda)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      SiOH4         NO2         NO3         NH4         PO4          NT 
   4.066588    3.489186    3.634643   16.867288    8.819736    4.908553 
         PT        Chla           T           S     Sigma_t 
   6.835572    2.264012 5417.455601 8388.550079 6878.896122 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">vif.cca</span>(spe_rda_pars)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>S 
1 </code></pre>
</div>
</div>
<p>Now, both the model and the first canonical axis resulting from the RDA are statistically significant (p &lt; 0.05). The VIF of salinity is only 1. This RDA model is interpretable. Lets plot it.</p>
</section>
<section id="rda-plot" class="level3" data-number="8.1.4">
<h3 data-number="8.1.4" class="anchored" data-anchor-id="rda-plot"><span class="header-section-number">8.1.4</span> RDA plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Preparation of the data for the plot</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="co"># View analysis results</span></span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>ii <span class="ot">&lt;-</span> <span class="fu">summary</span>(spe_rda_pars)</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Depending on the drawing result</span></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a><span class="co"># the drawing data can be enlarged or</span></span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a><span class="co"># reduced to a certain extent, as follows</span></span>
<span id="cb132-9"><a href="#cb132-9" aria-hidden="true" tabindex="-1"></a>sp <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(ii<span class="sc">$</span>species[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">*</span> <span class="dv">2</span></span>
<span id="cb132-10"><a href="#cb132-10" aria-hidden="true" tabindex="-1"></a>sp_top <span class="ot">&lt;-</span> sp[<span class="fu">order</span>(<span class="fu">abs</span>(sp<span class="sc">$</span>RDA1), <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ][<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, ]</span>
<span id="cb132-11"><a href="#cb132-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-12"><a href="#cb132-12" aria-hidden="true" tabindex="-1"></a>st <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(ii<span class="sc">$</span>sites[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])</span>
<span id="cb132-13"><a href="#cb132-13" aria-hidden="true" tabindex="-1"></a>st <span class="ot">&lt;-</span> <span class="fu">merge</span>(st,</span>
<span id="cb132-14"><a href="#cb132-14" aria-hidden="true" tabindex="-1"></a>      metadata[<span class="st">"Geo"</span>],</span>
<span id="cb132-15"><a href="#cb132-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">by =</span> <span class="st">"row.names"</span>)</span>
<span id="cb132-16"><a href="#cb132-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-17"><a href="#cb132-17" aria-hidden="true" tabindex="-1"></a>yz <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">as.data.frame</span>(ii<span class="sc">$</span>biplot[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]))</span>
<span id="cb132-18"><a href="#cb132-18" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(yz) <span class="ot">&lt;-</span> <span class="st">"Salinity"</span></span>
<span id="cb132-19"><a href="#cb132-19" aria-hidden="true" tabindex="-1"></a>yz <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(yz)</span>
<span id="cb132-20"><a href="#cb132-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-21"><a href="#cb132-21" aria-hidden="true" tabindex="-1"></a>eigen_values <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="dv">100</span> <span class="sc">*</span>ii<span class="sc">$</span>cont[[<span class="dv">1</span>]][<span class="dv">2</span>,], <span class="at">digits=</span><span class="dv">4</span>)</span>
<span id="cb132-22"><a href="#cb132-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb132-23"><a href="#cb132-23" aria-hidden="true" tabindex="-1"></a><span class="co">#plot</span></span>
<span id="cb132-24"><a href="#cb132-24" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb132-25"><a href="#cb132-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> st, <span class="at">size =</span> <span class="dv">4</span>,</span>
<span id="cb132-26"><a href="#cb132-26" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(<span class="at">x =</span> RDA1, <span class="at">y =</span> PC1,</span>
<span id="cb132-27"><a href="#cb132-27" aria-hidden="true" tabindex="-1"></a>                 <span class="at">shape =</span> Geo, <span class="at">fill =</span> Geo)) <span class="sc">+</span></span>
<span id="cb132-28"><a href="#cb132-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="dv">21</span><span class="sc">:</span><span class="dv">25</span>)) <span class="sc">+</span></span>
<span id="cb132-29"><a href="#cb132-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data =</span> sp_top,</span>
<span id="cb132-30"><a href="#cb132-30" aria-hidden="true" tabindex="-1"></a>               <span class="at">arrow =</span> <span class="fu">arrow</span>(<span class="at">angle =</span> <span class="fl">22.5</span>,</span>
<span id="cb132-31"><a href="#cb132-31" aria-hidden="true" tabindex="-1"></a>                             <span class="at">length =</span> <span class="fu">unit</span>(<span class="fl">0.35</span>, <span class="st">"cm"</span>),</span>
<span id="cb132-32"><a href="#cb132-32" aria-hidden="true" tabindex="-1"></a>                             <span class="at">type =</span> <span class="st">"closed"</span>),</span>
<span id="cb132-33"><a href="#cb132-33" aria-hidden="true" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="fl">0.6</span>, <span class="at">colour =</span> <span class="st">"red"</span>,</span>
<span id="cb132-34"><a href="#cb132-34" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">xend =</span> RDA1, <span class="at">yend =</span> PC1)) <span class="sc">+</span></span>
<span id="cb132-35"><a href="#cb132-35" aria-hidden="true" tabindex="-1"></a>  ggrepel<span class="sc">::</span><span class="fu">geom_text_repel</span>(<span class="at">data =</span> sp_top,</span>
<span id="cb132-36"><a href="#cb132-36" aria-hidden="true" tabindex="-1"></a>                           <span class="fu">aes</span>(<span class="at">x =</span> RDA1, <span class="at">y =</span> PC1, <span class="at">label =</span> <span class="fu">row.names</span>(sp_top))) <span class="sc">+</span></span>
<span id="cb132-37"><a href="#cb132-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data =</span> yz,</span>
<span id="cb132-38"><a href="#cb132-38" aria-hidden="true" tabindex="-1"></a>               <span class="at">arrow =</span> <span class="fu">arrow</span>(<span class="at">angle =</span> <span class="fl">22.5</span>,</span>
<span id="cb132-39"><a href="#cb132-39" aria-hidden="true" tabindex="-1"></a>                             <span class="at">length =</span> <span class="fu">unit</span>(<span class="fl">0.35</span>,<span class="st">"cm"</span>),</span>
<span id="cb132-40"><a href="#cb132-40" aria-hidden="true" tabindex="-1"></a>                             <span class="at">type =</span> <span class="st">"closed"</span>),</span>
<span id="cb132-41"><a href="#cb132-41" aria-hidden="true" tabindex="-1"></a>               <span class="at">linetype =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="fl">0.6</span>, <span class="at">colour =</span> <span class="st">"blue"</span>,</span>
<span id="cb132-42"><a href="#cb132-42" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">xend =</span> RDA1, <span class="at">yend =</span> PC1)) <span class="sc">+</span></span>
<span id="cb132-43"><a href="#cb132-43" aria-hidden="true" tabindex="-1"></a>  ggrepel<span class="sc">::</span><span class="fu">geom_text_repel</span>(<span class="at">data =</span> yz, <span class="fu">aes</span>(RDA1, PC1, <span class="at">label=</span><span class="fu">row.names</span>(yz)))<span class="sc">+</span></span>
<span id="cb132-44"><a href="#cb132-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste</span>(<span class="st">"RDA 1 ("</span>, eigen_values[<span class="dv">1</span>], <span class="st">"%)"</span>, <span class="at">sep =</span> <span class="st">""</span>),</span>
<span id="cb132-45"><a href="#cb132-45" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="fu">paste</span>(<span class="st">"PC 1 ("</span>, eigen_values[<span class="dv">2</span>], <span class="st">"%)"</span>, <span class="at">sep =</span> <span class="st">""</span>))<span class="sc">+</span></span>
<span id="cb132-46"><a href="#cb132-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>,<span class="at">linetype =</span> <span class="dv">3</span>,<span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span> </span>
<span id="cb132-47"><a href="#cb132-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>,<span class="at">linetype =</span> <span class="dv">3</span>,<span class="at">size =</span> <span class="dv">1</span>)<span class="sc">+</span></span>
<span id="cb132-48"><a href="#cb132-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">shape =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="cn">NULL</span>,</span>
<span id="cb132-49"><a href="#cb132-49" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb132-50"><a href="#cb132-50" aria-hidden="true" tabindex="-1"></a>         <span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="cn">NULL</span>))<span class="sc">+</span></span>
<span id="cb132-51"><a href="#cb132-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb132-52"><a href="#cb132-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>One of the most powerful aspects of RDA is the simultaneous visualization of your response and explanatory variables (i.e.&nbsp;species and environmental variables). From this ordination, we can really say now that salinity is the main environmental driver measured shaping bacterial communities. Among all the ASVs, some are more related to this gradient of salinity. This is the case of ASV 12 and 11 for which abundance increase when salinity decreases and ASV 7 which presents the opposite pattern. These differential abundance patterns can be explored with many kind of analyses (see next chapter) but what is really powerful with RDA is that you highlight gradient relationships not a difference of abundance between two conditions. However, a large part of the variance in the bacterial community remains unexplained. Variance in species communities can be explained by deterministic processes such as species sorting (influence of the environment as we’ve seen here) but also by stochastic processes such as dispersal which depend, among other things, of the distance between communities. Since we have this information, lets take a look at a very common pattern in community ecology: the distance-decay pattern.</p>
</section>
</section>
<section id="multiple-regression-on-dissimilarity-matrices-mrm" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="multiple-regression-on-dissimilarity-matrices-mrm"><span class="header-section-number">8.2</span> Multiple Regression on dissimilarity Matrices (MRM)</h2>
<p>The decay of assemblage similarity with spatial distance can be explained by alternative mechanisms: dispersal limitation and species sorting. To understand their relative contributions, we compare the decay in bacterial similarity with spatial distance and, independently, with environmental distance. You will find excellent articles on distance-decay relationship <a href="https://onlinelibrary.wiley.com/doi/epdf/10.1111/ecog.03693" target="_blank">here</a> and <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2021.702016/full" target="_blank">here</a>.</p>
<p>A combination of Mantel correlation and multiple regression, multiple regression on distance matrices (MRM; Manly, 1986; Smouse et al., 1986; Legendre et al., 1994) allows a regression-type analysis of two or more (dis)similarity matrices, using permutations to determine the significance of the coefficients of determination. One matrix must contain (dis)similarities calculated from response data, such as OTU abundances, and the other matrices must contain (dis)similarities calculated from explanatory data (e.g.&nbsp;environmental parameters or space).</p>
<p>First we calculate the spatial distance matrix. In order to calculate kilometric distance bewtween sampling points from geographic coordinates, we used the <strong>SpatialEpi</strong> package and the <em>latlong2grid()</em> function. Here, you will load the result of this function because there is a conflict with this package and the betapart package we use after.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co">#library(SpatialEpi)</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="co">#ANFcoord &lt;- read.table("Location_coordinates.txt", sep = "\t", row.names = 1, header = T)</span></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="co">#ANF_km &lt;- latlong2grid(ANFcoord[,1:2])</span></span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a><span class="co">#rownames(ANF_km) &lt;- rownames(ANFcoord)</span></span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>ANF_km <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,<span class="st">"beta_diversity"</span>,<span class="st">"spatial_distance.rds"</span>))</span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a>ANF_km_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(ANF_km)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then, The relationship between microbial pairwise similarity and spatial distance is assessed by fitting negative exponential function describing the decay in microbial similarity with spatial distance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Calculate and add model to the plot</span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>ANF_decay_exp <span class="ot">&lt;-</span> betapart<span class="sc">::</span><span class="fu">decay.model</span>(physeq_clr_dist<span class="sc">/</span><span class="dv">100</span>,</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>                                       ANF_km_dist,</span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">y.type=</span><span class="st">"dissim"</span>,</span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">model.type=</span><span class="st">"exp"</span>,</span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">perm=</span><span class="dv">100</span>)</span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-9"><a href="#cb135-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot Distance decay relationships</span></span>
<span id="cb135-10"><a href="#cb135-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ANF_km_dist, physeq_clr_dist<span class="sc">/</span><span class="dv">100</span>,</span>
<span id="cb135-11"><a href="#cb135-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(physeq_clr_dist<span class="sc">/</span><span class="dv">100</span>)),</span>
<span id="cb135-12"><a href="#cb135-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(ANF_km_dist)),</span>
<span id="cb135-13"><a href="#cb135-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Distance (km)"</span>, <span class="at">ylab =</span> <span class="st">"Dissimilarity (CLR)"</span>)</span>
<span id="cb135-14"><a href="#cb135-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-15"><a href="#cb135-15" aria-hidden="true" tabindex="-1"></a>betapart<span class="sc">::</span><span class="fu">plot.decay</span>(ANF_decay_exp, <span class="at">col =</span> <span class="st">"blue"</span>,</span>
<span id="cb135-16"><a href="#cb135-16" aria-hidden="true" tabindex="-1"></a>                     <span class="at">remove.dots =</span> <span class="cn">TRUE</span>, <span class="at">add =</span> <span class="cn">TRUE</span>)</span>
<span id="cb135-17"><a href="#cb135-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-18"><a href="#cb135-18" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>,</span>
<span id="cb135-19"><a href="#cb135-19" aria-hidden="true" tabindex="-1"></a>       <span class="fu">paste</span>(<span class="st">"exp: (Beta ="</span>, <span class="fu">round</span>(ANF_decay_exp<span class="sc">$</span>second.parameter, <span class="dv">4</span>),</span>
<span id="cb135-20"><a href="#cb135-20" aria-hidden="true" tabindex="-1"></a>             <span class="st">", Rsqr ="</span>, <span class="fu">round</span>(ANF_decay_exp<span class="sc">$</span>pseudo.r.squared, <span class="dv">2</span>),</span>
<span id="cb135-21"><a href="#cb135-21" aria-hidden="true" tabindex="-1"></a>             <span class="st">", p ="</span>, <span class="fu">round</span>(ANF_decay_exp<span class="sc">$</span>p.value, <span class="dv">2</span>)),</span>
<span id="cb135-22"><a href="#cb135-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-69-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The negative exponential model significantly explained the decay in similarity with spatial distance (p &lt; 0.01). But what is the contribution of dispersal and species sorting in this pattern? Lets find out by decomposing the variance between the spatial and the envirnomental matrices.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Variance partitioning</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Microbiam matrix (response)</span></span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>physeq_clr_dist_square <span class="ot">&lt;-</span> phyloseq<span class="sc">::</span><span class="fu">distance</span>(physeq_clr,</span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">method =</span> <span class="st">"euclidean"</span>,</span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">diag =</span> <span class="cn">TRUE</span>,</span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a>                                             <span class="at">upper =</span> <span class="cn">TRUE</span>)</span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Spatial matrix (explicative)</span></span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a>ANF_km_dist_square <span class="ot">&lt;-</span> <span class="fu">dist</span>(ANF_km, <span class="at">diag =</span> <span class="cn">TRUE</span>, <span class="at">upper =</span> <span class="cn">TRUE</span>)</span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a><span class="co">#environmental matrix (explicative)</span></span>
<span id="cb136-12"><a href="#cb136-12" aria-hidden="true" tabindex="-1"></a>envdata <span class="ot">&lt;-</span> <span class="fu">dist</span>(metadata[,<span class="dv">11</span><span class="sc">:</span><span class="dv">21</span>], <span class="at">diag =</span> <span class="cn">TRUE</span>, <span class="at">upper =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Multiple regressions on Matrices (MRM) - attention les colonnes et lignes des matrices doivent correspondrent (pas besoin d'avoir les mêmes noms)</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>ecodist<span class="sc">::</span><span class="fu">MRM</span>(physeq_clr_dist_square <span class="sc">~</span> envdata <span class="sc">+</span> ANF_km_dist_square, <span class="at">nperm=</span><span class="dv">1000</span>) <span class="co"># 0.366</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$coef
                   physeq_clr_dist_square  pval
Int                           13.52659906 1.000
envdata                        1.27215830 0.001
ANF_km_dist_square             0.02024728 0.001

$r.squared
       R2      pval 
0.3401107 0.0010000 

$F.test
       F   F.pval 
38.65542  0.00100 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>ecodist<span class="sc">::</span><span class="fu">MRM</span>(physeq_clr_dist_square <span class="sc">~</span> envdata, <span class="at">nperm=</span><span class="dv">1000</span>) <span class="co"># 0.212</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$coef
        physeq_clr_dist_square  pval
Int                  15.288591 0.999
envdata               1.630614 0.002

$r.squared
       R2      pval 
0.1819084 0.0020000 

$F.test
       F   F.pval 
33.57591  0.00200 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>ecodist<span class="sc">::</span><span class="fu">MRM</span>(physeq_clr_dist_square <span class="sc">~</span> ANF_km_dist_square, <span class="at">nperm=</span><span class="dv">1000</span>) <span class="co"># 0.238</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$coef
                   physeq_clr_dist_square  pval
Int                           16.38701869 0.941
ANF_km_dist_square             0.02402992 0.003

$r.squared
       R2      pval 
0.2352174 0.0030000 

$F.test
       F   F.pval 
46.44173  0.00300 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>modEvA<span class="sc">::</span><span class="fu">varPart</span>(<span class="at">A =</span> <span class="fl">0.212</span>, <span class="at">B =</span> <span class="fl">0.238</span>, <span class="at">AB =</span> <span class="fl">0.366</span>,</span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">A.name =</span> <span class="st">"Environmental"</span>,</span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">B.name =</span> <span class="st">"Dispersal limitation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-71-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                                   Proportion
Environmental                           0.128
Dispersal limitation                    0.154
Environmental_Dispersal limitation      0.084
Unexplained                             0.634</code></pre>
</div>
</div>
<p>Using multiple regression on distance matrices (MRM), spatial and environmental variables proved significant predictors of beta diversity and together explained 36.7 % of variation in dissimilarity of microbial communities. Variance partitioning was subsequently used to partition the variation into purely spatial, purely environmental and spatially-structured environmental components. With 15,4%, the amount of variation in dissimilarity explained by the purely spatial component was higher than the variation explained by the environmental component, indicating that dispersal is an important process shaping our communities.</p>
<p>Similarly to indirect gradient analyses, many different types of direct gradient analysis are available outthere. In the following graph, we offer suggestions of some of the appropriate choices based on data input structure and expected relationships among variables.</p>
<p><img src="../img/Direct_gradient_analysis_decision.png" class="img-fluid"></p>
</section>
</section>
<section id="diffential-abundance-analysis-daa" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Diffential abundance analysis (DAA)</h1>
<p>The goal of differential abundance testing is to identify specific taxa associated with metadata variables of interest. This is a difficult task. It is also one of the more controversial areas in microbiome data analysis as illustrated in this <a href="https://arxiv.org/abs/2104.07266" target="_blank">preprint</a>. This is related to concerns that normalization and testing approaches have generally failed to control false discovery rates. For more details see these papers <a href="https://www.biorxiv.org/content/10.1101/559831v1" target="_blank">here</a> and <a href="https://www.nature.com/articles/s41467-019-10656-5" target="_blank">here</a>.</p>
<p>There are many tools to perform DAA. The most popular tools, without going into evaluating whether or not they perform well for this task, are:<br>
-<a href="https://bioconductor.org/packages/release/bicoc/html/ALDEx2.html" target="_blank">ALDEx2</a><br>
-<a href="https://bioconductor.org/packages/release/cbioc/html/ANCOMBC.html" target="_blank">ANCOM-BC</a><br>
-<a href="https://cran.r-project.org/web/packages/corcncob/index.html" target="_blank">conrcob</a><br>
-<a href="https://bioconductor.org/packages/release/bicoc/html/DESeq2.html" target="_blank">DESeq2</a><br>
-<a href="https://bioconductor.org/packages/release/biocc/html/edgeR.html" target="_blank">edgeR</a><br>
-<a href="https://bioconductor.org/packages/release/biocc/html/lefser.html" target="_blank">LEFse</a><br>
-<a href="https://bioconductor.org/packages/releasce/bioc/html/limma.html" target="_blank">limma voom</a><br>
-<a href="https://genomebiology.biomedcentral.com/articcles/10.1186/s13059-022-02655-5" target="_blank">LinDA</a><br>
-<a href="https://www.bioconductor.org/packages/relecase/bioc/html/Maaslin2.html" target="_blank">MaAsLin2</a><br>
-<a href="https://www.bioconductor.org/packagesc/release/bioc/html/metagenomeSeq.html" target="_blank">metagenomeSeq</a><br>
-<a href="https://www.rdocumentation.org/packages/labdcsv/versions/2.0-1/topics/indval" target="_blank">IndVal</a><br>
-<a href="https://www.rdocumentation.org/packages/statcs/versions/3.6.2/topics/t.test" target="_blank">t-test</a><br>
-<a href="https://www.rdocumentation.org/packagces/stats/versions/3.6.2/topics/wilcox.test" target="_blank">Wilcoxon test</a><br>
</p>
<p>Nearing et al.&nbsp;(<a href="https://www.nature.com/articles/s41467-022-28034-z" target="_blank">2022</a>) compared all these listed methods across 38 different datasets and showed that ALDEx2 and ANCOM-BC produce the most consistent results across studies. Because different methods use different approaches (parametric vs non-parametric, different normalization techniques, assumptions etc.), results can differ between methods.Therefore, it is highly recommended to pick several methods to get an idea about how robust and potentially reproducible your findings are depending on the method. Here, we will apply 3 methods that are currently used in microbial ecology or that can be recommended based on recent literature (ANCOM-BC, ALDEx2 and LEFse) and we will compare the results between them. For this, we will use the recent <strong>microbiome_marker</strong> package.</p>
<section id="linear-discriminant-analysis-effect-size-lefse" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="linear-discriminant-analysis-effect-size-lefse"><span class="header-section-number">9.1</span> Linear discriminant analysis Effect Size (LEFse)</h2>
<p>LEFSE has been developped by Segata et al.&nbsp;(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3218848/" target="_blank">2011</a>). LEFse first use the non-parametric factorial Kruskal-Wallis (KW) sum-rank test to detect features with significant differential abundance with respect to the class of interest; biological consistency is subsequently investigated using a set of pairwise tests among subclasses using the (unpaired) Wilcoxon rank-sum test. As a last step, LEfSe uses LDA to estimate the effect size of each differentially abundant features.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co">#LEFSE</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>mm_lefse <span class="ot">&lt;-</span> microbiomeMarker<span class="sc">::</span><span class="fu">run_lefse</span>(physeq, <span class="at">norm =</span> <span class="st">"CPM"</span>,</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">wilcoxon_cutoff =</span> <span class="fl">0.01</span>,</span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">group =</span> <span class="st">"Geo"</span>,</span>
<span id="cb145-5"><a href="#cb145-5" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">taxa_rank =</span> <span class="st">"none"</span>,</span>
<span id="cb145-6"><a href="#cb145-6" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">kw_cutoff =</span> <span class="fl">0.01</span>,</span>
<span id="cb145-7"><a href="#cb145-7" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">multigrp_strat =</span> <span class="cn">TRUE</span>,</span>
<span id="cb145-8"><a href="#cb145-8" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">lda_cutoff =</span> <span class="dv">4</span>)</span>
<span id="cb145-9"><a href="#cb145-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-10"><a href="#cb145-10" aria-hidden="true" tabindex="-1"></a>mm_lefse_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(mm_lefse<span class="sc">@</span>marker_table)</span>
<span id="cb145-11"><a href="#cb145-11" aria-hidden="true" tabindex="-1"></a>mm_lefse_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         feature enrich_group   ef_lda       pvalue         padj
marker1    ASV11        North 4.746047 0.0015574784 0.0015574784
marker2    ASV12        North 4.699727 0.0045142882 0.0045142882
marker3    ASV10        North 4.661376 0.0022950748 0.0022950748
marker4    ASV18        North 4.460565 0.0045142882 0.0045142882
marker5    ASV35        North 4.183570 0.0045142882 0.0045142882
marker6    ASV49        North 4.025863 0.0045142882 0.0045142882
marker7     ASV2        South 4.950924 0.0039173223 0.0039173223
marker8     ASV8        South 4.706448 0.0020814438 0.0020814438
marker9     ASV7        South 4.670957 0.0010275895 0.0010275895
marker10    ASV3        South 4.433849 0.0091897421 0.0091897421
marker11   ASV13        South 4.406032 0.0073724319 0.0073724319
marker12   ASV27        South 4.333577 0.0008112059 0.0008112059</code></pre>
</div>
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>p_LDAsc <span class="ot">&lt;-</span> microbiomeMarker<span class="sc">::</span><span class="fu">plot_ef_bar</span>(mm_lefse)</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>y_labs <span class="ot">&lt;-</span> <span class="fu">ggplot_build</span>(p_LDAsc)<span class="sc">$</span>layout<span class="sc">$</span>panel_params[[<span class="dv">1</span>]]<span class="sc">$</span>y<span class="sc">$</span><span class="fu">get_labels</span>()</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>p_abd <span class="ot">&lt;-</span> microbiomeMarker<span class="sc">::</span><span class="fu">plot_abundance</span>(mm_lefse, <span class="at">group =</span> <span class="st">"Geo"</span>) <span class="sc">+</span></span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">limits =</span> y_labs)</span>
<span id="cb147-5"><a href="#cb147-5" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(p_LDAsc, p_abd, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>LEFse identifies 12 biomarkers and among them ASV 7, 11 and 12 that we already identifies ealier with other methods.</p>
</section>
<section id="differential-analysis-of-compositions-of-microbiomes-with-bias-correction-ancom-bc" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="differential-analysis-of-compositions-of-microbiomes-with-bias-correction-ancom-bc"><span class="header-section-number">9.2</span> Differential analysis of compositions of microbiomes with bias correction (ANCOM-BC)</h2>
<p>The ANCOM-BC <a href="https://www.nature.com/articles/s41467-020-17041-7" target="_blank">methodology</a> assumes that the observed sample is an unknown fraction of a unit volume of the ecosystem, and the sampling fraction varies from sample to sample. ANCOM-BC accounts for sampling fraction by introducing a sample-specific offset term in a linear regression framework, that is estimated from the observed data. The offset term serves as the bias correction, and the linear regression framework in log scale is analogous to log-ratio transformation to deal with the compositionality of microbiome data. Furthermore, this method provides p-values and confidence intervals for each taxon. It also controls the FDR and it is computationally simple to implement.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co">#ancomBC</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>mm_ancombc <span class="ot">&lt;-</span> <span class="fu">run_ancombc_patched</span>(</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>  physeq,</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">"Geo"</span>,</span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">taxa_rank =</span> <span class="st">"none"</span>,</span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pvalue_cutoff =</span> <span class="fl">0.001</span>,</span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">p_adjust =</span> <span class="st">"fdr"</span></span>
<span id="cb148-8"><a href="#cb148-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb148-9"><a href="#cb148-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-10"><a href="#cb148-10" aria-hidden="true" tabindex="-1"></a>mm_ancombc_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(mm_ancombc<span class="sc">@</span>marker_table)</span>
<span id="cb148-11"><a href="#cb148-11" aria-hidden="true" tabindex="-1"></a>mm_ancombc_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         feature enrich_group      ef_W       pvalue         padj
marker1     ASV2        South  3.980197 6.885820e-05 7.230111e-04
marker2     ASV7        South  4.341347 1.416118e-05 1.652137e-04
marker3     ASV8        South  4.532481 5.829496e-06 1.020162e-04
marker4    ASV10        North -4.775089 1.796277e-06 6.286968e-05
marker5    ASV11        North -5.811580 6.188594e-09 3.249012e-07
marker6    ASV12        North -4.466839 7.938375e-06 1.041912e-04
marker7    ASV18        North -4.561024 5.090471e-06 1.020162e-04
marker8    ASV27        South  5.874154 4.250091e-09 3.249012e-07
marker9    ASV35        North -4.483869 7.330158e-06 1.041912e-04
marker10   ASV49        North -4.680720 2.858686e-06 7.504051e-05</code></pre>
</div>
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>an_ef <span class="ot">&lt;-</span> microbiomeMarker<span class="sc">::</span><span class="fu">plot_ef_bar</span>(mm_ancombc)</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>y_labs <span class="ot">&lt;-</span> <span class="fu">ggplot_build</span>(an_ef)<span class="sc">$</span>layout<span class="sc">$</span>panel_params[[<span class="dv">1</span>]]<span class="sc">$</span>y<span class="sc">$</span><span class="fu">get_labels</span>()</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>an_abd <span class="ot">&lt;-</span> microbiomeMarker<span class="sc">::</span><span class="fu">plot_abundance</span>(mm_ancombc, <span class="at">group =</span> <span class="st">"Geo"</span>) <span class="sc">+</span></span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">limits =</span> y_labs)</span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(an_ef, an_abd, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="beta_diversity_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>ANCOM-BC identifies 10 biomarkers and all in common with the results of the LEFse analysis.</p>
</section>
<section id="anova-like-differential-expression-aldex2" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="anova-like-differential-expression-aldex2"><span class="header-section-number">9.3</span> ANOVA-like differential expression (ALDEx2)</h2>
<p>ALDEx2 estimates technical variation within each sample per taxon by utilizing the Dirichlet distribution. It furthermore applies the centered-log-ratio transformation (or closely related log-ratio transforms). Depending on the experimental setup, it will perform a two sample Welch’s T-test and Wilcoxon-test or a one-way ANOVA and Kruskal-Wallis-test. The Benjamini-Hochberg procedure is applied in any case to correct for multiple testing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>mm_aldex <span class="ot">&lt;-</span> microbiomeMarker<span class="sc">::</span><span class="fu">run_aldex</span>(physeq, <span class="at">group =</span> <span class="st">"Geo"</span>,</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">norm =</span> <span class="st">"CPM"</span>,</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">taxa_rank =</span> <span class="st">"none"</span>,</span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">p_adjust =</span> <span class="st">"fdr"</span>)</span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a>mm_aldex_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(mm_aldex<span class="sc">@</span>marker_table)</span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>mm_aldex_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        feature enrich_group ef_aldex       pvalue       padj
marker1   ASV27        North 2.095543 0.0003582814 0.03277281</code></pre>
</div>
</div>
<p>ALDEx2 is much more stringent and identifies only 1 biomarker, ASV 27 which has been identified by the two other DAA methods. The others do not reach the FDR cut-off used here; although, they likely have “largish” effect sizes. Often, if I consider performing DA testing, I will run several models and focus on the intersection of OTUs given by at least two methods. Here it would be the 10 ASV identified with the ANCOM-BC.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        for (let i = 0; i < 2; i++) {
          container.appendChild(note.children[i].cloneNode(true));
        }
        return container.innerHTML
      } else {
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        console.log("RESIZE");
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/ANF-MetaBioDiv\.github\.io\/course-material\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p><a href="http://abims.sb-roscoff.fr/"><img src="../img/logo_abims.png" alt="ABiMS" height="35"></a> <a href="https://www.umr-beep.fr/"><img src="../img/logo_beep.png" alt="beep" height="40"></a> <a href="https://www.france-bioinformatique.fr/"><img src="../img/logo_ifb.png" alt="IFB" height="40"></a> <a href="https://www.cnrs.fr//"><img src="../img/logo_cnrs.svg" alt="CNRS" height="40"></a> <a href="https://www.ird.fr/"><img src="../img/logo_ird.png" alt="IRD" height="40"></a> <a href="https://www.mio.osupytheas.fr/"><img src="../img/logo_mio.png" alt="MIO" height="40"></a> <a href="https://umr-marbec.fr/"><img src="../img/logo_marbec.jpeg" alt="marbec" height="40"></a> <a href="https://umr-marbec.fr/"><img src="../img/logo_tara.png" alt="marbec" height="30"></a></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>