<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.36">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2022-08-01">

<title>ANF metabarcoding - Part I- Data preprocessing : From Dada2 to Phyloseq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ANF metabarcoding</span>
  </a>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html">Home</a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html">About</a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#set-working-directory-data" id="toc-set-working-directory-data" class="nav-link active" data-scroll-target="#set-working-directory-data">Set working directory, Data</a></li>
  <li><a href="#iii--read-the-sequencing-files" id="toc-iii--read-the-sequencing-files" class="nav-link" data-scroll-target="#iii--read-the-sequencing-files"><span style="color: steelblue;"><strong>III- Read the sequencing files</strong></span></a>
  <ul class="collapse">
  <li><a href="#acreate-a-list-of-sequencing-file-paths" id="toc-acreate-a-list-of-sequencing-file-paths" class="nav-link" data-scroll-target="#acreate-a-list-of-sequencing-file-paths"><span style="color: steelblue;"><strong>a/Create a list of sequencing file paths</strong></span></a></li>
  <li><a href="#b-extract-the-names-of-the-samples" id="toc-b-extract-the-names-of-the-samples" class="nav-link" data-scroll-target="#b-extract-the-names-of-the-samples"><span style="color: steelblue;"><strong>b/ Extract the names of the samples</strong></span></a></li>
  <li><a href="#iv-sequence-quality-check" id="toc-iv-sequence-quality-check" class="nav-link" data-scroll-target="#iv-sequence-quality-check"><span style="color: steelblue;"><strong>IV-Sequence Quality Check</strong></span></a></li>
  <li><a href="#v-sequence-trimming" id="toc-v-sequence-trimming" class="nav-link" data-scroll-target="#v-sequence-trimming"><span style="color: steelblue;"><strong>V-Sequence Trimming</strong></span></a>
  <ul class="collapse">
  <li><a href="#a-primer-removal" id="toc-a-primer-removal" class="nav-link" data-scroll-target="#a-primer-removal"><span style="color: steelblue;"><strong>a/ Primer Removal</strong></span></a></li>
  <li><a href="#b-quality-trimming-process" id="toc-b-quality-trimming-process" class="nav-link" data-scroll-target="#b-quality-trimming-process"><span style="color: steelblue;">b/ Quality Trimming process</span></a></li>
  </ul></li>
  <li><a href="#vi-learn-errors" id="toc-vi-learn-errors" class="nav-link" data-scroll-target="#vi-learn-errors"><span style="color: steelblue;"><strong>VI-Learn « Errors »</strong></span></a>
  <ul class="collapse">
  <li><a href="#apply-the-function" id="toc-apply-the-function" class="nav-link" data-scroll-target="#apply-the-function"><strong>Apply the function</strong></a></li>
  </ul></li>
  <li><a href="#vii--dereplication-identical-sequences" id="toc-vii--dereplication-identical-sequences" class="nav-link" data-scroll-target="#vii--dereplication-identical-sequences"><span style="color: steelblue;"><strong>VII- Dereplication (identical sequences)</strong></span></a></li>
  <li><a href="#viii-amplicon-sequence-variant-asv" id="toc-viii-amplicon-sequence-variant-asv" class="nav-link" data-scroll-target="#viii-amplicon-sequence-variant-asv"><span style="color: steelblue;"><strong>VIII-Amplicon Sequence Variant =ASV</strong></span></a></li>
  <li><a href="#ix--assembly" id="toc-ix--assembly" class="nav-link" data-scroll-target="#ix--assembly"><span style="color: steelblue;"><strong>IX- Assembly</strong></span></a></li>
  <li><a href="#x--remove-chimera" id="toc-x--remove-chimera" class="nav-link" data-scroll-target="#x--remove-chimera"><span style="color: steelblue;"><strong>X- Remove Chimera</strong></span></a></li>
  <li><a href="#xi-summarize-the-pre-processing" id="toc-xi-summarize-the-pre-processing" class="nav-link" data-scroll-target="#xi-summarize-the-pre-processing"><span style="color: steelblue;"><strong>XI – Summarize the pre-processing</strong></span></a></li>
  <li><a href="#xii--taxonomic-assignment-using-dada2" id="toc-xii--taxonomic-assignment-using-dada2" class="nav-link" data-scroll-target="#xii--taxonomic-assignment-using-dada2"><span style="color: steelblue;"><strong>XII- Taxonomic Assignment using Dada2</strong></span></a></li>
  <li><a href="#export-asv-table" id="toc-export-asv-table" class="nav-link" data-scroll-target="#export-asv-table">Export ASV table</a>
  <ul class="collapse">
  <li><a href="#r-objects" id="toc-r-objects" class="nav-link" data-scroll-target="#r-objects">R objects</a></li>
  <li><a href="#text-files" id="toc-text-files" class="nav-link" data-scroll-target="#text-files">Text files</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Part I- Data preprocessing : From Dada2 to Phyloseq</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Armougom/Garel </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 1, 2022</p>
    </div>
  </div>
    
  </div>
  

</header>

<section id="set-working-directory-data" class="level1">
<h1>Set working directory, Data</h1>
<p><strong>Your sequencing data is in the directory data so:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Put the path directory containing your data (sequencing files) in a « variable » called path_to_fastqs</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>path_to_fastqs <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,<span class="st">"raw"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#download reference databases</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>refdb_folder <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,<span class="st">"refdb"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(refdb_folder)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(refdb_folder): '/home/nicolas/Documents/anf-metabarcoding/
data/refdb' existe déjà</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getOption</span>(<span class="st">'timeout'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 60</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout=</span><span class="dv">1200</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>silva_train_set <span class="ot">&lt;-</span> <span class="fu">file.path</span>(refdb_folder,<span class="st">"silva_nr99_v138.1_train_set.fa.gz"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(refdb_folder,<span class="st">"silva_nr99_v138.1_train_set.fa.gz"</span>))){</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(<span class="st">"https://zenodo.org/record/4587955/files/silva_nr99_v138.1_train_set.fa.gz"</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>              silva_train_set,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">quiet=</span><span class="cn">TRUE</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>silva_species_assignment <span class="ot">&lt;-</span> <span class="fu">file.path</span>(refdb_folder,<span class="st">"silva_species_assignment_v138.1.fa.gz"</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(refdb_folder,<span class="st">"silva_species_assignment_v138.1.fa.gz"</span>))){</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(<span class="st">"https://zenodo.org/record/4587955/files/silva_species_assignment_v138.1.fa.gz"</span>,</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>              silva_species_assignment,</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>              <span class="at">quiet=</span><span class="cn">TRUE</span>)</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout=</span><span class="dv">60</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Attached functions from functions_dada4.R located in the R folder</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"R"</span>,<span class="st">"functions_dada4.R"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="iii--read-the-sequencing-files" class="level1">
<h1><span style="color: steelblue;"><strong>III- Read the sequencing files</strong></span></h1>
<section id="acreate-a-list-of-sequencing-file-paths" class="level3">
<h3 class="anchored" data-anchor-id="acreate-a-list-of-sequencing-file-paths"><span style="color: steelblue;"><strong>a/Create a list of sequencing file paths</strong></span></h3>
<p>The « forward » &amp; « reverse » files are in gzipped (compressed) fastq format with the label:</p>
<p><span style="color: red;">SAMPLENAME_R1.fastq.gz</span> for the forward files and <span style="color: red;">SAMPLENAME_R2.fastq.gz</span> for the reverses files.</p>
<p>R1 is Forward, R2 is Reverse</p>
<p>Function : list.files</p>
<p>https://stat.ethz.ch/R-manual/R-devel/library/base/html/list.files.html</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>fnFs <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(path_to_fastqs, <span class="at">pattern=</span><span class="st">"_R1.fastq.gz"</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>))[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>fnRs <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(path_to_fastqs, <span class="at">pattern=</span><span class="st">"_R2.fastq.gz"</span>, <span class="at">full.names=</span><span class="cn">TRUE</span>))[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>To understand: What fnFs &amp; fnRs variables contain?</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>fnFs <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/nicolas/Documents/anf-metabarcoding/data/raw/S11B_R1.fastq.gz"
[2] "/home/nicolas/Documents/anf-metabarcoding/data/raw/S1B_R1.fastq.gz" </code></pre>
</div>
</div>
</section>
<section id="b-extract-the-names-of-the-samples" class="level3">
<h3 class="anchored" data-anchor-id="b-extract-the-names-of-the-samples"><span style="color: steelblue;"><strong>b/ Extract the names of the samples</strong></span></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sample.names <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="fu">basename</span>(fnFs), <span class="st">"_"</span>), <span class="st">`</span><span class="at">[</span><span class="st">`</span>, <span class="dv">1</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># with regular expressions</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># sample.names &lt;- gsub("^.+/|_.+$","",fnFs)</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>sample.names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "S11B" "S1B" </code></pre>
</div>
</div>
<p><strong>To understand:</strong></p>
<p><strong>strsplit</strong> : split character chain according a defined pattern</p>
<p>https://www.rdocumentation.org/packages/base/versions/3.6.0/topics/strsplit</p>
<p><strong>basename</strong>: simplify the « PATH »</p>
<p>https://www.rdocumentation.org/packages/base/versions/3.6.0/topics/basename</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">strsplit</span>(<span class="fu">basename</span>(fnFs), <span class="st">"_"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "S11B"        "R1.fastq.gz"

[[2]]
[1] "S1B"         "R1.fastq.gz"</code></pre>
</div>
</div>
<p>“strsplit” function?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="fu">basename</span>(fnFs), <span class="st">"_"</span>), <span class="st">`</span><span class="at">[</span><span class="st">`</span>, <span class="dv">2</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "R1.fastq.gz" "R1.fastq.gz"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="fu">basename</span>(fnFs), <span class="st">"_"</span>), <span class="st">`</span><span class="at">[</span><span class="st">`</span>, <span class="dv">3</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA NA</code></pre>
</div>
</div>
</section>
<section id="iv-sequence-quality-check" class="level2">
<h2 class="anchored" data-anchor-id="iv-sequence-quality-check"><span style="color: steelblue;"><strong>IV-Sequence Quality Check</strong></span></h2>
<p>Function: plotQualityProfile :</p>
<p>https://www.rdocumentation.org/packages/dada2/versions/1.0.3/topics/plotQualityProfile</p>
<p>We use a function implemented in functions_dada4.R, which takes the list of R1 files and R2 files and put all quality plot results in one pdf file and save in your current directory <strong>TP_METABARCODING</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>qualityprofile</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>function (fileFw, fileRev, outfile) 
{
    pdf(outfile, onefile = TRUE)
    for (i in 1:length(fileFw)) {
        f &lt;- dada2::plotQualityProfile(fileFw[[i]])
        r &lt;- dada2::plotQualityProfile(fileRev[[i]])
        print(f)
        print(r)
    }
    dev.off()
}</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a directory for the outputs</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>quality_folder <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>,<span class="st">"dada2"</span>,<span class="st">"quality_plots"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(quality_folder,<span class="at">recursive =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(quality_folder, recursive = TRUE): '/home/nicolas/
Documents/anf-metabarcoding/outputs/dada2/quality_plots' existe déjà</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qualityprofile</span>(fnFs,</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>               fnRs,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>               <span class="fu">file.path</span>(quality_folder,<span class="st">"quality_plots.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; =
"none")` instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.
`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.
`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>png 
  2 </code></pre>
</div>
</div>
<p><strong>Open the qualityplot.pdf</strong></p>
</section>
<section id="v-sequence-trimming" class="level2">
<h2 class="anchored" data-anchor-id="v-sequence-trimming"><span style="color: steelblue;"><strong>V-Sequence Trimming</strong></span></h2>
<section id="a-primer-removal" class="level3">
<h3 class="anchored" data-anchor-id="a-primer-removal"><span style="color: steelblue;"><strong>a/ Primer Removal</strong></span></h3>
<section id="put-primer-sequences-in-variables" class="level4">
<h4 class="anchored" data-anchor-id="put-primer-sequences-in-variables"><strong>Put primer sequences in variables</strong></h4>
</section>
<section id="create-a-sub-directory-nop-in-.data-directory-with-file-names" class="level4">
<h4 class="anchored" data-anchor-id="create-a-sub-directory-nop-in-.data-directory-with-file-names"><strong>Create a sub-directory NOP in ./data directory &amp; with file names</strong></h4>
<p>Prepare empty files that are going to collect the results (e.g.&nbsp;sequences with no primers)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">#files are empty at this stage</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>path_to_trimmed_reads <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>,<span class="st">"dada2"</span>,<span class="st">"trimmed"</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(path_to_trimmed_reads)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(path_to_trimmed_reads): '/home/nicolas/Documents/anf-
metabarcoding/outputs/dada2/trimmed' existe déjà</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>nopFw <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_to_trimmed_reads, <span class="fu">basename</span>(fnFs))</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>nopRv <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_to_trimmed_reads, <span class="fu">basename</span>(fnRs))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co">#see</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>nopFw <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/nicolas/Documents/anf-metabarcoding/outputs/dada2/trimmed/S11B_R1.fastq.gz"
[2] "/home/nicolas/Documents/anf-metabarcoding/outputs/dada2/trimmed/S1B_R1.fastq.gz" </code></pre>
</div>
</div>
</section>
<section id="remove-primers" class="level4">
<h4 class="anchored" data-anchor-id="remove-primers"><strong>Remove primers</strong></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>FWD  <span class="ot">&lt;-</span> <span class="st">"CCTACGGGNBGCASCAG"</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>REV  <span class="ot">&lt;-</span> <span class="st">"GACTACNVGGGTATCTAAT"</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Enlever les Primers  dans fichiers R1=fnFs et R2=fnRs et remplir les nouveaux fichiers sans les primers : nopFs (R1) et nopRs (R2)</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>dada2<span class="sc">::</span><span class="fu">removePrimers</span>(fnFs, nopFw, <span class="at">primer.fwd=</span>FWD, <span class="at">orient=</span><span class="cn">TRUE</span>, <span class="at">max.mismatch=</span><span class="dv">0</span>,<span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Overwriting file:/home/nicolas/Documents/anf-metabarcoding/outputs/dada2/trimmed/S11B_R1.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Read in 2000, output 1959 (98%) filtered sequences.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Overwriting file:/home/nicolas/Documents/anf-metabarcoding/outputs/dada2/trimmed/S1B_R1.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Read in 2000, output 1949 (97.4%) filtered sequences.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>dada2<span class="sc">::</span><span class="fu">removePrimers</span>(fnRs, nopRv, <span class="at">primer.fwd=</span>REV, <span class="at">orient=</span><span class="cn">TRUE</span>, <span class="at">max.mismatch=</span><span class="dv">0</span>,<span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Overwriting file:/home/nicolas/Documents/anf-metabarcoding/outputs/dada2/trimmed/S11B_R2.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Read in 2000, output 1442 (72.1%) filtered sequences.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Overwriting file:/home/nicolas/Documents/anf-metabarcoding/outputs/dada2/trimmed/S1B_R2.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Read in 2000, output 1487 (74.3%) filtered sequences.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="do">#####</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="b-quality-trimming-process" class="level3">
<h3 class="anchored" data-anchor-id="b-quality-trimming-process"><span style="color: steelblue;">b/ Quality Trimming process</span></h3>
<section id="create-a-sub-directory-filtered-files-for-collecting-the-filtered-results" class="level4">
<h4 class="anchored" data-anchor-id="create-a-sub-directory-filtered-files-for-collecting-the-filtered-results"><strong>Create a sub-directory Filtered &amp; files for collecting the filtered results</strong></h4>
<p>Same as previously see for primers, but now for sequence quality trimming</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>path_to_filtered_reads <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>,<span class="st">"dada2"</span>,<span class="st">"filtered"</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(path_to_filtered_reads)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(path_to_filtered_reads): '/home/nicolas/Documents/anf-
metabarcoding/outputs/dada2/filtered' existe déjà</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>filtFs <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_to_filtered_reads, <span class="fu">basename</span>(fnFs))</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>filtRs <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_to_filtered_reads, <span class="fu">basename</span>(fnRs))</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(filtFs) <span class="ot">&lt;-</span> sample.names</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(filtRs) <span class="ot">&lt;-</span> sample.names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="apply-quality-trimming-function" class="level4">
<h4 class="anchored" data-anchor-id="apply-quality-trimming-function"><strong>Apply quality trimming function</strong></h4>
<p><strong>Function filterAndTrim</strong>: https://rdrr.io/bioc/dada2/man/filterAndTrim.html</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">filterAndTrim</span>(nopFw, filtFs, nopRv, filtRs, <span class="at">minLen=</span><span class="dv">150</span>, <span class="at">matchIDs=</span><span class="cn">TRUE</span>,</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">maxN=</span><span class="dv">0</span>, <span class="at">maxEE=</span><span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">3</span>), <span class="at">truncQ=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co">#See</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 reads.in reads.out
S11B_R1.fastq.gz     1959       950
S1B_R1.fastq.gz      1949      1016</code></pre>
</div>
</div>
<p><strong>Details:</strong></p>
<p><strong>nopFw</strong> : Input, your row Forward list (path) of sequencing files without primer.</p>
<p><strong>filtFs</strong> : Output, the empty files created at the previous stage that you are now going to fill with this command.</p>
<p><strong>nopRv, filRs</strong> : Same as above, but for the Reverse data</p>
<p><strong>TruncLen</strong> : Truncate reads after truncLen bases. Reads shorter than this are discarded.</p>
<p>Exple : TruncLen=c(200,150), means R1 forward are cut at 200 pb &amp; the R2 reverse at 150 pb.</p>
<p><strong>TrimLeft</strong> : The number of nucleotides to remove from the start of each read, from the left</p>
<p><strong>Trimright</strong> : The number of nucleotides to remove from the start of each read, from the right</p>
<p><strong>maxN</strong> : Max number of ambiguous bases accepted</p>
<p><strong>maxEE</strong> : Quality system to remove low quality read. Standard maxEE=c(2,2).</p>
<p>If you want to be more flexible (accept more low quality), increase the value, maxEE=c(2,5). 2 is for the forward, 5 for le reverse.</p>
<p><strong>TruncQ=2</strong>. Truncate reads at the first instance of a quality score less than or equal to truncQ.</p>
</section>
</section>
</section>
<section id="vi-learn-errors" class="level2">
<h2 class="anchored" data-anchor-id="vi-learn-errors"><span style="color: steelblue;"><strong>VI-Learn « Errors »</strong></span></h2>
<p><strong>Function: learnErrors</strong> https://rdrr.io/bioc/dada2/man/learnErrors.html</p>
<p>Parametric model to distinguish sequencing artefacts from true biological variations</p>
<section id="apply-the-function" class="level3">
<h3 class="anchored" data-anchor-id="apply-the-function"><strong>Apply the function</strong></h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>errF <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">learnErrors</span>(filtFs, <span class="at">randomize =</span><span class="cn">TRUE</span>, <span class="at">multithread=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>541515 total bases in 1966 reads from 2 samples will be used for learning the error rates.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>errR <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">learnErrors</span>(filtRs, <span class="at">randomize =</span><span class="cn">TRUE</span>, <span class="at">multithread=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>554402 total bases in 1966 reads from 2 samples will be used for learning the error rates.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">#See plot</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>dada2<span class="sc">::</span><span class="fu">plotErrors</span>(errF, <span class="at">nominalQ=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis
Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output-display">
<p><img src="PART_I-DADA2_ANF_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="vii--dereplication-identical-sequences" class="level2">
<h2 class="anchored" data-anchor-id="vii--dereplication-identical-sequences"><span style="color: steelblue;"><strong>VII- Dereplication (identical sequences)</strong></span></h2>
<p><strong>Function: derepFastq</strong></p>
<p>https://www.rdocumentation.org/packages/dada2/versions/1.0.3/topics/derepFastq</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>derepFs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">derepFastq</span>(filtFs, <span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Dereplicating sequence entries in Fastq file: /home/nicolas/Documents/anf-metabarcoding/outputs/dada2/filtered/S11B_R1.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Encountered 607 unique sequences from 950 total sequences read.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Dereplicating sequence entries in Fastq file: /home/nicolas/Documents/anf-metabarcoding/outputs/dada2/filtered/S1B_R1.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Encountered 664 unique sequences from 1016 total sequences read.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>derepRs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">derepFastq</span>(filtRs, <span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Dereplicating sequence entries in Fastq file: /home/nicolas/Documents/anf-metabarcoding/outputs/dada2/filtered/S11B_R2.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Encountered 691 unique sequences from 950 total sequences read.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Dereplicating sequence entries in Fastq file: /home/nicolas/Documents/anf-metabarcoding/outputs/dada2/filtered/S1B_R2.fastq.gz</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Encountered 733 unique sequences from 1016 total sequences read.</code></pre>
</div>
</div>
</section>
<section id="viii-amplicon-sequence-variant-asv" class="level2">
<h2 class="anchored" data-anchor-id="viii-amplicon-sequence-variant-asv"><span style="color: steelblue;"><strong>VIII-Amplicon Sequence Variant =ASV</strong></span></h2>
<p><span style="color: steelblue;"><strong>Error Corrections: function dada</strong></span></p>
<p>https://www.quantargo.com/help/r/latest/packages/dada2/1.18.0/dada</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>dadaFs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">dada</span>(derepFs, <span class="at">err=</span>errF, <span class="at">multithread=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample 1 - 950 reads in 607 unique sequences.
Sample 2 - 1016 reads in 664 unique sequences.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>dadaRs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">dada</span>(derepRs, <span class="at">err=</span>errR, <span class="at">multithread=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample 1 - 950 reads in 691 unique sequences.
Sample 2 - 1016 reads in 733 unique sequences.</code></pre>
</div>
</div>
</section>
<section id="ix--assembly" class="level2">
<h2 class="anchored" data-anchor-id="ix--assembly"><span style="color: steelblue;"><strong>IX- Assembly</strong></span></h2>
<p><strong>Function: mergePairs :</strong></p>
<p>https://www.rdocumentation.org/packages/dada2/versions/1.0.3/topics/mergePairs</p>
<p><em>Merge the forwards &amp; reverses together</em> maxMismatch : how many mismatch do you accept in the overlapping region?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>mergers <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">mergePairs</span>(dadaFs, derepFs, dadaRs, derepRs, <span class="at">maxMismatch=</span> <span class="dv">0</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>709 paired-reads (in 28 unique pairings) successfully merged out of 768 (in 45 pairings) input.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>677 paired-reads (in 32 unique pairings) successfully merged out of 758 (in 57 pairings) input.</code></pre>
</div>
</div>
<section id="make-otu-table-dada2-format-have-sequences-as-identifiers" class="level4">
<h4 class="anchored" data-anchor-id="make-otu-table-dada2-format-have-sequences-as-identifiers"><strong>Make OTU table: dada2 format have sequences as identifiers!</strong></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>seqtab <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">makeSequenceTable</span>(mergers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="x--remove-chimera" class="level2">
<h2 class="anchored" data-anchor-id="x--remove-chimera"><span style="color: steelblue;"><strong>X- Remove Chimera</strong></span></h2>
<p><strong>Function: removeBimeraDenovo</strong></p>
<p>https://www.rdocumentation.org/packages/dada2/versions/1.0.3/topics/removeBimeraDenovo</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>seqtab.nochim <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">removeBimeraDenovo</span>(seqtab, <span class="at">method=</span><span class="st">"consensus"</span>, <span class="at">multithread=</span><span class="cn">TRUE</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Identified 1 bimeras out of 53 input sequences.</code></pre>
</div>
</div>
</section>
<section id="xi-summarize-the-pre-processing" class="level2">
<h2 class="anchored" data-anchor-id="xi-summarize-the-pre-processing"><span style="color: steelblue;"><strong>XI – Summarize the pre-processing</strong></span></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>getN <span class="ot">&lt;-</span> <span class="cf">function</span>(x) <span class="fu">sum</span>(dada2<span class="sc">::</span><span class="fu">getUniques</span>(x))</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>track <span class="ot">&lt;-</span> base<span class="sc">::</span><span class="fu">cbind</span>(out, <span class="fu">sapply</span>(dadaFs, getN), <span class="fu">sapply</span>(dadaRs, getN),<span class="fu">sapply</span>(mergers, getN),<span class="fu">rowSums</span>(seqtab.nochim),<span class="fu">rowSums</span>(seqtab.nochim)<span class="sc">/</span>out[,<span class="dv">1</span>]<span class="sc">*</span><span class="dv">100</span>)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(track) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"input"</span>, <span class="st">"filtered"</span>, <span class="st">"denoisedF"</span>, <span class="st">"denoisedR"</span>, <span class="st">"merged"</span>, <span class="st">"nonchim"</span>,<span class="st">"%retained"</span>)</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(track) <span class="ot">&lt;-</span> sample.names</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>track</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     input filtered denoisedF denoisedR merged nonchim %retained
S11B  1959      950       830       809    709     630  32.15926
S1B   1949     1016       794       859    677     662  33.96614</code></pre>
</div>
</div>
</section>
<section id="xii--taxonomic-assignment-using-dada2" class="level2">
<h2 class="anchored" data-anchor-id="xii--taxonomic-assignment-using-dada2"><span style="color: steelblue;"><strong>XII- Taxonomic Assignment using Dada2</strong></span></h2>
<p><span style="color: steelblue;"><strong>Function “assignTaxonomy”</strong></span>: <strong>Assignment from Phylum to Genus level</strong></p>
<p>https://www.quantargo.com/help/r/latest/packages/dada2/1.18.0/assignTaxonomy</p>
<p><span style="color: steelblue;"><strong>Function “addSpecies”</strong></span>: <strong>Assignment to the species level if possible (100% identity).</strong></p>
<p>https://www.quantargo.com/help/r/latest/packages/dada2/1.18.0/addSpecies</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">assignTaxonomy</span>(seqtab.nochim,</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>                              silva_train_set,</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">taxLevels =</span> <span class="fu">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>,<span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>,<span class="st">"Genus"</span>, <span class="st">"Species"</span>),</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">multithread =</span> <span class="cn">TRUE</span>,</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">minBoot =</span> <span class="dv">60</span>)</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>taxa <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">addSpecies</span>(taxa,</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>                          silva_species_assignment,</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">allowMultiple =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="export-asv-table" class="level2">
<h2 class="anchored" data-anchor-id="export-asv-table">Export ASV table</h2>
<section id="r-objects" class="level3">
<h3 class="anchored" data-anchor-id="r-objects">R objects</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>export_folder <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>,<span class="st">"dada2"</span>,<span class="st">"asv_table"</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(export_folder)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in dir.create(export_folder): '/home/nicolas/Documents/anf-
metabarcoding/outputs/dada2/asv_table' existe déjà</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="at">object =</span> seqtab.nochim,</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="fu">file.path</span>(export_folder,<span class="st">"seqtab_nochim.rds"</span>))</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="at">object =</span> taxa,</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="fu">file.path</span>(export_folder,<span class="st">"taxa.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="text-files" class="level3">
<h3 class="anchored" data-anchor-id="text-files">Text files</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>taxo_export <span class="ot">&lt;-</span> <span class="fu">apply</span>(taxa,<span class="dv">1</span>,paste,<span class="at">collapse=</span><span class="st">";"</span>)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>seqtab_export <span class="ot">&lt;-</span> <span class="fu">t</span>(seqtab.nochim)</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>seqtab_export <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> taxo_export,</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> seqtab_export,</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">by=</span><span class="dv">0</span>,</span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">all.x=</span><span class="cn">TRUE</span>,</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">sort=</span><span class="cn">FALSE</span>)</span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-10"><a href="#cb78-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(seqtab_export)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"sequence"</span>,<span class="st">"taxonomy"</span>)</span>
<span id="cb78-11"><a href="#cb78-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-12"><a href="#cb78-12" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(seqtab_export,</span>
<span id="cb78-13"><a href="#cb78-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">file=</span><span class="fu">file.path</span>(export_folder,<span class="st">"asv_table.tsv"</span>),</span>
<span id="cb78-14"><a href="#cb78-14" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote=</span><span class="cn">FALSE</span>,</span>
<span id="cb78-15"><a href="#cb78-15" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb78-16"><a href="#cb78-16" aria-hidden="true" tabindex="-1"></a>            <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>