---
title: "Part I- Data preprocessing : From Dada2 to Phyloseq"
date: "August 2022"
output: html_document
author:
  - "Fabrice Armougom, MIO"
  - "Marc Garel, MIO"
editor_options: 
  chunk_output_type: console
---

```{r}
library(phyloseq)
```

```{r}
asv_table <- readRDS(here::here("outputs","dada2","asv_table","seqtab_nochim.rds"))
taxonomy <- readRDS(here::here("outputs","dada2","asv_table","taxa.rds"))
context <- read.table(here::here("data","context","mapfileFA.txt"),
                      header = TRUE,
                      row.names = 1)

context
```

```{r}
asv_seq <- colnames(asv_table)
asv_id <- sapply(asv_seq,
                 function(x) digest::digest(x,algo="sha1")) |> unname()
row.names(taxonomy) <- colnames(asv_table) <- names(asv_seq) <- asv_id
```

## [**XIV- Get a phyloseq object**]{style="color: steelblue;"}

### **Mapfile**

Must contain sample names identically as you defined it at the beginning of the TP

```{r}
#Check it, if necessary: 
sample.names
```

### **Create the phyloseq object (OTU_table, Tax_table, tree)**

```{r}
physeq <- phyloseq(otu_table(asv_table, taxa_are_rows=FALSE),
                 tax_table(taxonomy),
                 sample_data(context))
```

## [**XV- Add the ASV sequences to your Phyloseq object & change ASV ID**]{style="color: steelblue;"}
 
```{r}
nucl <- Biostrings::DNAStringSet(asv_seq)
physeq <- merge_phyloseq(physeq, nucl)
```

### **See the final object for the alpha and beta analyses**

```{r}
#See Sequence object attached to Final1
Final1
```

```{r}
#See the change of ASV names 
taxa_names(Final1)[1:2]

```


## [**XIII- Phylogenetic Tree**]{style="color: steelblue;"}

NB: This step can be skipped if you have some problems to perform it or to many ASVs.

### **Get the ASV sequences**

```{r}
seqs <- dada2::getSequences(seqtab.nochim)
```

### **Alignment by DECIPHER**

```{r}
aln <- DECIPHER::AlignSeqs(refseq(physeq), anchor=NA) 
#See alignment
DECIPHER::BrowseSeqs(aln, highlight=0)
```

### **Transformed by Phydat for distance matrix/tree**

```{r}
phang.align <- phangorn::phyDat(as(aln, "matrix"), type="DNA")
```

-   

    #### **Build distance matrix**

```{r}
#Distance matrix
dm <- phangorn::dist.ml(phang.align)
#NJ tree 
treeNJ <- phangorn::NJ(dm)
#root
treeNJ<- phangorn::midpoint(tree = treeNJ)

physeq <- merge_phyloseq(physeq, treeNJ)
```

```{r}
plot_bar(physeq, fill = "Class")

plot_tree(physeq, color = "groupe", label.tips = "Family", ladderize = "left", justify = "left" , size = "Abundance")
```



