<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Fabrice Armougom, MIO">
<meta name="author" content="Marc Garel, MIO">
<meta name="author" content="Nicolas Henry, ABiMS">
<meta name="author" content="Charlotte Berthelier, ABiMS">
<meta name="dcterms.date" content="2025-09-01">

<title>ANF MetaBioDiv - Preprocessing : dada2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">ANF MetaBioDiv</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">    
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/rforbeginer.html" rel="" target="">
 <span class="dropdown-text">R for beginners</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/introduction.pdf" rel="" target="">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/dada2.html" rel="" target="_blank">
 <span class="dropdown-text">Analysing metabarcoding data with dada2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/phyloseq.html" rel="" target="_blank">
 <span class="dropdown-text">Introduction to phyloseq</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/preprocessing_dada2.html" rel="" target="">
 <span class="dropdown-text">Preprocessing: dada2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals/preprocessing_phyloseq.html" rel="" target="">
 <span class="dropdown-text">Preprocessing: phyloseq</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/alpha_diversity.pdf" rel="" target="_blank">
 <span class="dropdown-text">Alpha diversity</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/stat_mean_comparison.pdf" rel="" target="">
 <span class="dropdown-text">Introduction to univariate statistics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../courses/ampliseq_presention_ANF.html" rel="" target="_blank">
 <span class="dropdown-text">Introduction to Ampliseq</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/alpha_diversity.html" rel="" target="">
 <span class="dropdown-text">Alpha diversity</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../practicals/ampliseq.html" rel="" target="">
 <span class="dropdown-text">Workflow Ampliseq</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-4" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Day 4</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-4">    
        <li class="dropdown-header">Courses</li>
        <li>
    <a class="dropdown-item" href="../courses/beta_diversity.pdf" rel="" target="_blank">
 <span class="dropdown-text">Beta diversity</span></a>
  </li>  
        <li class="dropdown-header">Practicals</li>
        <li>
    <a class="dropdown-item" href="../practicals/beta_diversity.html" rel="" target="">
 <span class="dropdown-text">Beta diversity</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/ANF-MetaBioDiv/course-material" rel="" target="_blank"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#prepare-your-working-directory" id="toc-prepare-your-working-directory" class="nav-link active" data-scroll-target="#prepare-your-working-directory"><span class="header-section-number">1</span> Prepare your working directory</a>
  <ul class="collapse">
  <li><a href="#get-the-repository-files" id="toc-get-the-repository-files" class="nav-link" data-scroll-target="#get-the-repository-files"><span class="header-section-number">1.1</span> Get the repository files</a></li>
  <li><a href="#dowload-the-reference-database" id="toc-dowload-the-reference-database" class="nav-link" data-scroll-target="#dowload-the-reference-database"><span class="header-section-number">1.2</span> Dowload the reference database</a></li>
  <li><a href="#attach-custom-functions" id="toc-attach-custom-functions" class="nav-link" data-scroll-target="#attach-custom-functions"><span class="header-section-number">1.3</span> Attach custom functions</a></li>
  </ul></li>
  <li><a href="#inputs-files" id="toc-inputs-files" class="nav-link" data-scroll-target="#inputs-files"><span class="header-section-number">2</span> Inputs files</a>
  <ul class="collapse">
  <li><a href="#locate-the-sequencing-files" id="toc-locate-the-sequencing-files" class="nav-link" data-scroll-target="#locate-the-sequencing-files"><span class="header-section-number">2.1</span> Locate the sequencing files</a></li>
  <li><a href="#extract-sample-names" id="toc-extract-sample-names" class="nav-link" data-scroll-target="#extract-sample-names"><span class="header-section-number">2.2</span> Extract sample names</a></li>
  </ul></li>
  <li><a href="#sequence-quality-check" id="toc-sequence-quality-check" class="nav-link" data-scroll-target="#sequence-quality-check"><span class="header-section-number">3</span> Sequence quality check</a></li>
  <li><a href="#primer-removal" id="toc-primer-removal" class="nav-link" data-scroll-target="#primer-removal"><span class="header-section-number">4</span> Primer removal</a>
  <ul class="collapse">
  <li><a href="#prepare-outputs" id="toc-prepare-outputs" class="nav-link" data-scroll-target="#prepare-outputs"><span class="header-section-number">4.1</span> Prepare outputs</a></li>
  <li><a href="#remove-primers" id="toc-remove-primers" class="nav-link" data-scroll-target="#remove-primers"><span class="header-section-number">4.2</span> Remove primers</a></li>
  <li><a href="#for-more-complex-situations" id="toc-for-more-complex-situations" class="nav-link" data-scroll-target="#for-more-complex-situations"><span class="header-section-number">4.3</span> For more complex situations</a></li>
  </ul></li>
  <li><a href="#trimming-and-quality-filtering" id="toc-trimming-and-quality-filtering" class="nav-link" data-scroll-target="#trimming-and-quality-filtering"><span class="header-section-number">5</span> Trimming and quality filtering</a>
  <ul class="collapse">
  <li><a href="#prepare-outputs-1" id="toc-prepare-outputs-1" class="nav-link" data-scroll-target="#prepare-outputs-1"><span class="header-section-number">5.0.1</span> Prepare outputs</a></li>
  <li><a href="#use-dada2filterandtrim" id="toc-use-dada2filterandtrim" class="nav-link" data-scroll-target="#use-dada2filterandtrim"><span class="header-section-number">5.0.2</span> Use <code>dada2::filterAndTrim()</code></a></li>
  </ul></li>
  <li><a href="#denoising" id="toc-denoising" class="nav-link" data-scroll-target="#denoising"><span class="header-section-number">6</span> Denoising</a>
  <ul class="collapse">
  <li><a href="#learn-the-error-model" id="toc-learn-the-error-model" class="nav-link" data-scroll-target="#learn-the-error-model"><span class="header-section-number">6.1</span> Learn the error model</a></li>
  <li><a href="#dereplication" id="toc-dereplication" class="nav-link" data-scroll-target="#dereplication"><span class="header-section-number">6.2</span> Dereplication</a></li>
  <li><a href="#run-dada" id="toc-run-dada" class="nav-link" data-scroll-target="#run-dada"><span class="header-section-number">6.3</span> Run dada</a></li>
  </ul></li>
  <li><a href="#merge-paired-end-reads" id="toc-merge-paired-end-reads" class="nav-link" data-scroll-target="#merge-paired-end-reads"><span class="header-section-number">7</span> Merge paired-end reads</a></li>
  <li><a href="#build-the-asv-table" id="toc-build-the-asv-table" class="nav-link" data-scroll-target="#build-the-asv-table"><span class="header-section-number">8</span> Build the ASV table</a></li>
  <li><a href="#remove-chimeras" id="toc-remove-chimeras" class="nav-link" data-scroll-target="#remove-chimeras"><span class="header-section-number">9</span> Remove chimeras</a></li>
  <li><a href="#taxonomic-assignment-from-dada2" id="toc-taxonomic-assignment-from-dada2" class="nav-link" data-scroll-target="#taxonomic-assignment-from-dada2"><span class="header-section-number">10</span> Taxonomic assignment from dada2</a></li>
  <li><a href="#export" id="toc-export" class="nav-link" data-scroll-target="#export"><span class="header-section-number">11</span> Export</a>
  <ul class="collapse">
  <li><a href="#r-objects" id="toc-r-objects" class="nav-link" data-scroll-target="#r-objects"><span class="header-section-number">11.1</span> R objects</a></li>
  <li><a href="#text-files" id="toc-text-files" class="nav-link" data-scroll-target="#text-files"><span class="header-section-number">11.2</span> Text files</a></li>
  <li><a href="#log" id="toc-log" class="nav-link" data-scroll-target="#log"><span class="header-section-number">11.3</span> Log</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ANF-MetaBioDiv/course-material/edit/main/practicals/preprocessing_dada2.Rmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/ANF-MetaBioDiv/course-material/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Preprocessing : dada2</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Fabrice Armougom, MIO </p>
             <p>Marc Garel, MIO </p>
             <p>Nicolas Henry, ABiMS </p>
             <p>Charlotte Berthelier, ABiMS </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 1, 2025</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="prepare-your-working-directory" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Prepare your working directory</h1>
<section id="get-the-repository-files" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="get-the-repository-files"><span class="header-section-number">1.1</span> Get the repository files</h2>
<p>To begin with, download the course repository on your computer or virtual machine.</p>
<p>To do so, visit the (<a href="https://github.com/ANF-MetaBioDiv/course-material" target="_blank">ANF-metabarcoding</a>) repository on github and download it as a zip file.</p>
<p>You can also download the repository directly from the terminal using this commmand:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://github.com/ANF-MetaBioDiv/course-material/archive/refs/heads/main.zip</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once on your machine, unzip the file and place the resulting folder in the most convenient location (<code>~/Documents/</code> for example).</p>
</section>
<section id="dowload-the-reference-database" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="dowload-the-reference-database"><span class="header-section-number">1.2</span> Dowload the reference database</h2>
<p>Save as a variable the path to the folder where you will place the references databases.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>refdb_folder <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"refdb"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>refdb_folder</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/rstudio/mydatalocal/course-material/data/refdb"</code></pre>
</div>
</div>
<p>In this course we use as much as possible the <code>::</code> operator. It allows calling a function from a given package without having to load the entire package using <code>library(package)</code>. This way it is clear where the functions come from. In that case, the function <code>here()</code> comes from the package <code>here</code>.</p>
<p>The reason why we use here::here() is that when you render a Rmarkdown file, the working directory is where the Rmarkdown file is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Whereas here::here() point to the root of the R project</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>here<span class="sc">::</span><span class="fu">here</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s create the folder directly from R:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(refdb_folder)) <span class="fu">dir.create</span>(refdb_folder, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can also create the folder from RStudio in the <code>Files</code> window</p>
<p><strong>Tip:</strong> You can access the documentation of any R function using <code>?</code> in the console. If you to know everything about the function <code>dir.create()</code>, simply run <code>?dir.create()</code></p>
<p>The Silva reference database, commonly used to assign 16S metabarcoding data, will be used in practical.</p>
<p>In case you are working with 18S sequences, you will have better assignments using <a href="https://pr2-database.org/" target="_blank">PR2</a> or <a href="https://doi.org/10.5281/zenodo.6327891" target="_blank">EukRibo</a> especially if you are interested in protists.</p>
<p>The following code downloads dada2 formatted silva reference databases. If you are not comfortable with it, you can simply download the reference database files from your web browser <a href="https://zenodo.org/record/4587955/files/silva_nr99_v138.1_train_set.fa.gz%22">here</a> and <a href="&quot;https://zenodo.org/record/4587955/files/silva_species_assignment_v138.1.fa.gz" target="_blank">here</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># R stop downloading after timeout which is</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># 60 seconds by default</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">getOption</span>(<span class="st">"timeout"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 60</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># so we change timeout to be 20 minutes</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="dv">1200</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we save in variable the path to the refdb</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># in the working space</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>silva_train_set <span class="ot">&lt;-</span> <span class="fu">file.path</span>(refdb_folder,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"silva_nr99_v138.1_train_set.fa.gz"</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>silva_species_assignment <span class="ot">&lt;-</span> <span class="fu">file.path</span>(refdb_folder,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"silva_species_assignment_v138.1.fa.gz"</span>)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># then we download the files if they don't already exist</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(silva_train_set)) {</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/4587955/files/silva_nr99_v138.1_train_set.fa.gz"</span>,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    silva_train_set,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">quiet =</span> <span class="cn">TRUE</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">file.exists</span>(silva_species_assignment)) {</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">download.file</span>(</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">"https://zenodo.org/record/4587955/files/silva_species_assignment_v138.1.fa.gz"</span>,</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>    silva_species_assignment,</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>    <span class="at">quiet =</span> <span class="cn">TRUE</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="attach-custom-functions" class="level2" data-number="1.3">
<h2 data-number="1.3" class="anchored" data-anchor-id="attach-custom-functions"><span class="header-section-number">1.3</span> Attach custom functions</h2>
<p>We will use in this practical R functions especially written for this course. The “classic” way to import functions is to use <code>source()</code> with the name of the R script to source.</p>
<p>Instead, we use <code>devtools::load_all()</code>. This function will source all the scripts from the folder <code>R/</code> along with the documentation in <code>man/</code> :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">load_all</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="inputs-files" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Inputs files</h1>
<section id="locate-the-sequencing-files" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="locate-the-sequencing-files"><span class="header-section-number">2.1</span> Locate the sequencing files</h2>
<p>Save the path to the directory containing your raw data (paired-end sequencing fastq files) in a variable named <code>path_to_fastqs</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>path_to_fastqs <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"raw"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <a href="https://en.wikipedia.org/wiki/Gzip" target="_blank">gzipped</a> (compressed) <a href="https://en.wikipedia.org/wiki/FASTQ_format" target="_blank">FASTQ</a> formatted “forward” (R1) and “reverse” (R2) files are named as follow:</p>
<ul>
<li><code>${SAMPLENAME}_R1.fastq.gz</code> for the forward files</li>
<li><code>${SAMPLENAME}_R2.fastq.gz</code> for the reverse files.</li>
</ul>
<p>We list the forward files using the function <code>list.files()</code>. The argument <code>pattern</code> gives you the possibility to only select file names matching a regular expression. In our case, we select all file names finishing by <code>_R1.fastq.gz</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>fnFs <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(path_to_fastqs,</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">pattern =</span> <span class="st">"_R1.fastq.gz"</span>,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">full.names =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We do the same for reverse samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>fnRs <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(path_to_fastqs,</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">pattern =</span> <span class="st">"_R2.fastq.gz"</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">full.names =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>To understand:</strong> What fnFs &amp; fnRs variables contain?</p>
</section>
<section id="extract-sample-names" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="extract-sample-names"><span class="header-section-number">2.2</span> Extract sample names</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sample_names <span class="ot">&lt;-</span> <span class="fu">basename</span>(fnFs) <span class="sc">|&gt;</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">strsplit</span>(<span class="at">split =</span> <span class="st">"_"</span>) <span class="sc">|&gt;</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(head, <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>To understand:</strong></p>
<p><code>basename()</code>: remove path to only keep file name.</p>
<p><code>|&gt;</code>: R “pipe”. It allows you to chain functions, avoiding intermediate variables and nested parenthesis. It basically transfers the output of the left expression to the input of the right expression. You need R &gt; 4.1 to use this pipe, otherwise use <code>%&gt;%</code> from <a href="https://magrittr.tidyverse.org/" target="_blank"><code>magrittr</code></a>.</p>
<p><code>strsplit()</code>: split character chain according to a defined pattern. <code>?strsplit</code> for documentation.</p>
<p><code>sapply()</code>: apply a function to each element of a list or vector. The output is simplified to be vector.</p>
<p>Let’s go step by step. First list the R1 file names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">basename</span>(fnFs) <span class="sc">|&gt;</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "S11B_R1.fastq.gz" "S1B_R1.fastq.gz"  "S2B_R1.fastq.gz"  "S2S_R1.fastq.gz" 
[5] "S3B_R1.fastq.gz"  "S3S_R1.fastq.gz" </code></pre>
</div>
</div>
<p>We can see that the sample name is before the first <code>_</code>. With <code>strsplit()</code>, we can split each file name into a 2 elements vector. The result is a list of 2 elements vectors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">basename</span>(fnFs) <span class="sc">|&gt;</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">strsplit</span>(<span class="at">split =</span> <span class="st">"_"</span>) <span class="sc">|&gt;</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "S11B"        "R1.fastq.gz"

[[2]]
[1] "S1B"         "R1.fastq.gz"

[[3]]
[1] "S2B"         "R1.fastq.gz"

[[4]]
[1] "S2S"         "R1.fastq.gz"

[[5]]
[1] "S3B"         "R1.fastq.gz"

[[6]]
[1] "S3S"         "R1.fastq.gz"</code></pre>
</div>
</div>
<p>Now, We just have to extract the first element for each file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">basename</span>(fnFs) <span class="sc">|&gt;</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">strsplit</span>(<span class="at">split =</span> <span class="st">"_"</span>) <span class="sc">|&gt;</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(head, <span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "S11B" "S1B"  "S2B"  "S2S"  "S3B"  "S3S" </code></pre>
</div>
</div>
<p><strong>Tip:</strong> you can achieve the same thing using regular expressions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gsub</span>(<span class="st">"^.+/|_.+$"</span>, <span class="st">""</span>, fnFs) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "S11B" "S1B"  "S2B"  "S2S"  "S3B"  "S3S" </code></pre>
</div>
</div>
<p>Regular expressions are extremely useful. If you are keen to learn how to use them, have a look <a href="https://bookdown.org/rdpeng/rprogdatascience/regular-expressions.html" target="_blank">here</a></p>
</section>
</section>
<section id="sequence-quality-check" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Sequence quality check</h1>
<p>We use a custom function, <code>qualityprofile()</code>, implemented in <code>R/preprocessing.R</code> to check the quality of the raw sequences.</p>
<p>Run <code>?qualityprofile</code> to know more about this function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a directory for the outputs</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>quality_folder <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"dada2"</span>,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"quality_plots"</span>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(quality_folder)) {</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(quality_folder, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="fu">qualityprofile</span>(fnFs,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>               fnRs,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>               <span class="fu">file.path</span>(quality_folder, <span class="st">"quality_plots.pdf"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Open the the pdf file generated by <code>qualityprofile()</code></strong></p>
</section>
<section id="primer-removal" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Primer removal</h1>
<section id="prepare-outputs" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="prepare-outputs"><span class="header-section-number">4.1</span> Prepare outputs</h2>
<p>We first create a folder where to save the reads once they are trimmed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>path_to_trimmed_reads <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"outputs"</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"dada2"</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"trimmed"</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_to_trimmed_reads)) <span class="fu">dir.create</span>(path_to_trimmed_reads, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-primers" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="remove-primers"><span class="header-section-number">4.2</span> Remove primers</h2>
<p>The data you are working with correspond to the V3-V4 region using the primers Pro341F (<code>CCTACGGGNBGCASCAG</code>) and Pro805R (<code>GACTACNVGGGTATCTAAT</code>). Save into variables the forward and reverse primers:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>primer_fwd  <span class="ot">&lt;-</span> <span class="st">"CCTACGGGNBGCASCAG"</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>primer_rev  <span class="ot">&lt;-</span> <span class="st">"GACTACNVGGGTATCTAAT"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s have a closer look at sequences and find the primers.</p>
<p>Forward reads would contain <code>CCTACGGGNBGCASCAG</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">readDNAStringSet</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  fnFs[<span class="dv">1</span>],</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"fastq"</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrec =</span> <span class="dv">10</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 10:
     width seq                                              names               
 [1]   293 CCTACGGGGGGCAGCAGTAGGGA...ACATCGGCTTAACCGATGAAGT M01522:260:000000...
 [2]   293 CCTACGGGTGGCACCAGTAGGGA...CGGGGCTTAACCTCGGAACTGC M01522:260:000000...
 [3]   292 CCTACGGGGCGCAGCAGGCGCGA...GGGACCGGGAGAGGTGTGAGGT M01522:260:000000...
 [4]   293 CCTACGGGGTGCAGCAGTAGGGA...TCAAAACTCCCAGTCTAGAGTT M01522:260:000000...
 [5]   291 CCTACGGGTGGCAGCAGTGGGGA...GCAGTGGAAACTGTTGGGCTTG M01522:260:000000...
 [6]   293 CCTACGGGATGCAGCAGGCGCGA...GGGACCGGGAGAGGTGTGGGGG M01522:260:000000...
 [7]   292 CCTACGGGATGCAGCAGTGGGGA...TTTAATCCTGATGAGCTAGAAA M01522:260:000000...
 [8]   293 CCTACGGGGCGCAGCAGTAGGGA...TTAAAACTTTTGTTCTGGAATT M01522:260:000000...
 [9]   292 CCTACGGGTTGCAGCAGTGGGGA...ATTAAAACTTTTCAGCTAGAGT M01522:260:000000...
[10]   293 CCTACGGGAGGCAGCAGTGGGGA...CCCGGGCTCAACCTGGGAACGG M01522:260:000000...</code></pre>
</div>
</div>
<p>And reverse reads should contain <code>GACTACNVGGGTATCTAAT</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">readDNAStringSet</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  fnRs[<span class="dv">1</span>],</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">format =</span> <span class="st">"fastq"</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrec =</span> <span class="dv">10</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 10:
     width seq                                              names               
 [1]   301 GACTACCAGGGTATCTAATCCTG...GGCTGCTGGCACGAAGTTCGCC M01522:260:000000...
 [2]   301 GACTACCGGGGTATCTAATCCTG...GGCTGCTGGCACGGAGTTAGCC M01522:260:000000...
 [3]   300 AATCCGGTTCGTGCCCCTAGGCT...TCTTTCCCAGCCCTTATTCCAA M01522:260:000000...
 [4]   301 GACTACCGGGGTATCTAATCCTG...GGCTGCTGGCACGGAGTTAGCC M01522:260:000000...
 [5]   301 GACTACCGGGGTATCTAATCCCT...GGCTGCTGGCCCGGAATTAGCC M01522:260:000000...
 [6]   301 GGTATCTAATCCGGTTCGTGCCC...CACCGTCCTTACCCCCCCCTTT M01522:260:000000...
 [7]   301 GGTATCTAATCTTGTTTGCTCCC...CCCGACGTTAGCCGGGGCTTCT M01522:260:000000...
 [8]   301 GACTACGAGGGTATCTAATCCCG...GGCTGCTGGCACGGAATTAGCC M01522:260:000000...
 [9]   301 GGTATCTAATCCTCTTCGCTACC...CACGAAGTTAGCCGGACCTTCT M01522:260:000000...
[10]   301 GACTACGGGGGTATCTAATCCTG...GGCTGCCGGCACGGGGTTAGCC M01522:260:000000...</code></pre>
</div>
</div>
<p>We use a custom function, <code>primer_trim()</code>, implemented in <code>R/preprocessing.R</code> to remove the primers using cutadapt. To work, <code>primer_trim()</code> needs cutadapt to be installed on your computer.</p>
<p>Run <code>?primer_trim</code> to know more about this function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>(primer_log <span class="ot">&lt;-</span> <span class="fu">primer_trim</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">forward_files =</span> fnFs,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reverse_files =</span> fnRs,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">primer_fwd =</span> primer_fwd,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">primer_rev =</span> primer_rev,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">output_dir =</span> path_to_trimmed_reads,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_size =</span> <span class="dv">200</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   sample status in_reads   in_bp too_short too_long too_many_n out_reads
1    S11B     OK     2000 1186767         0        0          0      1863
2     S1B     OK     2000 1186613         1        0          0      1855
3     S2B     OK     2000 1186942         0        0          0      1839
4     S2S     OK     2000 1186868         0        0          0      1833
5     S3B     OK     2000 1186650         0        0          0      1860
6     S3S     OK     2000 1186475         1        0          0      1880
7     S4B     OK     2000 1186331         2        0          0      1867
8     S4S     OK     2000 1186681         0        0          0      1872
9     S5B     OK     2000 1186386         1        0          0      1841
10    S5S     OK     2000 1186501         1        0          0      1861
11    S6B     OK     2000 1186261         2        0          0      1839
12    S6S     OK     2000 1187078         1        0          0      1835
13    S7B     OK     2000 1186888         0        0          0      1825
14    S7S     OK     2000 1186299         3        0          0      1845
15    S8B     OK     2000 1186354         3        0          0      1840
16    S8S     OK     2000 1186610         1        0          0      1848
17    S9B     OK     2000 1187038         0        0          0      1834
18    S9S     OK     2000 1186867         0        0          0      1835
   w/adapters qualtrim_bp out_bp w/adapters2 qualtrim2_bp out2_bp
1        1986           0 513149        1876            0  528595
2        1975           0 511096        1877            0  525893
3        1987           0 506659        1850            0  521371
4        1989           0 504998        1843            0  519979
5        1989           0 512326        1870            0  527518
6        1989           0 517598        1891            0  532758
7        1980           0 514342        1884            0  529379
8        1987           0 515511        1884            0  530555
9        1984           0 506972        1856            0  522013
10       1991           0 512539        1869            0  527592
11       1981           0 506577        1857            0  521787
12       1982           0 505929        1851            0  520562
13       1987           0 503033        1836            0  517931
14       1987           0 508524        1857            0  523039
15       1993           0 507178        1847            0  522137
16       1982           0 509177        1865            0  524085
17       1983           0 505424        1851            0  520706
18       1979           0 505519        1853            0  520103</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>nopFw <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(path_to_trimmed_reads, <span class="at">pattern =</span> <span class="st">"R1"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>nopRv <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">list.files</span>(path_to_trimmed_reads, <span class="at">pattern =</span> <span class="st">"R2"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="for-more-complex-situations" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="for-more-complex-situations"><span class="header-section-number">4.3</span> For more complex situations</h2>
<p>If you have to deal with mix-orientated paired-end reads, you may find inspiration <a href="https://gitlab.sb-roscoff.fr/nhenry/abims-metabarcoding-pipeline-dada2" target="_blank">here</a></p>
</section>
</section>
<section id="trimming-and-quality-filtering" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Trimming and quality filtering</h1>
<section id="prepare-outputs-1" class="level3" data-number="5.0.1">
<h3 data-number="5.0.1" class="anchored" data-anchor-id="prepare-outputs-1"><span class="header-section-number">5.0.1</span> Prepare outputs</h3>
<p>Same as before, create a folder</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>path_to_filtered_reads <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>, <span class="st">"dada2"</span>, <span class="st">"filtered"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(path_to_filtered_reads)) <span class="fu">dir.create</span>(path_to_filtered_reads, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and list paths:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>filtFs <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_to_filtered_reads, <span class="fu">basename</span>(fnFs))</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>filtRs <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path_to_filtered_reads, <span class="fu">basename</span>(fnRs))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To make the link between files and sample names, simply name vector of file names using the sample names</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(filtFs) <span class="ot">&lt;-</span> sample_names</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(filtRs) <span class="ot">&lt;-</span> sample_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="use-dada2filterandtrim" class="level3" data-number="5.0.2">
<h3 data-number="5.0.2" class="anchored" data-anchor-id="use-dada2filterandtrim"><span class="header-section-number">5.0.2</span> Use <code>dada2::filterAndTrim()</code></h3>
<p>Let’s have a look at what the function is doing. To do so, type <code>?dada2::filterAndTrim()</code> in the console.</p>
<p>Let’s run the function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>(out <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">filterAndTrim</span>(</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">fwd =</span> nopFw,</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">filt =</span> filtFs,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">rev =</span> nopRv,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">filt.rev =</span> filtRs,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">minLen =</span> <span class="dv">150</span>,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">matchIDs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">maxN =</span> <span class="dv">0</span>,</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">maxEE =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">3</span>),</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">truncQ =</span> <span class="dv">2</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 reads.in reads.out
S11B_R1.fastq.gz     1863      1200
S1B_R1.fastq.gz      1855      1251
S2B_R1.fastq.gz      1839      1255
S2S_R1.fastq.gz      1833      1244
S3B_R1.fastq.gz      1860      1244
S3S_R1.fastq.gz      1880      1312
S4B_R1.fastq.gz      1867      1262
S4S_R1.fastq.gz      1872      1328
S5B_R1.fastq.gz      1841      1255
S5S_R1.fastq.gz      1861      1244
S6B_R1.fastq.gz      1839      1251
S6S_R1.fastq.gz      1835      1239
S7B_R1.fastq.gz      1825      1203
S7S_R1.fastq.gz      1845      1182
S8B_R1.fastq.gz      1840      1169
S8S_R1.fastq.gz      1848      1267
S9B_R1.fastq.gz      1834      1195
S9S_R1.fastq.gz      1835      1249</code></pre>
</div>
</div>
<p><strong>What happened?</strong></p>
<p>Details about the function arguments: * <strong>nopFw</strong> : input, where the forward reads without primers are (path) * <strong>filtFs</strong> : output, where forward filtered reads are written (path) * <strong>nopRv</strong> and <strong>filRs</strong> : same as above, but wiht reverse reads * <strong>TruncLen</strong> : truncate reads after truncLen bases. Reads shorter than that are discarded (TruncLen=c(200,150), means forward and reverse reads are cut at 200 bp and 150 bp respectively) * <strong>TrimLeft</strong> : number of nucleotides to remove from the start * <strong>Trimright</strong> : number of nucleotides to remove from the end * <strong>maxN</strong> : max number of ambiguous bases accepted * <strong>maxEE</strong> : read expected errors (EE) threshold. The EE of a read is the sum of the error probability of each base composing it. Increase that value to accept more low quality reads. The first value refers to the forward reads and the second to the reverse reads. * <strong>TruncQ=2</strong>: truncate reads at the first instance of a quality score less than or equal to truncQ.</p>
</section>
</section>
<section id="denoising" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Denoising</h1>
<section id="learn-the-error-model" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="learn-the-error-model"><span class="header-section-number">6.1</span> Learn the error model</h2>
<p>To be able to denoise your data, you need an error model. The error model will tell you at which rate a nucleotide is replace by another for a given quality score. For example, for a quality score Q of 30, what is the probability of an A being wrongly read as a T.</p>
<p>This error model can be learnt directly from the data with the function <code>dada2::learnErrors()</code>. You can come back to the <a href="../courses/dada2.html#/dada2-workflow-denoising-approach" target="_blank">course</a> for more details about the maths behind.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>errF <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">learnErrors</span>(filtFs,</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">randomize =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>6157072 total bases in 22350 reads from 18 samples will be used for learning the error rates.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>errR <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">learnErrors</span>(filtRs,</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">randomize =</span> <span class="cn">TRUE</span>,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>6337638 total bases in 22350 reads from 18 samples will be used for learning the error rates.</code></pre>
</div>
</div>
<p>You can visualise the resulting error model using the function <code>dada2::plotErrors()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>dada2<span class="sc">::</span><span class="fu">plotErrors</span>(errF, <span class="at">nominalQ=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Transformation introduced infinite values in continuous y-axis
Transformation introduced infinite values in continuous y-axis</code></pre>
</div>
<div class="cell-output-display">
<p><img src="preprocessing_dada2_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="dereplication" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="dereplication"><span class="header-section-number">6.2</span> Dereplication</h2>
<p>Before denoising, we need to dereplicate the sequences. It means, for each unique sequence, count the number of reads.</p>
<p>The dereplication is achieved using the function <code>dada2::derepFastq()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>derepFs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">derepFastq</span>(filtFs, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>derepRs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">derepFastq</span>(filtRs, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-dada" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="run-dada"><span class="header-section-number">6.3</span> Run dada</h2>
<p>Now we are ready to run the denoising algorithm with <code>dada2::dada()</code>. As input, we need the error model and the dereplicated sequences.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>dadaFs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">dada</span>(derepFs, <span class="at">err =</span> errF, <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample 1 - 1200 reads in 754 unique sequences.
Sample 2 - 1251 reads in 779 unique sequences.
Sample 3 - 1255 reads in 789 unique sequences.
Sample 4 - 1244 reads in 762 unique sequences.
Sample 5 - 1244 reads in 772 unique sequences.
Sample 6 - 1312 reads in 763 unique sequences.
Sample 7 - 1262 reads in 738 unique sequences.
Sample 8 - 1328 reads in 638 unique sequences.
Sample 9 - 1255 reads in 782 unique sequences.
Sample 10 - 1244 reads in 663 unique sequences.
Sample 11 - 1251 reads in 696 unique sequences.
Sample 12 - 1239 reads in 657 unique sequences.
Sample 13 - 1203 reads in 691 unique sequences.
Sample 14 - 1182 reads in 675 unique sequences.
Sample 15 - 1169 reads in 697 unique sequences.
Sample 16 - 1267 reads in 714 unique sequences.
Sample 17 - 1195 reads in 685 unique sequences.
Sample 18 - 1249 reads in 677 unique sequences.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>dadaRs <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">dada</span>(derepRs, <span class="at">err =</span> errR, <span class="at">multithread =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample 1 - 1200 reads in 928 unique sequences.
Sample 2 - 1251 reads in 948 unique sequences.
Sample 3 - 1255 reads in 968 unique sequences.
Sample 4 - 1244 reads in 925 unique sequences.
Sample 5 - 1244 reads in 948 unique sequences.
Sample 6 - 1312 reads in 967 unique sequences.
Sample 7 - 1262 reads in 953 unique sequences.
Sample 8 - 1328 reads in 904 unique sequences.
Sample 9 - 1255 reads in 975 unique sequences.
Sample 10 - 1244 reads in 887 unique sequences.
Sample 11 - 1251 reads in 914 unique sequences.
Sample 12 - 1239 reads in 846 unique sequences.
Sample 13 - 1203 reads in 881 unique sequences.
Sample 14 - 1182 reads in 874 unique sequences.
Sample 15 - 1169 reads in 879 unique sequences.
Sample 16 - 1267 reads in 967 unique sequences.
Sample 17 - 1195 reads in 892 unique sequences.
Sample 18 - 1249 reads in 911 unique sequences.</code></pre>
</div>
</div>
</section>
</section>
<section id="merge-paired-end-reads" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Merge paired-end reads</h1>
<p>Once forward and reverse reads have been denoised, we can merge them with <code>dada2::mergePairs()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>mergers <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">mergePairs</span>(</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">dadaF =</span> dadaFs,</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">derepF =</span> derepFs,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">dadaR =</span> dadaRs,</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">derepR =</span> derepRs,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">maxMismatch =</span> <span class="dv">0</span>,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">TRUE</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="build-the-asv-table" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Build the ASV table</h1>
<p>At this point we have ASVs and we know their number of reads in each sample. We have enough information to build an ASV table.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>seqtab <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">makeSequenceTable</span>(mergers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remove-chimeras" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Remove chimeras</h1>
<p>Chimeras are artifact sequences formed by two or more biological sequences incorrectly joined together. We find and remove bimeras (two-parent chimeras) using the function <code>dada2::removeBimeraDenovo()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>seqtab_nochim <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">removeBimeraDenovo</span>(seqtab,</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">method =</span> <span class="st">"consensus"</span>,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">multithread =</span> <span class="cn">TRUE</span>,</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="taxonomic-assignment-from-dada2" class="level1" data-number="10">
<h1 data-number="10"><span class="header-section-number">10</span> Taxonomic assignment from dada2</h1>
<p>The ASV table is ready. But without a clue about the ASVs taxonomic identity, we won’t go far in our ecological interpretations. We can have an idea of ASV taxonomic identity comparing their sequences to references databases such as SILVA.</p>
<p>The taxonomic assignment is done in two steps.</p>
<p>First, each ASV is assigned to a taxonomy using the RDP Naive Bayesian Classifier algorithm described in <a href="https://doi.org/10.1128/aem.00062-07" target="_blank">Wang et al.&nbsp;2007</a> called by the function <code>dada2::assignTaxonomy()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>taxonomy <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">assignTaxonomy</span>(</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">seqs =</span> seqtab_nochim,</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">refFasta =</span> silva_train_set,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">taxLevels =</span> <span class="fu">c</span>(<span class="st">"Kingdom"</span>, <span class="st">"Phylum"</span>, <span class="st">"Class"</span>,</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>,</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>                <span class="st">"Species"</span>),</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">multithread =</span> <span class="cn">TRUE</span>,</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">minBoot =</span> <span class="dv">60</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The method is robust, however, it often fails to assign at the species level.</p>
<p>If you consider that in case an ASV is 100% similar to a reference sequence, it belongs to the same species, then you can use <code>dada2::addSpecies()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>taxonomy <span class="ot">&lt;-</span> dada2<span class="sc">::</span><span class="fu">addSpecies</span>(</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  taxonomy,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  silva_species_assignment,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">allowMultiple =</span> <span class="cn">FALSE</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function assign to the species level ASVs which are identical to a reference sequence.</p>
</section>
<section id="export" class="level1" data-number="11">
<h1 data-number="11"><span class="header-section-number">11</span> Export</h1>
<p>All the preprocessing is done. Now we export our results.</p>
<section id="r-objects" class="level2" data-number="11.1">
<h2 data-number="11.1" class="anchored" data-anchor-id="r-objects"><span class="header-section-number">11.1</span> R objects</h2>
<p>The results can be exported as a R objects, one object for the ASV table and another one for the taxonomy.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>export_folder <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"outputs"</span>, <span class="st">"dada2"</span>, <span class="st">"asv_table"</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">dir.exists</span>(export_folder)) <span class="fu">dir.create</span>(export_folder, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="at">object =</span> seqtab_nochim,</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="fu">file.path</span>(export_folder, <span class="st">"seqtab_nochim.rds"</span>))</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="at">object =</span> taxonomy,</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="fu">file.path</span>(export_folder, <span class="st">"taxonomy.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="text-files" class="level2" data-number="11.2">
<h2 data-number="11.2" class="anchored" data-anchor-id="text-files"><span class="header-section-number">11.2</span> Text files</h2>
<p>We recommand to export your results as text files. They are then reusable by other programs/languages.</p>
<p>But before, we need to format the data a little bit.</p>
<p>First we create a new variable to collect the ASV sequences:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>asv_seq <span class="ot">&lt;-</span> <span class="fu">colnames</span>(seqtab_nochim)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We create unique ids for each ASV. The sequence itself is an unique id, but we would like to have something shorter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>ndigits <span class="ot">&lt;-</span> <span class="fu">nchar</span>(<span class="fu">length</span>(asv_seq))</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>asv_id <span class="ot">&lt;-</span> <span class="fu">sprintf</span>(<span class="fu">paste0</span>(<span class="st">"ASV_%0"</span>, ndigits, <span class="st">"d"</span>), <span class="fu">seq_along</span>(asv_seq))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and rename the different variables with the new ids</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(taxonomy) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(seqtab_nochim) <span class="ot">&lt;-</span> <span class="fu">names</span>(asv_seq) <span class="ot">&lt;-</span> asv_id</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before exporting the data frames (<code>taxonomy</code> and <code>seqtab_nochim</code>) as a text file, we convert their row names (ASV ids) into a new column named <code>asv</code>. This is achieved using the custom function <code>df_export()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>taxonomy_export <span class="ot">&lt;-</span> <span class="fu">df_export</span>(taxonomy, <span class="at">new_rn =</span> <span class="st">"asv"</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>seqtab_nochim_export <span class="ot">&lt;-</span> <span class="fu">t</span>(seqtab_nochim)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>seqtab_nochim_export <span class="ot">&lt;-</span> <span class="fu">df_export</span>(seqtab_nochim_export, <span class="at">new_rn =</span> <span class="st">"asv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we can export the taxonomy</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(taxonomy_export,</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="fu">file.path</span>(export_folder, <span class="st">"taxonomy.tsv"</span>),</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>the ASV table</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(seqtab_nochim_export,</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">file =</span> <span class="fu">file.path</span>(export_folder, <span class="st">"asv_table.tsv"</span>),</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and the sequences as a fasta file</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"&gt;"</span>, <span class="fu">names</span>(asv_seq), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>, asv_seq),</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep =</span> <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>,</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">file.path</span>(export_folder, <span class="st">"asv.fasta"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="log" class="level2" data-number="11.3">
<h2 data-number="11.3" class="anchored" data-anchor-id="log"><span class="header-section-number">11.3</span> Log</h2>
<p>Statistics about each preprocessing step can also be exported.</p>
<p>First this table need to be assembled:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>getN <span class="ot">&lt;-</span> <span class="cf">function</span>(x) <span class="fu">sum</span>(dada2<span class="sc">::</span><span class="fu">getUniques</span>(x))</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>log_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">input =</span> primer_log<span class="sc">$</span>in_reads,</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">with_fwd_primer =</span> primer_log<span class="sc">$</span><span class="st">`</span><span class="at">w/adapters</span><span class="st">`</span>,</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">with_rev_primer =</span> primer_log<span class="sc">$</span><span class="st">`</span><span class="at">w/adapters2</span><span class="st">`</span> ,</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">with_both_primers =</span> out[, <span class="dv">1</span>],</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">filtered =</span> out[, <span class="dv">2</span>],</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">denoisedF =</span> <span class="fu">sapply</span>(dadaFs, getN),</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">denoisedR =</span> <span class="fu">sapply</span>(dadaRs, getN),</span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">merged =</span> <span class="fu">sapply</span>(mergers, getN),</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">nonchim =</span> <span class="fu">rowSums</span>(seqtab_nochim),</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">perc_retained =</span> <span class="fu">rowSums</span>(seqtab_nochim) <span class="sc">/</span> out[, <span class="dv">1</span>] <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(log_table) <span class="ot">&lt;-</span> sample_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then it can be exported:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">df_export</span>(log_table, <span class="at">new_rn =</span> <span class="st">"sample"</span>) <span class="sc">|&gt;</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(<span class="at">file =</span> <span class="fu">file.path</span>(export_folder, <span class="st">"log_table.tsv"</span>),</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">quote =</span> <span class="cn">FALSE</span>,</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/ANF-MetaBioDiv\.github\.io\/course-material\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center"><a href="http://abims.sb-roscoff.fr/"><img src="../img/logo_abims.png" alt="ABiMS" height="35"></a> <a href="https://www.umr-beep.fr/"><img src="../img/logo_beep.png" alt="beep" height="40"></a> <a href="https://www.france-bioinformatique.fr/"><img src="../img/logo_ifb.png" alt="IFB" height="40"></a> <a href="https://www.cnrs.fr//"><img src="../img/logo_cnrs.svg" alt="CNRS" height="40"></a> <a href="https://www.ird.fr/"><img src="../img/logo_ird.png" alt="IRD" height="40"></a> <a href="https://www.mio.osupytheas.fr/"><img src="../img/logo_mio.png" alt="MIO" height="40"></a> <a href="https://umr-marbec.fr/"><img src="../img/logo_marbec.jpeg" alt="marbec" height="40"></a> <a href="https://umr-marbec.fr/"><img src="../img/logo_tara.png" alt="marbec" height="30"></a></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>