---
title: "Preprocessing : phyloseq"
date: "August 2022"
output:
  html_notebook:
  html_document:
author:
  - "Fabrice Armougom, MIO"
  - "Marc Garel, MIO"
editor_options: 
  chunk_output_type: inline
---

```{r include = FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warnings = FALSE)
```

```{r}
library(phyloseq)
```

```{r}
asv_table <- readRDS(here::here("outputs",
                                "dada2",
                                "asv_table",
                                "seqtab_nochim.rds"))

taxonomy <- readRDS(here::here("outputs",
                               "dada2",
                               "asv_table",
                               "taxonomy.rds"))

context <- read.table(here::here("data",
                                 "context",
                                 "mapfileFA.txt"),
                      header = TRUE,
                      row.names = 1)

head(context)
```

```{r}
asv_seq <- colnames(asv_table)
asv_id <- sapply(asv_seq,
                 function(x) digest::digest(x, algo = "sha1")) |> unname()
row.names(taxonomy) <- colnames(asv_table) <- names(asv_seq) <- asv_id
```

## Get a phyloseq object

### Check sample file

Make sure sample names are the same as in the ASV table

```{r}
row.names(context)
row.names(asv_table)

setdiff(row.names(asv_table), row.names(context))
```

### Assemble ASV table, taxonomy and contextual data

```{r}
physeq <- phyloseq::phyloseq(
  phyloseq::otu_table(asv_table, taxa_are_rows = FALSE),
  phyloseq::tax_table(taxonomy),
                 sample_data(context))
```

## Add the ASV sequences to your Phyloseq object & change ASV ID

```{r}
nucl <- Biostrings::DNAStringSet(asv_seq)
physeq <- merge_phyloseq(physeq, nucl)
```

## [**XIII- Phylogenetic Tree**]{style="color: steelblue;"}

NB: This step can be skipped if you have some problems to perform it or to many ASVs.

### **Alignment by DECIPHER**

```{r}
aln <- DECIPHER::AlignSeqs(refseq(physeq), anchor=NA) 
#See alignment
DECIPHER::BrowseSeqs(aln, highlight=0)
```

### **Transformed by Phydat for distance matrix/tree**

```{r}
phang.align <- phangorn::phyDat(as(aln, "matrix"), type="DNA")
```

-   

    #### **Build distance matrix**

```{r}
#Distance matrix
dm <- phangorn::dist.ml(phang.align)
#NJ tree 
treeNJ <- phangorn::NJ(dm)
#root
treeNJ<- phangorn::midpoint(tree = treeNJ)

physeq <- merge_phyloseq(physeq, treeNJ)
```

```{r}
plot_bar(physeq, fill = "Class")

plot_tree(physeq, color = "groupe", label.tips = "Family", ladderize = "left", justify = "left" , size = "Abundance")
```
