<!DOCTYPE html>
<html lang="en"><head>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.361">

  <meta name="author" content="Nicolas Henry">
  <title>ANF MetaBioDiv - Analysing metabarcoding data with dada2</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Analysing metabarcoding data with dada2</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Nicolas Henry 
</div>
</div>
</div>

</section>
<section>
<section id="introduction" class="title-slide slide level1 center">
<h1>Introduction</h1>

</section>
<section id="i-received-data" class="slide level2">
<h2>I received data</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,<span class="st">"raw"</span>) <span class="sc">|&gt;</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>    <span class="fu">list.files</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "S11B_R1.fastq.gz" "S11B_R2.fastq.gz" "S1B_R1.fastq.gz"  "S1B_R2.fastq.gz" 
 [5] "S2B_R1.fastq.gz"  "S2B_R2.fastq.gz"  "S2S_R1.fastq.gz"  "S2S_R2.fastq.gz" 
 [9] "S3B_R1.fastq.gz"  "S3B_R2.fastq.gz"  "S3S_R1.fastq.gz"  "S3S_R2.fastq.gz" 
[13] "S4B_R1.fastq.gz"  "S4B_R2.fastq.gz"  "S4S_R1.fastq.gz"  "S4S_R2.fastq.gz" 
[17] "S5B_R1.fastq.gz"  "S5B_R2.fastq.gz"  "S5S_R1.fastq.gz"  "S5S_R2.fastq.gz" 
[21] "S6B_R1.fastq.gz"  "S6B_R2.fastq.gz"  "S6S_R1.fastq.gz"  "S6S_R2.fastq.gz" 
[25] "S7B_R1.fastq.gz"  "S7B_R2.fastq.gz"  "S7S_R1.fastq.gz"  "S7S_R2.fastq.gz" 
[29] "S8B_R1.fastq.gz"  "S8B_R2.fastq.gz"  "S8S_R1.fastq.gz"  "S8S_R2.fastq.gz" 
[33] "S9B_R1.fastq.gz"  "S9B_R2.fastq.gz"  "S9S_R1.fastq.gz"  "S9S_R2.fastq.gz" </code></pre>
</div>
</div>
</section>
<section id="what-am-i-going-to-do-with-that" class="slide level2">
<h2>What am I going to do with that?</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>,<span class="st">"raw"</span>,<span class="st">"S11B_R1.fastq.gz"</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>    <span class="fu">gzfile</span>() <span class="sc">|&gt;</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>    <span class="fu">readLines</span>(<span class="at">n =</span> <span class="dv">4</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>    <span class="fu">cat</span>(<span class="at">sep=</span><span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>@M01522:260:000000000-C3YFC:1:2114:21886:9214 1:N:0:CTTGTA
CCTACGGGGGGCAGCAGTAGGGAATTTTGCGCAATGGGCGAAAGCCTGACGCAGCAACGCCGCGTGATCGATGAAGCTTCTAGGAGCGTAAAGATCTGTCGTGAGGGAAGAAAAATTCGGAGGTAAATAATTTTCGTTTTTGACGGTACCTCACAAGAAAGCACCGGCTAACTTCGTGCCAGCAGCCGCGGTAATACGAGGGGTGCTAGCGTTGTCCGGAATCATTGGGCGTAAAGGGTCCGTAGGCGGTTTAACAAGTCAATTGTTAAAGACATCGGCTTAACCGATGAAGT
+
GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGFGGGGGGGGGGGGGGFGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGFGGGGGGGGGGGGGDFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEEGGGGCGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEEGGGGGGC8CD8CGGGGGCE:FGGGGGGGFFG7FFFGFFGF:@?FFBEFFFFFFF3&gt;BFFF*</code></pre>
</div>
</div>
<p>We need a bioinformatic pipeline to:</p>
<ul>
<li>Filter out wrong sequences, such as sequencing errors</li>
<li>Get a format suitable for diversity and community composition analyses of our samples</li>
</ul>
</section></section>
<section>
<section id="fastq-records" class="title-slide slide level1 center">
<h1>FASTQ records</h1>

</section>
<section id="overview" class="slide level2">
<h2>Overview</h2>
<p>Two files (R1 and R2) per sequencing run or per sample (100 first nucleotides):</p>
<p>R1:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>@M01522:260:000000000-C3YFC:1:2114:21886:9214 1:N:0:CTTGTA
CCTACGGGGGGCAGCAGTAGGGAATTTTGCGCAATGGGCGAAAGCCTGACGCAGCAACGCCGCGTGATCGATGAAGCTTCTAGGAGCGTAAAGATCTGTC
+
GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGFGGGGGGGGGGGGGGFGGGFGGGGGGGGGGGGGGGGGGGGGGGGG</code></pre>
</div>
</div>
<p>R2:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>@M01522:260:000000000-C3YFC:1:2114:21886:9214 2:N:0:CTTGTA
GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTCCCTCAGCGTCAGTTTTAGGCCAGAAAGTTGCCTTCGCCATTGGTGTTCCTTCTGAT
+
CCCCCFGGGGGGGGGGGGGGGGGGGCEGGGGGGGGGGEEEFGG&gt;GGGGGGGGGGGGFGGGG?FGGGGGFGFGGFGGFGGCGEGGGGGGGG&gt;FGGGGGGGG</code></pre>
</div>
</div>
<p>Per record: identifier, sequence, quality</p>
</section>
<section id="identifier" class="slide level2">
<h2>Identifier</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>@M01522:260:000000000-C3YFC:1:2114:21886:9214 1:N:0:CTTGTA</code></pre>
</div>
</div>
<table>
<thead>
<tr class="header">
<th>Value</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>M05074</td>
<td>the unique instrument id</td>
</tr>
<tr class="even">
<td>31</td>
<td>the run id</td>
</tr>
<tr class="odd">
<td>000000000-BC3R3</td>
<td>the flowcell id</td>
</tr>
<tr class="even">
<td>1</td>
<td>flowcell lane</td>
</tr>
<tr class="odd">
<td>1101</td>
<td>tile number within the flowcell lane</td>
</tr>
<tr class="even">
<td>13622</td>
<td>‘x’-coordinate within the tile</td>
</tr>
<tr class="odd">
<td>1477</td>
<td>‘y’-coordinate within the tile</td>
</tr>
<tr class="even">
<td>1</td>
<td>the member of a pair</td>
</tr>
<tr class="odd">
<td>N</td>
<td>Y if the read is filtered (did not pass)</td>
</tr>
<tr class="even">
<td>0</td>
<td>0: none of the control bits are on</td>
</tr>
<tr class="odd">
<td>1</td>
<td>index sequence</td>
</tr>
</tbody>
</table>
</section>
<section id="sequence" class="slide level2">
<h2>Sequence</h2>
<p><br></p>

<img data-src="../img/read_example_1.svg" class="r-stretch"><ul>
<li><p>The tag informs you from which sample the read come from</p></li>
<li><p>The primer used for the amplification</p></li>
<li><p>The targeted sequence (metabarcode)</p></li>
</ul>
</section>
<section id="quality" class="slide level2">
<h2>Quality</h2>
<p>Quality score encoded using ASCII characters:</p>
<div class="columns">
<div class="column" style="width:50%;">
<table>
<thead>
<tr class="header">
<th style="text-align: center;">ASCII character</th>
<th style="text-align: center;">Quality score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">!</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">”</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">#</td>
<td style="text-align: center;">2</td>
</tr>
</tbody>
</table>
</div><div class="column" style="width:50%;">
<table>
<thead>
<tr class="header">
<th style="text-align: center;">ASCII character</th>
<th style="text-align: center;">Quality score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">G</td>
<td style="text-align: center;">38</td>
</tr>
<tr class="even">
<td style="text-align: center;">H</td>
<td style="text-align: center;">39</td>
</tr>
<tr class="odd">
<td style="text-align: center;">I</td>
<td style="text-align: center;">40</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The Phred quality score Q is related to the base calling error probability P according to the following formula:</p>
<p><span class="math display">\[Q = -10\log_{10}P\]</span></p>
<p><span class="math display">\[P=10^{Q/-10}\]</span></p>
<aside class="notes">
<p>In raw data reads, the quality score rarely exceed 40, but higher scores are possible when different raw reads are assembled</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section></section>
<section>
<section id="bioinformatic-pipeline" class="title-slide slide level1 center">
<h1>Bioinformatic pipeline</h1>

</section>
<section id="strategies" class="slide level2">
<h2>Strategies</h2>
<p><br></p>
<ul>
<li>Genetic clustering based approaches producing OTUs:
<ul>
<li><a href="https://doi.org/10.1038/nmeth.2604">UPARSE</a></li>
<li><a href="https://doi.org/10.1093/bioinformatics/btab493">Swarm</a> (could be seen as denoising too)</li>
</ul></li>
<li>Denoising approaches producing ASVs:
<ul>
<li><a href="https://doi.org/10.1128/mSystems.00191-16">Deblur</a></li>
<li><a href="https://doi.org/10.1038%2Fnmeth.3869">DADA2</a></li>
</ul></li>
</ul>
</section>
<section id="dada2-quest-ce-que-cest" class="slide level2">
<h2>DADA2, qu’est-ce que c’est?</h2>
<p><br></p>
<ul>
<li><p>DADA (Divisive Amplicon Denoising Algorithm) an algorithm to denoise Roche’s 454 platform errors (<a href="https://doi.org/10.1186/1471-2105-13-283" target="_blank">Rosen et al.&nbsp;2012</a>)</p></li>
<li><p>DADA2 implements a new quality-aware model of Illumina amplicon errors (<a href="https://doi.org/10.1038/nmeth.3869" target="_blank">Callahan et al.&nbsp;2016</a>)</p></li>
<li><p>DADA2 is an open-source R package <a href="https://github.com/benjjneb/dada2">https://github.com/benjjneb/dada2</a></p></li>
</ul>
</section>
<section id="dada2-the-full-amplicon-workflow" class="slide level2">
<h2>DADA2, the full amplicon workflow</h2>
<ul>
<li>Quality assessment → <code>plotQualityProfile()</code></li>
<li>Length trimming → <code>filterAndTrim()</code></li>
<li>Quality filtering → <code>filterAndTrim()</code></li>
<li>Denoising → <code>dada()</code></li>
<li>Merging pair-end reads → <code>mergePairs()</code></li>
<li>Chimera identification → <code>removeBimeraDenovo()</code></li>
<li>Taxonomic assignment → <code>assignTaxonomy()</code></li>
</ul>
<p>But some preliminary steps are missing…</p>
</section></section>
<section>
<section id="before-using-dada2" class="title-slide slide level1 center">
<h1>Before using Dada2</h1>

</section>
<section id="artificial-sequences" class="slide level2">
<h2>Artificial sequences</h2>

<img data-src="../img/read_example_1.svg" class="r-stretch"><p><strong>Tags</strong> are used to encode the sample provenance of the reads. Reads need to be grouped by sample provenance (demultiplexing)</p>
<p><strong>Primer</strong> sequences will bias the error model built by DADA2 and need to be removed (primer trimming).</p>
<p>Both task can be achieved using <strong>Cutadapt</strong>, a command line-tool to find and remove error-tolerantly adapters from reads (<a href="https://doi.org/10.14806/ej.17.1.200" target="_blank">Martin et al.&nbsp;2011</a>).</p>
</section>
<section id="demultiplexing-using-cutadapt" class="slide level2">
<h2>Demultiplexing using cutadapt</h2>
<p>If your tags are in a fasta file with corresponding sample names as header, you can use this command-line:</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb8"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>    <span class="at">-g</span> file:<span class="va">${BARCODES}</span> <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>    <span class="at">-o</span> {name}_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4"></a>    <span class="at">-p</span> {name}_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5"></a>    <span class="va">${R1_FILE}</span> <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6"></a>    <span class="va">${R2_FILE}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column" style="width:50%;">
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># </span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="co"># tags to look for at the beginning (5') of R1 files. ${BARCODES} is a fasta file containing the tags</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="co"># demultiplexed R1 file output. name will be replace by the name of the tag</span></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co"># same as above but with R2 files</span></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co"># input R1 file</span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co"># input R2 file</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>You end up with as many R1 and R2 files as samples you have</p>
<div class="callout callout-warning callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Help! My reads are mixed-orientated</strong></p>
</div>
<div class="callout-content">
<p>Run <code>cutadapt</code> a second time, looking for tags in R2.</p>
<p>Keep the outputs of the two rounds separated for the rest of the workflow.</p>
</div>
</div>
</div>
</section>
<section id="primer-removal-using-cutadapt" class="slide level2">
<h2>Primer removal using cutadapt</h2>
<p>To remove forward and reverse primer sequences form pair-end read files:</p>
<div class="columns">
<div class="column" style="width:50%;">
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="ex">cutadapt</span> <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>    <span class="at">-e</span> 0.1 <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3"></a>    <span class="at">-g</span> <span class="va">${PRIMER_FORWARD}</span> <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>    <span class="at">-G</span> <span class="va">${PRIMER_REVERSE}</span> <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5"></a>    <span class="at">--report</span><span class="op">=</span>minimal <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6"></a>    <span class="at">--discard-untrimmed</span> <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7"></a>    <span class="at">--minimum-length</span> <span class="va">${MIN_LENGTH}</span> <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8"></a>    <span class="at">--no-indels</span> <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9"></a>    <span class="at">-o</span> <span class="va">${FORWARD}</span> <span class="dt">\</span></span>
<span id="cb10-10"><a href="#cb10-10"></a>    <span class="at">-p</span> <span class="va">${REVERSE}</span> <span class="dt">\</span></span>
<span id="cb10-11"><a href="#cb10-11"></a>    <span class="va">${R1_FILE}</span> <span class="dt">\</span></span>
<span id="cb10-12"><a href="#cb10-12"></a>    <span class="va">${R2_FILE}</span> <span class="dv">1</span><span class="op">&gt;</span> <span class="va">${LOG}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div><div class="column" style="width:50%;">
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="co"># </span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="co"># error tolerance (default value)</span></span>
<span id="cb11-3"><a href="#cb11-3"></a><span class="co"># forward primer</span></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co"># reverse primer</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="co"># ask to print primer trimming statistics</span></span>
<span id="cb11-6"><a href="#cb11-6"></a><span class="co"># reads not containing primers are discarded</span></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="co"># read with a length below this threshold after trimming are discarded</span></span>
<span id="cb11-8"><a href="#cb11-8"></a><span class="co"># no indels allowed when mathing primer to reads</span></span>
<span id="cb11-9"><a href="#cb11-9"></a><span class="co"># R1 output</span></span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="co"># R2 output</span></span>
<span id="cb11-11"><a href="#cb11-11"></a><span class="co"># R1 input</span></span>
<span id="cb11-12"><a href="#cb11-12"></a><span class="co"># R2 input; 1&gt; ${LOG} export the report in the file ${LOG}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>As for demultiplexing, if reads are mix-orientated, run <code>cutadapt</code> twice</p>
</section></section>
<section>
<section id="dada2-workflow-read-preparation" class="title-slide slide level1 center">
<h1>DADA2 workflow: read preparation</h1>

</section>
<section id="check-reads-quality" class="slide level2">
<h2>Check reads quality</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>Check the overall quality of a sequencing run using <code>plotQualityProfile()</code></p>
<p>Outside of DADA2, you can also use <a href="https://github.com/s-andrews/FastQC" target="_blank">FASTQC</a></p>
<p>If the overall quality is too low, you will probably have to resequence your samples</p>
</div><div class="column" style="width:50%;">
<p><img data-src="../img/quality_plot_example.svg"></p>
</div>
</div>
<p>A quality drop is often observed in the end of the reads</p>
<aside class="notes">
<ul>
<li>green line: mean</li>
<li>orange line (plain): median</li>
<li>orange line (dashed): 25th and 75th quantiles</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="trimming-and-filtering" class="slide level2">
<h2>Trimming and filtering</h2>
<p>Trimming, at a given length, will improve the overall read quality</p>
<div class="callout callout-caution callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Danger zone</strong></p>
</div>
<div class="callout-content">
<p>After trimming, make sure that forward and reverse reads are still long enough to overlap</p>
</div>
</div>
</div>
<p>Reads of low quality and/or with ambiguous nucleotides (N) after trimming are filtered out.</p>
<p>Both length trimming and quality filtering are achieved using the function <code>filterAndTrim()</code></p>
</section></section>
<section>
<section id="dada2-workflow-denoising-approach" class="title-slide slide level1 center">
<h1>DADA2 workflow: denoising approach</h1>

</section>
<section id="denoising" class="slide level2">
<h2>Denoising</h2>
<p>Is sequence <span class="math inline">\(i\)</span> generated by sequence <span class="math inline">\(j\)</span> because of sequencing errors?</p>

<img data-src="../img/alignment_example_1.svg" class="r-stretch"><p>In order to define if <span class="math inline">\(i\)</span> is an error of <span class="math inline">\(j\)</span> and perform denoising using DADA2, we need to compute:</p>
<ul>
<li>the error rate <span class="math inline">\(\lambda_{ji}\)</span></li>
<li>the abundance p-value <span class="math inline">\(p_A(j \rightarrow i)\)</span></li>
</ul>
</section>
<section id="error-rate" class="slide level2">
<h2>Error rate</h2>
<blockquote>
<p>The rate at which an amplicon read with sequence i is produced from sample sequence j is reduced to the product over the transition probabilities between the L aligned nucleotides:</p>
</blockquote>
<p><span class="math display">\[\lambda_{ji} = \prod_{l=0}^L p(j(l) \rightarrow i(l),q_i(l))\]</span></p>
</section>
<section id="the-abundance-p-value" class="slide level2">
<h2>The abundance p-value</h2>
<p>The abundance p-value (<span class="math inline">\(p_A\)</span>) is the probability of all reads of <span class="math inline">\(i\)</span> (<span class="math inline">\(a_i\)</span>) or more being produced from <span class="math inline">\(j\)</span> during amplification ore sequencing.</p>
<p><span class="math display">\[p_A(j \rightarrow i) = \frac{1}{1- \rho_{\mathrm{pois}}(n_j\lambda_{ji},0)} \sum_{a=a_i}^\infty \rho_{\mathrm{pois}}(n_j\lambda_{ji},a)\]</span></p>
<p>A low p-value indicate that it is unlikely that <span class="math inline">\(i\)</span> is noise from amplifying and sequencing <span class="math inline">\(j\)</span></p>
<aside class="notes">
<p>The discrete nature of the Poisson distribution is also why this is a probability mass function and not a density function</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="the-divisive-partitioning-algorithm" class="slide level2">
<h2>The divisive partitioning algorithm</h2>
<ol type="1">
<li>all unique sequences into a single partition</li>
<li><span class="math inline">\(p_A\)</span> of all sequences against most abundant sequence (center)</li>
<li>if smallest p-value below <span class="math inline">\(\Omega_A\)</span>: new partition</li>
<li><span class="math inline">\(p_A\)</span> of all sequences against center of new partition</li>
<li>every sequence join the partition most likely to have produced it</li>
</ol>
<p>Repeat steps 3 to 5 until all abundance p-values are greater than <span class="math inline">\(\Omega_A\)</span></p>
</section></section>
<section>
<section id="dada2-workflow-denoising-in-practice" class="title-slide slide level1 center">
<h1>DADA2 workflow: denoising in practice</h1>

</section>
<section id="learn-the-error-model" class="slide level2">
<h2>Learn the error model</h2>
<p>How do we compute <span class="math inline">\(\lambda_{ji}\)</span> if we don’t know the error rate for each possible transition?</p>
<p>The error rates will be learned from the data using <code>learnErrors()</code></p>
<blockquote>
<p>The learnErrors method learns this error model from the data, by alternating estimation of the error rates and inference of sample composition until they converge on a jointly consistent solution</p>
</blockquote>
</section>
<section id="visualise-the-error-model" class="slide level2">
<h2>Visualise the error model</h2>
<p>You can visualise the estimated error rates using the function <code>plotErrors()</code></p>

<img data-src="../img/error_plot.svg" class="r-stretch quarto-figure-center"></section>
<section id="run-the-dada2-algorithm" class="slide level2">
<h2>Run the DADA2 algorithm</h2>
<ul>
<li><p>After dereplicating your sequences (<code>derepFastq()</code>), denoise using the function <code>dada()</code></p></li>
<li><p>By default sample inference is performed on each sample individually (<code>pool = FALSE</code>).</p></li>
<li><p>If you are interested in rare variants present in several samples use <code>pool = TRUE</code></p></li>
<li><p>When working on big data, <code>pool = "pseudo"</code> is an interesting alternative to <code>pool = TRUE</code></p></li>
</ul>
</section></section>
<section>
<section id="dada2-workflow-build-the-asv-table" class="title-slide slide level1 center">
<h1>DADA2 workflow: build the ASV table</h1>

</section>
<section id="merge-paired-reads" class="slide level2">
<h2>Merge paired reads</h2>
<p>Merge forward and reverse reads using <code>mergePairs()</code></p>

<img data-src="../img/merging_example.svg" class="r-stretch"><ul>
<li><code>minOverlap</code>: minimum size of overlap</li>
<li><code>maxMismatch</code>: maximum number of mismatches</li>
<li><code>justConcatenate</code>: in case your reads don’t overlap</li>
</ul>
</section>
<section id="remove-chimeras" class="slide level2">
<h2>Remove chimeras</h2>
<ul>
<li><p>Chimeras are artifact sequences formed by two or more biological sequences incorrectly joined together.</p></li>
<li><p>We find and remove bimeras (two-parent chimeras) using the function <code>removeBimeraDenovo()</code></p></li>
<li><p>Chimeric sequences are identified if they can be exactly reconstructed by combining a left-segment and a right-segment from two more abundant “parent” sequences.</p></li>
<li><p>Can be done per sample (<code>method="per-sample"</code>), for the entire dataset (<code>method="pooled"</code>) or in a consensus way (<code>method="consensus"</code>)</p></li>
</ul>
<aside class="notes">
<p>Chimeras are artifact sequences formed by two or more biological sequences incorrectly joined together. This often occurs during PCR reactions using mixed templates (i.e., uncultured environmental samples). Incomplete extensions during PCR allow subsequent PCR cycles to use a partially extended strand to bind to the template of a different, but similar, sequence. This partially extended strand then acts as a primer to extend and form a chimeric sequence. Once created, the chimeric sequence is then further amplified in subsequent cycles. The end result is a PCR artifact that does not represent a sequence that exists in nature.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="taxonomic-assignment" class="slide level2">
<h2>Taxonomic assignment</h2>
<ul>
<li><p>Taxonomic assignment within DADA2 workflow, <code>assignTaxonomy()</code>, is achieved using a native implementation of the naive Bayesian classifier method (<a href="https://doi.org/10.1128/aem.00062-07" target="_blank">Wang et al.&nbsp;2007</a>).</p></li>
<li><p>Reference fastas for the three most common 16S databases: Silva, RDP and GreenGenes are maintained by dada2’s team.</p></li>
<li><p>For eukaryotes, a dada2 formatted version of PR2 can be found <a href="https://doi.org/10.1128/aem.00062-07" target="_blank">here</a>.</p></li>
</ul>
</section>
<section id="extra-filtering-steps" class="slide level2">
<h2>Extra filtering steps</h2>
<p><br></p>
<p>ASVs can be filtered out if:</p>
<ul>
<li>sequence has an unexpected length</li>
<li>not assigned to the desired taxonomic group</li>
<li>considered as contaminant (present in negative controls)</li>
</ul>
</section></section>
<section id="now-it-is-your-turn" class="title-slide slide level1 center">
<h1>Now it is your turn!</h1>

<div class="footer footer-default">
<p><a href="../index.html"><i class="fa-solid fa-house" aria-hidden="true"></i></a></p>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>